{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Probability of no variant within k=41 bases on one side: 0.973028 (FN = 0.026972)\n",
      "Probability of no variant within k=41 bases on at least one side: 0.999273 (FN = 0.000727)\n",
      "Probability of no sequencing error within k=41 bases on one side: 0.959809 (FN = 0.040191)\n",
      "Probability of no sequencing error within k=41 bases on at least one side: 0.998385 (FN = 0.001615)\n",
      "Probability of no variant AND no sequencing error within k=41 bases on one side: 0.933896 (FN = 0.066104)\n",
      "Probability of no variant AND no sequencing error within k=41 bases on at least one side: 0.995630 (FN = 0.004370)\n",
      "Probability of detecting mutation of interest near edge: 0.972434 (FN = 0.027566)\n",
      "Probability of detecting mutation of interest near middle: 0.999272 (FN = 0.000728)\n",
      "Probability a single read is with a linked variant is usable: 0.698160\n",
      "Probability of getting at least 3 usable reads out of 5 reads with linked variant: 0.834477 (FN = 0.165523)\n",
      "\n",
      "Probability of no variant within k=47 bases on one side: 0.969142 (FN = 0.030858)\n",
      "Probability of no variant within k=47 bases on at least one side: 0.999048 (FN = 0.000952)\n",
      "Probability of no sequencing error within k=47 bases on one side: 0.954065 (FN = 0.045935)\n",
      "Probability of no sequencing error within k=47 bases on at least one side: 0.997890 (FN = 0.002110)\n",
      "Probability of no variant AND no sequencing error within k=47 bases on one side: 0.924596 (FN = 0.075404)\n",
      "Probability of no variant AND no sequencing error within k=47 bases on at least one side: 0.994314 (FN = 0.005686)\n",
      "Probability of detecting mutation of interest near edge: 0.968267 (FN = 0.031733)\n",
      "Probability of detecting mutation of interest near middle: 0.999048 (FN = 0.000952)\n",
      "Probability a single read is with a linked variant is usable: 0.655780\n",
      "Probability of getting at least 3 usable reads out of 5 reads with linked variant: 0.773737 (FN = 0.226263)\n",
      "\n",
      "Probability of no variant within k=51 bases on one side: 0.966561 (FN = 0.033439)\n",
      "Probability of no variant within k=51 bases on at least one side: 0.998882 (FN = 0.001118)\n",
      "Probability of no sequencing error within k=51 bases on one side: 0.950254 (FN = 0.049746)\n",
      "Probability of no sequencing error within k=51 bases on at least one side: 0.997525 (FN = 0.002475)\n",
      "Probability of no variant AND no sequencing error within k=51 bases on one side: 0.918447 (FN = 0.081553)\n",
      "Probability of no variant AND no sequencing error within k=51 bases on at least one side: 0.993349 (FN = 0.006651)\n",
      "Probability of detecting mutation of interest near edge: 0.965458 (FN = 0.034542)\n",
      "Probability of detecting mutation of interest near middle: 0.998882 (FN = 0.001118)\n",
      "Probability a single read is with a linked variant is usable: 0.627796\n",
      "Probability of getting at least 3 usable reads out of 5 reads with linked variant: 0.729386 (FN = 0.270614)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import binom\n",
    "\n",
    "ks = [41, 47, 51]\n",
    "read_coverage = 5\n",
    "count_threshold = 3\n",
    "variant_rate = 1 / 1500\n",
    "error_rate = 1 / 1000\n",
    "read_length = 150\n",
    "\n",
    "p = variant_rate + error_rate  # for general case\n",
    "q = 1 - p\n",
    "\n",
    "def calculate_probalities(p, k):\n",
    "    prob_no_variant_when_mutation_of_interest_is_on_end = (1 - p) ** k\n",
    "    prob_clean_flank_both = (1 - p) ** (2 * k)\n",
    "    prob_variant_on_both_sides = 1 - 2 * prob_no_variant_when_mutation_of_interest_is_on_end + prob_clean_flank_both\n",
    "    prob_no_variant_within_k_when_mutation_of_interest_is_in_middle = 1 - prob_variant_on_both_sides\n",
    "\n",
    "    return prob_no_variant_when_mutation_of_interest_is_on_end, prob_no_variant_within_k_when_mutation_of_interest_is_in_middle\n",
    "\n",
    "for k in ks:\n",
    "    prob_no_variant_when_mutation_of_interest_is_on_end, prob_no_variant_within_k_when_mutation_of_interest_is_in_middle = calculate_probalities(variant_rate, k)\n",
    "    prob_no_sequencing_error_when_mutation_of_interest_is_on_end, prob_no_sequencing_error_within_k_when_mutation_of_interest_is_in_middle = calculate_probalities(error_rate, k)\n",
    "    prob_no_variant_OR_sequencing_error_when_mutation_of_interest_is_on_end, prob_no_variant_OR_sequencing_error_within_k_when_mutation_of_interest_is_in_middle = calculate_probalities(p, k)\n",
    "\n",
    "    # Edge case: no variant in flank, only sequencing error matters\n",
    "    prob_that_COUNT_THRESHOLD_out_of_READ_COVERAGE_reads_have_no_sequencing_error_flank = float(binom.sf(count_threshold - 1, read_coverage, prob_no_sequencing_error_when_mutation_of_interest_is_on_end))\n",
    "    prob_mutation_of_interest_is_detected_all_flank = prob_no_variant_when_mutation_of_interest_is_on_end * prob_that_COUNT_THRESHOLD_out_of_READ_COVERAGE_reads_have_no_sequencing_error_flank\n",
    "\n",
    "    # Middle case: no variant in flank, only sequencing error matters\n",
    "    prob_that_COUNT_THRESHOLD_out_of_READ_COVERAGE_reads_have_no_sequencing_error_both = float(binom.sf(count_threshold - 1, read_coverage, prob_no_sequencing_error_within_k_when_mutation_of_interest_is_in_middle))\n",
    "    prob_mutation_of_interest_is_detected_all_middle = prob_no_variant_within_k_when_mutation_of_interest_is_in_middle * prob_that_COUNT_THRESHOLD_out_of_READ_COVERAGE_reads_have_no_sequencing_error_both\n",
    "\n",
    "    # What if we have linked variants -- then we need (1) at least count_threshold reads will have the variant in position between (1, read_length-k) and (2) at least count_threshold reads will have no sequencing error in the followin k-1 bases (assumes linked variant is before variant of interest -- symmetrical in other case)\n",
    "    P_pos = (read_length - k) / read_length  # Probability that variant is in a valid position in the read\n",
    "    P_variant = variant_rate\n",
    "    P_no_error = (1 - error_rate) ** (k - 1)  # Probability that the following k-1 bases are error-free\n",
    "    P_total = P_pos * P_no_error  # Combined probability that a read is usable (has variant in valid position and no sequencing error in the following k-1 bases)\n",
    "    prob_at_least_threshold = 1 - binom.cdf(count_threshold - 1, read_coverage, P_total)  # Step 4: Binomial model â€” probability of getting >= count_threshold usable reads\n",
    "\n",
    "    print(f\"Probability of no variant within k={k} bases on one side: {prob_no_variant_when_mutation_of_interest_is_on_end:.6f} (FN = {1 - prob_no_variant_when_mutation_of_interest_is_on_end:.6f})\")\n",
    "    print(f\"Probability of no variant within k={k} bases on at least one side: {prob_no_variant_within_k_when_mutation_of_interest_is_in_middle:.6f} (FN = {1 - prob_no_variant_within_k_when_mutation_of_interest_is_in_middle:.6f})\")\n",
    "    print(f\"Probability of no sequencing error within k={k} bases on one side: {prob_no_sequencing_error_when_mutation_of_interest_is_on_end:.6f} (FN = {1 - prob_no_sequencing_error_when_mutation_of_interest_is_on_end:.6f})\")\n",
    "    print(f\"Probability of no sequencing error within k={k} bases on at least one side: {prob_no_sequencing_error_within_k_when_mutation_of_interest_is_in_middle:.6f} (FN = {1 - prob_no_sequencing_error_within_k_when_mutation_of_interest_is_in_middle:.6f})\")\n",
    "    print(f\"Probability of no variant AND no sequencing error within k={k} bases on one side: {prob_no_variant_OR_sequencing_error_when_mutation_of_interest_is_on_end:.6f} (FN = {1 - prob_no_variant_OR_sequencing_error_when_mutation_of_interest_is_on_end:.6f})\")\n",
    "    print(f\"Probability of no variant AND no sequencing error within k={k} bases on at least one side: {prob_no_variant_OR_sequencing_error_within_k_when_mutation_of_interest_is_in_middle:.6f} (FN = {1 - prob_no_variant_OR_sequencing_error_within_k_when_mutation_of_interest_is_in_middle:.6f})\")\n",
    "    print(f\"Probability of detecting mutation of interest near edge: {prob_mutation_of_interest_is_detected_all_flank:.6f} (FN = {1 - prob_mutation_of_interest_is_detected_all_flank:.6f})\")\n",
    "    print(f\"Probability of detecting mutation of interest near middle: {prob_mutation_of_interest_is_detected_all_middle:.6f} (FN = {1 - prob_mutation_of_interest_is_detected_all_middle:.6f})\")\n",
    "    print(f\"Probability a single read is with a linked variant is usable: {P_total:.6f}\")\n",
    "    print(f\"Probability of getting at least {count_threshold} usable reads out of {read_coverage} reads with linked variant: {prob_at_least_threshold:.6f} (FN = {1 - prob_at_least_threshold:.6f})\")\n",
    "    print()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RLSRP_2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
