{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8a4b538-0583-45e9-91d1-eec859bae4c1",
   "metadata": {},
   "source": [
    "# Extract cancer variant reads for alignment visualization\n",
    "\n",
    "## In this notebook, we use a single RNA-seq fastq file from a melanoma cancer cell line from the the CCLE project. Learn more about the project here:\n",
    "- sequencing data (ENA): https://www.ebi.ac.uk/ena/browser/view/PRJNA523380\n",
    "- paper: https://www.nature.com/articles/s41586-019-1186-3\n",
    "\n",
    "### Requirements: kb, samtools, and bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9cae3de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import varseek as vk\n",
    "except ImportError:\n",
    "    print(\"varseek not found, installing...\")\n",
    "    !pip install -U -q varseek\n",
    "try:\n",
    "    import RLSRWP_2025\n",
    "except ImportError:\n",
    "    print(\"RLSRWP_2025 not found, installing...\")\n",
    "    !pip install -q git+https://github.com/pachterlab/RLSRWP_2025.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "400e7db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gget\n",
    "import pyfastx\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "# pd.set_option('display.max_columns', None)\n",
    "import varseek as vk\n",
    "from varseek.utils import make_bus_df, add_variant_type\n",
    "\n",
    "RLSRWP_2025_dir = os.path.dirname(os.path.abspath(\"\"))  # if this notebook resides in RLSRWP_2025/notebooks/0_data_download.ipynb, then this retrieves RLSRWP_2025\n",
    "threads = \"2\"\n",
    "use_head = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "317e1ef9",
   "metadata": {},
   "source": [
    "### File path definitions and imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81b187a",
   "metadata": {},
   "outputs": [],
   "source": [
    "notebook_out_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"varseek_count_out_alignment_visualization_full\")\n",
    "kb_count_out_dir = os.path.join(notebook_out_dir, \"kb_count_out_vcrs\")\n",
    "adata_path = os.path.join(kb_count_out_dir, \"counts_unfiltered\", \"adata.h5ad\")\n",
    "aligned_reads_parent_dir = os.path.join(notebook_out_dir, \"pseudoaligned_reads_to_vcrs_reference\")\n",
    "bowtie_read_alignments = os.path.join(notebook_out_dir, \"bowtie_read_alignments\")\n",
    "\n",
    "# vk ref out directory and files - downloaded if not already present\n",
    "vk_ref_out_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"vk_ref_out\")\n",
    "vcrs_index = os.path.join(vk_ref_out_dir, \"vcrs_index.idx\")\n",
    "vcrs_t2g = os.path.join(vk_ref_out_dir, \"vcrs_t2g_filtered.txt\")\n",
    "vcrs_fasta = os.path.join(vk_ref_out_dir, \"vcrs_filtered.fa\")\n",
    "\n",
    "# fastq directories - fastqs_dir downloaded if not already present, and fastqs_processed_dir created with fastp if not already present\n",
    "fastqs_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"ccle_data_base\", \"RNASeq_MELHO_SKIN\")\n",
    "fastq_file = os.path.join(fastqs_dir, \"SRR8615233_1.fastq.gz\")\n",
    "technology = \"bulk\"\n",
    "\n",
    "# cosmic directory\n",
    "cosmic_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"reference\", \"cosmic\")\n",
    "cosmic_csv = os.path.join(cosmic_dir, \"CancerMutationCensus_AllData_Tsv_v101_GRCh37\", \"CancerMutationCensus_AllData_v101_GRCh37_mutation_workflow.csv\")\n",
    "\n",
    "# reference genome directory\n",
    "reference_genome_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"reference\", \"ensembl_grch37_release93\")\n",
    "\n",
    "# kb count to reference genome directory and files - created if not already present - only used if qc_against_gene_matrix=True\n",
    "qc_against_gene_matrix = False\n",
    "kb_count_reference_genome_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"kb_count_reference_genome\")\n",
    "reference_genome_index = os.path.join(reference_genome_dir, \"index.idx\")  # either already exists or will be created\n",
    "reference_genome_t2g = os.path.join(reference_genome_dir, \"t2g.txt\")  # either already exists or will be created\n",
    "reference_genome_fasta = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.dna.primary_assembly.fa\")  # if reference_genome_index/reference_genome_t2g do not exist, then I need to supply the reference genome fasta and gtf\n",
    "reference_genome_gtf = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.87.gtf\")  # if reference_genome_index/reference_genome_t2g do not exist, then I need to supply the reference genome fasta and gtf\n",
    "\n",
    "# for bowtie2 - also uses reference_genome_fasta above (I could use cdna instead of genome too if desired)\n",
    "bowtie_reference_dir = os.path.join(reference_genome_dir, \"bowtie_index_genome\")\n",
    "bowtie_reference_prefix = os.path.join(bowtie_reference_dir, \"index\")\n",
    "\n",
    "# general\n",
    "w = \"47\"  # used during creation of the index, so cannot be altered\n",
    "k = \"51\"  # used during creation of the index, so cannot be altered\n",
    "strand = \"unstranded\"\n",
    "parity = \"single\"  # although the original data is paired, we will only be using a single file, so we will run in single-end mode\n",
    "\n",
    "# software\n",
    "bustools = \"bustools\"\n",
    "bowtie2 = \"bowtie2\"\n",
    "bowtie2_build = \"bowtie2-build\"\n",
    "samtools = \"samtools\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f50ee94",
   "metadata": {},
   "source": [
    "### Download VCRS reference files with varseek ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ebc5dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(vcrs_index) or not os.path.exists(vcrs_t2g) or not os.path.exists(vcrs_fasta):\n",
    "    vk.ref(variants=\"cosmic_cmc\", sequences=\"cdna\", w=w, k=k, dlist_reference_source=\"t2t\", index_out=vcrs_index, t2g_out=vcrs_t2g, fasta_out=vcrs_fasta, download=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ffe24bf",
   "metadata": {},
   "source": [
    "### Download COSMIC if not already downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "184711a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/1725634623.py:13: DtypeWarning: Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  cosmic_df = pd.read_csv(cosmic_csv)\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "b1c09fa7-7b3f-4f5e-9019-dfb1c932fe39",
       "rows": [
        [
         "0",
         "ENST00000396153",
         "c.1468C>T",
         "p.Q490*",
         "COSV105286190",
         "1",
         "-",
         "g.51001067G>A",
         "FAF1",
         "48293736",
         "c.1920C>T",
         "ENST00000396153:c.1920C>T"
        ],
        [
         "1",
         "ENST00000445907",
         "c.162C>A",
         "p.L54=",
         "COSV57765075",
         "7",
         "+",
         "g.136699774C>A",
         "CHRM2",
         "60222372",
         "c.690C>A",
         "ENST00000445907:c.690C>A"
        ],
        [
         "2",
         "ENST00000445907",
         "c.617A>T",
         "p.Y206F",
         "COSV108138517",
         "7",
         "+",
         "g.136700229A>T",
         "CHRM2",
         "60229878",
         "c.1145A>T",
         "ENST00000445907:c.1145A>T"
        ],
        [
         "3",
         "ENST00000445907",
         "c.93A>T",
         "p.G31=",
         "COSV100282253",
         "7",
         "+",
         "g.136699705A>T",
         "CHRM2",
         "60227007",
         "c.621A>T",
         "ENST00000445907:c.621A>T"
        ],
        [
         "4",
         "ENST00000445907",
         "c.956A>C",
         "p.N319T",
         "COSV108138508",
         "7",
         "+",
         "g.136700568A>C",
         "CHRM2",
         "60213402",
         "c.1484A>C",
         "ENST00000445907:c.1484A>C"
        ]
       ],
       "shape": {
        "columns": 11,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000396153</td>\n",
       "      <td>c.1468C&gt;T</td>\n",
       "      <td>p.Q490*</td>\n",
       "      <td>COSV105286190</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.51001067G&gt;A</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>48293736</td>\n",
       "      <td>c.1920C&gt;T</td>\n",
       "      <td>ENST00000396153:c.1920C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.162C&gt;A</td>\n",
       "      <td>p.L54=</td>\n",
       "      <td>COSV57765075</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699774C&gt;A</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60222372</td>\n",
       "      <td>c.690C&gt;A</td>\n",
       "      <td>ENST00000445907:c.690C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.617A&gt;T</td>\n",
       "      <td>p.Y206F</td>\n",
       "      <td>COSV108138517</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700229A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60229878</td>\n",
       "      <td>c.1145A&gt;T</td>\n",
       "      <td>ENST00000445907:c.1145A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.93A&gt;T</td>\n",
       "      <td>p.G31=</td>\n",
       "      <td>COSV100282253</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699705A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60227007</td>\n",
       "      <td>c.621A&gt;T</td>\n",
       "      <td>ENST00000445907:c.621A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.956A&gt;C</td>\n",
       "      <td>p.N319T</td>\n",
       "      <td>COSV108138508</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700568A&gt;C</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60213402</td>\n",
       "      <td>c.1484A&gt;C</td>\n",
       "      <td>ENST00000445907:c.1484A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seq_ID   mutation mutation_aa GENOMIC_MUTATION_ID chromosome  \\\n",
       "0  ENST00000396153  c.1468C>T     p.Q490*       COSV105286190          1   \n",
       "1  ENST00000445907   c.162C>A      p.L54=        COSV57765075          7   \n",
       "2  ENST00000445907   c.617A>T     p.Y206F       COSV108138517          7   \n",
       "3  ENST00000445907    c.93A>T      p.G31=       COSV100282253          7   \n",
       "4  ENST00000445907   c.956A>C     p.N319T       COSV108138508          7   \n",
       "\n",
       "  strand mutation_genome gene_name  mutation_id mutation_cdna  \\\n",
       "0      -   g.51001067G>A      FAF1     48293736     c.1920C>T   \n",
       "1      +  g.136699774C>A     CHRM2     60222372      c.690C>A   \n",
       "2      +  g.136700229A>T     CHRM2     60229878     c.1145A>T   \n",
       "3      +  g.136699705A>T     CHRM2     60227007      c.621A>T   \n",
       "4      +  g.136700568A>C     CHRM2     60213402     c.1484A>C   \n",
       "\n",
       "                      header  \n",
       "0  ENST00000396153:c.1920C>T  \n",
       "1   ENST00000445907:c.690C>A  \n",
       "2  ENST00000445907:c.1145A>T  \n",
       "3   ENST00000445907:c.621A>T  \n",
       "4  ENST00000445907:c.1484A>C  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if cosmic_csv and not os.path.isfile(cosmic_csv):\n",
    "    gget.cosmic(\n",
    "        None,\n",
    "        grch_version=37,\n",
    "        cosmic_version=101,\n",
    "        out=cosmic_dir,\n",
    "        cosmic_project=\"cancer\",\n",
    "        download_cosmic=True,\n",
    "        gget_mutate=True,\n",
    "        keep_genome_info=True,\n",
    "        remove_duplicates=True\n",
    "    )\n",
    "\n",
    "cosmic_df = pd.read_csv(cosmic_csv)\n",
    "\n",
    "if \"mutation_cdna\" not in cosmic_df.columns:\n",
    "    reference_cdna_path = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.cdna.all.fa\")\n",
    "    reference_cds_path = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.cds.all.fa\")\n",
    "    if not os.path.exists(reference_cdna_path):\n",
    "        reference_cdna_dir = os.path.dirname(reference_cdna_path) if os.path.dirname(reference_cdna_path) else \".\"\n",
    "        !gget ref -w cdna -r 93 --out_dir {reference_cdna_dir} -d human_grch37\n",
    "        !gunzip {reference_cdna_path}.gz\n",
    "    if not os.path.exists(reference_cds_path):\n",
    "        reference_cds_dir = os.path.dirname(reference_cds_path) if os.path.dirname(reference_cds_path) else \".\"\n",
    "        !gget ref -w cdna -r 93 --out_dir {reference_cds_dir} -d human_grch37\n",
    "        !gunzip {reference_cds_path}.gz\n",
    "    print(\"Converting CDS to cDNA in COSMIC\")\n",
    "    cosmic_df, _ = vk.utils.convert_mutation_cds_locations_to_cdna(input_csv_path=cosmic_df, output_csv_path=cosmic_csv, cds_fasta_path=reference_cds_path, cdna_fasta_path=reference_cdna_path, verbose=True, strip_leading_Ns_cds=True)\n",
    "\n",
    "if \"variant_type\" not in cosmic_df.columns:\n",
    "    add_variant_type(cosmic_df, \"mutation_cdna\")\n",
    "\n",
    "if \"header\" not in cosmic_df.columns:\n",
    "    cosmic_df[\"header\"] = cosmic_df[\"seq_ID\"] + \":\" + cosmic_df[\"mutation_cdna\"]\n",
    "    cosmic_df.to_csv(cosmic_csv, index=False)\n",
    "\n",
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "492c02ce",
   "metadata": {},
   "source": [
    "### Make bowtie2 index files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17451c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(bowtie_reference_dir) or len(os.listdir(bowtie_reference_dir)) == 0:\n",
    "    os.makedirs(bowtie_reference_dir, exist_ok=True)\n",
    "    !{bowtie2_build} --threads {threads} {reference_genome_fasta} {bowtie_reference_prefix}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e3028e5",
   "metadata": {},
   "source": [
    "### Download fastq file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e2bae66",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isfile(fastq_file):\n",
    "    fastq_file_link = \"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\"  # [\"ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\", \"ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_2.fastq.gz\"]\n",
    "    os.makedirs(fastqs_dir, exist_ok=True)\n",
    "    !wget -c --tries=5 --retry-connrefused -O {fastq_file} {fastq_file_link}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dc6d8e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if use_head:\n",
    "    fastq_file_head = fastq_file.replace(\".fastq.gz\", \"_head.fastq\")\n",
    "    if not os.path.isfile(fastq_file_head):\n",
    "        !zcat {fastq_file} | head -1000000 > $fastq_file_head\n",
    "    fastq_file = fastq_file_head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "907dad2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not fastq_file.endswith(\".gz\"):\n",
    "    !head {fastq_file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202970af",
   "metadata": {},
   "source": [
    "### Make pyfastx index files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a971147c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_indexed = pyfastx.Fastq(fastq_file, build_index=True)\n",
    "vcrs_fasta_indexed = pyfastx.Fasta(vcrs_fasta, build_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4fc8333",
   "metadata": {},
   "source": [
    "### Make kb reference files if qc_against_gene_matrix=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5fc75df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "kb_count_reference_genome_adata = os.path.join(kb_count_reference_genome_dir, \"counts_unfiltered\", \"adata.h5ad\")\n",
    "if qc_against_gene_matrix and not os.path.exists(kb_count_reference_genome_adata):  # check if kb count was run\n",
    "    os.makedirs(kb_count_reference_genome_dir, exist_ok=True)\n",
    "    if not os.path.exists(reference_genome_index) or not os.path.exists(reference_genome_t2g):  # check if kb ref was run\n",
    "        if not os.path.exists(reference_genome_fasta) or not os.path.exists(reference_genome_gtf):\n",
    "            reference_genome_out_dir = os.path.dirname(reference_genome_fasta) if reference_genome_fasta else \".\"\n",
    "            # using grch37, ensembl 93 to agree with COSMIC\n",
    "            !gget ref -w dna,gtf -r 93 --out_dir {reference_genome_out_dir} -d human_grch37 && gunzip {reference_genome_fasta}.gz && gunzip {reference_genome_gtf}.gz\n",
    "        reference_genome_f1 = os.path.join(kb_count_reference_genome_dir, \"f1.fa\")\n",
    "        !kb ref -t {threads} -i {reference_genome_index} -g {reference_genome_t2g} -f1 {reference_genome_f1} {reference_genome_fasta} {reference_genome_gtf}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70df8866",
   "metadata": {},
   "source": [
    "### Perform variant screening with varseek count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca71a00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(adata_path):\n",
    "    vk_count_output_dict = vk.count(\n",
    "        fastq_file,\n",
    "        index=vcrs_index,\n",
    "        t2g=vcrs_t2g,\n",
    "        technology=technology,\n",
    "        out=notebook_out_dir,\n",
    "        k=k,\n",
    "        strand=strand,\n",
    "        parity=parity,\n",
    "        threads=threads,\n",
    "        mm=False,\n",
    "        union=False,\n",
    "        disable_fastqpp=True,\n",
    "        disable_clean=True,\n",
    "        # quality_control_fastqs=True, cut_front=True, cut_tail=True  # equivalent to the fastp step above\n",
    "        # qc_against_gene_matrix=qc_against_gene_matrix, reference_genome_index=reference_genome_index, reference_genome_t2g=reference_genome_t2g,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6038ec84",
   "metadata": {},
   "source": [
    "### Load in adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "0dc1900e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1 × 5329695"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(adata_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fef39c0",
   "metadata": {},
   "source": [
    "### View top non-mitochondrial 15 variants by total reads aligned - 5 substitution, 5 deletion, 5 insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ec5f4dbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "efea39e8-2063-4379-a288-e7e4865fa52e",
       "rows": [
        [
         "0",
         "ENST00000396153",
         "c.1468C>T",
         "p.Q490*",
         "COSV105286190",
         "1",
         "-",
         "g.51001067G>A",
         "FAF1",
         "48293736",
         "c.1920C>T",
         "ENST00000396153:c.1920C>T"
        ],
        [
         "1",
         "ENST00000445907",
         "c.162C>A",
         "p.L54=",
         "COSV57765075",
         "7",
         "+",
         "g.136699774C>A",
         "CHRM2",
         "60222372",
         "c.690C>A",
         "ENST00000445907:c.690C>A"
        ],
        [
         "2",
         "ENST00000445907",
         "c.617A>T",
         "p.Y206F",
         "COSV108138517",
         "7",
         "+",
         "g.136700229A>T",
         "CHRM2",
         "60229878",
         "c.1145A>T",
         "ENST00000445907:c.1145A>T"
        ],
        [
         "3",
         "ENST00000445907",
         "c.93A>T",
         "p.G31=",
         "COSV100282253",
         "7",
         "+",
         "g.136699705A>T",
         "CHRM2",
         "60227007",
         "c.621A>T",
         "ENST00000445907:c.621A>T"
        ],
        [
         "4",
         "ENST00000445907",
         "c.956A>C",
         "p.N319T",
         "COSV108138508",
         "7",
         "+",
         "g.136700568A>C",
         "CHRM2",
         "60213402",
         "c.1484A>C",
         "ENST00000445907:c.1484A>C"
        ]
       ],
       "shape": {
        "columns": 11,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000396153</td>\n",
       "      <td>c.1468C&gt;T</td>\n",
       "      <td>p.Q490*</td>\n",
       "      <td>COSV105286190</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.51001067G&gt;A</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>48293736</td>\n",
       "      <td>c.1920C&gt;T</td>\n",
       "      <td>ENST00000396153:c.1920C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.162C&gt;A</td>\n",
       "      <td>p.L54=</td>\n",
       "      <td>COSV57765075</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699774C&gt;A</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60222372</td>\n",
       "      <td>c.690C&gt;A</td>\n",
       "      <td>ENST00000445907:c.690C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.617A&gt;T</td>\n",
       "      <td>p.Y206F</td>\n",
       "      <td>COSV108138517</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700229A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60229878</td>\n",
       "      <td>c.1145A&gt;T</td>\n",
       "      <td>ENST00000445907:c.1145A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.93A&gt;T</td>\n",
       "      <td>p.G31=</td>\n",
       "      <td>COSV100282253</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699705A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60227007</td>\n",
       "      <td>c.621A&gt;T</td>\n",
       "      <td>ENST00000445907:c.621A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.956A&gt;C</td>\n",
       "      <td>p.N319T</td>\n",
       "      <td>COSV108138508</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700568A&gt;C</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60213402</td>\n",
       "      <td>c.1484A&gt;C</td>\n",
       "      <td>ENST00000445907:c.1484A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seq_ID   mutation mutation_aa GENOMIC_MUTATION_ID chromosome  \\\n",
       "0  ENST00000396153  c.1468C>T     p.Q490*       COSV105286190          1   \n",
       "1  ENST00000445907   c.162C>A      p.L54=        COSV57765075          7   \n",
       "2  ENST00000445907   c.617A>T     p.Y206F       COSV108138517          7   \n",
       "3  ENST00000445907    c.93A>T      p.G31=       COSV100282253          7   \n",
       "4  ENST00000445907   c.956A>C     p.N319T       COSV108138508          7   \n",
       "\n",
       "  strand mutation_genome gene_name  mutation_id mutation_cdna  \\\n",
       "0      -   g.51001067G>A      FAF1     48293736     c.1920C>T   \n",
       "1      +  g.136699774C>A     CHRM2     60222372      c.690C>A   \n",
       "2      +  g.136700229A>T     CHRM2     60229878     c.1145A>T   \n",
       "3      +  g.136699705A>T     CHRM2     60227007      c.621A>T   \n",
       "4      +  g.136700568A>C     CHRM2     60213402     c.1484A>C   \n",
       "\n",
       "                      header  \n",
       "0  ENST00000396153:c.1920C>T  \n",
       "1   ENST00000445907:c.690C>A  \n",
       "2  ENST00000445907:c.1145A>T  \n",
       "3   ENST00000445907:c.621A>T  \n",
       "4  ENST00000445907:c.1484A>C  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "dd33e1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map cosmic_df['gene_name'] to adata.var using 'header'\n",
    "adata.var['header'] = adata.var_names\n",
    "cosmic_df_gene_map = cosmic_df.drop_duplicates('header').set_index('header')[['gene_name', 'variant_type']]\n",
    "adata.var = adata.var.merge(cosmic_df_gene_map, on='header', how='left').set_index(adata.var.index)\n",
    "adata.var.index = adata.var['header']\n",
    "\n",
    "# Filter out columns where gene_name starts with \"mt-\" (case-insensitive)\n",
    "mask = ~adata.var['gene_name'].astype(str).str.upper().str.startswith('MT-', na=False)\n",
    "adata = adata[:, mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "6bb6edfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Variant",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Counts",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Gene_Name",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "a4c9a94b-912a-4f64-86db-5e0d0897eef3",
       "rows": [
        [
         "0",
         "ENST00000398606:c.562A>G",
         "13934",
         "GSTP1"
        ],
        [
         "1",
         "ENST00000449260:c.1988C>T",
         "12589",
         "PMEL"
        ],
        [
         "2",
         "ENST00000251453:c.80T>G",
         "11981",
         "RPS16"
        ],
        [
         "3",
         "ENST00000398606:c.804T>C",
         "11300",
         "GSTP1"
        ],
        [
         "4",
         "ENST00000343262:c.156C>T",
         "6296",
         "RPS2"
        ],
        [
         "5",
         "ENST00000322723:c.1024_1026del",
         "2501",
         "NCL"
        ],
        [
         "6",
         "ENST00000263574:c.750_755del",
         "254",
         "APLP2"
        ],
        [
         "7",
         "ENST00000367034:c.152_154del",
         "216",
         "MRPL18"
        ],
        [
         "8",
         "ENST00000314222:c.448_449insCCCCC",
         "145",
         "PHLDA2"
        ],
        [
         "9",
         "ENST00000373477:c.2126_2127insCGTAC",
         "145",
         "YARS"
        ],
        [
         "10",
         "ENST00000434724:c.239_247del",
         "115",
         "GSPT1"
        ],
        [
         "11",
         "ENST00000529010:c.1426_1428del",
         "108",
         "NUCB2"
        ],
        [
         "12",
         "ENST00000262710:c.2258_2259insATTCAC",
         "82",
         "ACIN1"
        ],
        [
         "13",
         "ENST00000372797:c.1497_1498insG",
         "27",
         "CAP1"
        ],
        [
         "14",
         "ENST00000435765:c.1262_1263insGTT",
         "26",
         "PSMC2"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 15
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Counts</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000398606:c.562A&gt;G</td>\n",
       "      <td>13934</td>\n",
       "      <td>GSTP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000449260:c.1988C&gt;T</td>\n",
       "      <td>12589</td>\n",
       "      <td>PMEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "      <td>11981</td>\n",
       "      <td>RPS16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000398606:c.804T&gt;C</td>\n",
       "      <td>11300</td>\n",
       "      <td>GSTP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000343262:c.156C&gt;T</td>\n",
       "      <td>6296</td>\n",
       "      <td>RPS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000322723:c.1024_1026del</td>\n",
       "      <td>2501</td>\n",
       "      <td>NCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000263574:c.750_755del</td>\n",
       "      <td>254</td>\n",
       "      <td>APLP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000367034:c.152_154del</td>\n",
       "      <td>216</td>\n",
       "      <td>MRPL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000314222:c.448_449insCCCCC</td>\n",
       "      <td>145</td>\n",
       "      <td>PHLDA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000373477:c.2126_2127insCGTAC</td>\n",
       "      <td>145</td>\n",
       "      <td>YARS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000434724:c.239_247del</td>\n",
       "      <td>115</td>\n",
       "      <td>GSPT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000529010:c.1426_1428del</td>\n",
       "      <td>108</td>\n",
       "      <td>NUCB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000262710:c.2258_2259insATTCAC</td>\n",
       "      <td>82</td>\n",
       "      <td>ACIN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000372797:c.1497_1498insG</td>\n",
       "      <td>27</td>\n",
       "      <td>CAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000435765:c.1262_1263insGTT</td>\n",
       "      <td>26</td>\n",
       "      <td>PSMC2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Variant  Counts Gene_Name\n",
       "0               ENST00000398606:c.562A>G   13934     GSTP1\n",
       "1              ENST00000449260:c.1988C>T   12589      PMEL\n",
       "2                ENST00000251453:c.80T>G   11981     RPS16\n",
       "3               ENST00000398606:c.804T>C   11300     GSTP1\n",
       "4               ENST00000343262:c.156C>T    6296      RPS2\n",
       "5         ENST00000322723:c.1024_1026del    2501       NCL\n",
       "6           ENST00000263574:c.750_755del     254     APLP2\n",
       "7           ENST00000367034:c.152_154del     216    MRPL18\n",
       "8      ENST00000314222:c.448_449insCCCCC     145    PHLDA2\n",
       "9    ENST00000373477:c.2126_2127insCGTAC     145      YARS\n",
       "10          ENST00000434724:c.239_247del     115     GSPT1\n",
       "11        ENST00000529010:c.1426_1428del     108     NUCB2\n",
       "12  ENST00000262710:c.2258_2259insATTCAC      82     ACIN1\n",
       "13       ENST00000372797:c.1497_1498insG      27      CAP1\n",
       "14     ENST00000435765:c.1262_1263insGTT      26     PSMC2"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_to_select_per_mutation_type = 5\n",
    "top15, top15_counts, top15_gene_names = [], [], []\n",
    "\n",
    "for mutation_type in [\"substitution\", \"insertion\", \"deletion\"]:\n",
    "    adata_tmp = adata[:, adata.var['variant_type'] == mutation_type].copy()  # select only mutation_type\n",
    "    variants_list = list(adata_tmp.var.index[np.argsort(np.array(adata_tmp.X.todense())[0])][::-1][:number_to_select_per_mutation_type])\n",
    "    top15.extend(variants_list)\n",
    "    top15_counts.extend(list(np.array(adata_tmp.X.todense())[0][np.argsort(np.array(adata_tmp.X.todense())[0])][::-1][:number_to_select_per_mutation_type].astype(int)))\n",
    "    top15_gene_names.extend(cosmic_df.set_index('header').loc[variants_list, 'gene_name'].tolist())\n",
    "\n",
    "top15_dict = dict(zip(top15, top15_counts))\n",
    "\n",
    "top15_df = pd.DataFrame({\n",
    "    'Variant': top15,\n",
    "    'Counts': top15_counts,\n",
    "    'Gene_Name': top15_gene_names\n",
    "})\n",
    "\n",
    "top15 = top15_df.sort_values(by=\"Counts\", ascending=False)[\"Variant\"].tolist()  # sort descending order\n",
    "top15_df = top15_df.sort_values(by=\"Counts\", ascending=False).reset_index(drop=True)\n",
    "top15_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baee92d7",
   "metadata": {},
   "source": [
    "### Map the reads to the VCRS to which they aligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2cf1c7a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/3798617008.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_bus_df[\"variant_names_str\"] = filtered_bus_df[\"variant_names\"].apply(lambda x: x[0])  # cast to string\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reads in filtered bus file: 59719\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "barcode",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "UMI",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "EC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "read_index",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fastq_header",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "file_index",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "variant_names",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "counted_in_count_matrix",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "variant_names_str",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "bc4848d0-a6ef-494d-a8a3-a0e6b68ae26c",
       "rows": [
        [
         "9665",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917225",
         "SRR8615233.917226",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9666",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917227",
         "SRR8615233.917228",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9667",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917230",
         "SRR8615233.917231",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9668",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917231",
         "SRR8615233.917232",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9669",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917232",
         "SRR8615233.917233",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ]
       ],
       "shape": {
        "columns": 9,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>UMI</th>\n",
       "      <th>EC</th>\n",
       "      <th>read_index</th>\n",
       "      <th>fastq_header</th>\n",
       "      <th>file_index</th>\n",
       "      <th>variant_names</th>\n",
       "      <th>counted_in_count_matrix</th>\n",
       "      <th>variant_names_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9665</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917225</td>\n",
       "      <td>SRR8615233.917226</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9666</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917227</td>\n",
       "      <td>SRR8615233.917228</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9667</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917230</td>\n",
       "      <td>SRR8615233.917231</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9668</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917231</td>\n",
       "      <td>SRR8615233.917232</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9669</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917232</td>\n",
       "      <td>SRR8615233.917233</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               barcode UMI    EC read_index       fastq_header file_index  \\\n",
       "9665  AAAAAAAAAAAAAAAA   T  1702     917225  SRR8615233.917226          0   \n",
       "9666  AAAAAAAAAAAAAAAA   T  1702     917227  SRR8615233.917228          0   \n",
       "9667  AAAAAAAAAAAAAAAA   T  1702     917230  SRR8615233.917231          0   \n",
       "9668  AAAAAAAAAAAAAAAA   T  1702     917231  SRR8615233.917232          0   \n",
       "9669  AAAAAAAAAAAAAAAA   T  1702     917232  SRR8615233.917233          0   \n",
       "\n",
       "                  variant_names  counted_in_count_matrix  \\\n",
       "9665  [ENST00000251453:c.80T>G]                     True   \n",
       "9666  [ENST00000251453:c.80T>G]                     True   \n",
       "9667  [ENST00000251453:c.80T>G]                     True   \n",
       "9668  [ENST00000251453:c.80T>G]                     True   \n",
       "9669  [ENST00000251453:c.80T>G]                     True   \n",
       "\n",
       "            variant_names_str  \n",
       "9665  ENST00000251453:c.80T>G  \n",
       "9666  ENST00000251453:c.80T>G  \n",
       "9667  ENST00000251453:c.80T>G  \n",
       "9668  ENST00000251453:c.80T>G  \n",
       "9669  ENST00000251453:c.80T>G  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bus_df = make_bus_df(kb_count_out_dir, fastq_file, technology=technology, parity=parity, bustools=bustools)\n",
    "bus_df = bus_df.drop(columns=[\"transcript_names\"]).rename(columns={\"gene_names\": \"variant_names\"})\n",
    "filtered_bus_df = bus_df[bus_df[\"variant_names\"].apply(lambda x: len(x) == 1)]  # remove multi-mapping reads\n",
    "filtered_bus_df[\"variant_names_str\"] = filtered_bus_df[\"variant_names\"].apply(lambda x: x[0])  # cast to string\n",
    "filtered_bus_df = filtered_bus_df[(filtered_bus_df[\"variant_names_str\"].isin(top15)) & (filtered_bus_df[\"counted_in_count_matrix\"])]\n",
    "print(f\"Number of reads in filtered bus file: {len(filtered_bus_df)}\")\n",
    "filtered_bus_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "32c55d55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENST00000398606:c.562A>G': 'vcrs_0',\n",
       " 'ENST00000449260:c.1988C>T': 'vcrs_1',\n",
       " 'ENST00000251453:c.80T>G': 'vcrs_2',\n",
       " 'ENST00000398606:c.804T>C': 'vcrs_3',\n",
       " 'ENST00000343262:c.156C>T': 'vcrs_4',\n",
       " 'ENST00000322723:c.1024_1026del': 'vcrs_5',\n",
       " 'ENST00000263574:c.750_755del': 'vcrs_6',\n",
       " 'ENST00000367034:c.152_154del': 'vcrs_7',\n",
       " 'ENST00000314222:c.448_449insCCCCC': 'vcrs_8',\n",
       " 'ENST00000373477:c.2126_2127insCGTAC': 'vcrs_9',\n",
       " 'ENST00000434724:c.239_247del': 'vcrs_10',\n",
       " 'ENST00000529010:c.1426_1428del': 'vcrs_11',\n",
       " 'ENST00000262710:c.2258_2259insATTCAC': 'vcrs_12',\n",
       " 'ENST00000372797:c.1497_1498insG': 'vcrs_13',\n",
       " 'ENST00000435765:c.1262_1263insGTT': 'vcrs_14'}"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make IDs since the gene names in HGVSC format do not make for good folder names or bowtie2 header names\n",
    "vcrs_header_to_id = {vcrs: f\"vcrs_{i}\" for i, vcrs in enumerate(list(top15))}\n",
    "filtered_bus_df[\"vcrs_ids\"] = filtered_bus_df[\"variant_names_str\"].map(vcrs_header_to_id)\n",
    "\n",
    "vcrs_header_to_id_file = os.path.join(notebook_out_dir, \"vcrs_header_to_id.txt\")\n",
    "with open(vcrs_header_to_id_file, \"w\") as f:\n",
    "    for key, value in vcrs_header_to_id.items():\n",
    "        f.write(f\"{key}\\t{value}\\n\")  # Tab-separated\n",
    "\n",
    "id_to_vcrs_header = {v: k for k, v in vcrs_header_to_id.items()}\n",
    "vcrs_header_to_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b504ecc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing vcrs_0\n",
      "Processing vcrs_1\n",
      "Processing vcrs_10\n",
      "Processing vcrs_11\n",
      "Processing vcrs_12\n",
      "Processing vcrs_13\n",
      "Processing vcrs_14\n",
      "Processing vcrs_2\n",
      "Processing vcrs_3\n",
      "Processing vcrs_4\n",
      "Processing vcrs_5\n",
      "Processing vcrs_6\n",
      "Processing vcrs_7\n",
      "Processing vcrs_8\n",
      "Processing vcrs_9\n"
     ]
    }
   ],
   "source": [
    "for vcrs_id in sorted(filtered_bus_df[\"vcrs_ids\"].unique()):  # Get unique gene names\n",
    "    print(f\"Processing {vcrs_id}\")\n",
    "    temp_df = filtered_bus_df[filtered_bus_df[\"vcrs_ids\"] == vcrs_id]  # Filter\n",
    "    fastq_headers = temp_df[\"fastq_header\"].tolist()  # Get values as a list\n",
    "\n",
    "    gene_dir = os.path.join(aligned_reads_parent_dir, vcrs_id)\n",
    "    os.makedirs(gene_dir, exist_ok=True)\n",
    "    \n",
    "    aligned_reads_file = os.path.join(gene_dir, \"1.fastq\")\n",
    "    with open(aligned_reads_file, \"w\") as f:\n",
    "        for header in fastq_headers:\n",
    "            sequence = fastq_indexed[header].seq\n",
    "            qualities = fastq_indexed[header].qual\n",
    "            f.write(f\"@{header}\\n{sequence}\\n+\\n{qualities}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f1ecf2",
   "metadata": {},
   "source": [
    "### View COSMIC metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "107a9752",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/2740002178.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cosmic_df_subset[\"id\"] = cosmic_df_subset[\"header\"].map(vcrs_header_to_id)\n",
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/2740002178.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cosmic_df_subset[\"counts\"] = cosmic_df_subset[\"header\"].map(lambda x: top15_dict[x])\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "counts",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "d2d0fb6d-3b42-4d13-9c63-c9d1475cd228",
       "rows": [
        [
         "5118134",
         "ENST00000398606",
         "c.313A>G",
         "p.I105V",
         "COSV66992376",
         "11",
         "+",
         "g.67352689A>G",
         "GSTP1",
         "48379894",
         "c.562A>G",
         "ENST00000398606:c.562A>G",
         "substitution",
         "vcrs_0",
         "13934"
        ],
        [
         "1545555",
         "ENST00000449260",
         "c.1977C>T",
         "p.N659=",
         "COSV59384225",
         "12",
         "-",
         "g.56348028G>A",
         "PMEL",
         "60961346",
         "c.1988C>T",
         "ENST00000449260:c.1988C>T",
         "substitution",
         "vcrs_1",
         "12589"
        ],
        [
         "4199446",
         "ENST00000251453",
         "c.27T>G",
         "p.S9=",
         "COSV52238647",
         "19",
         "-",
         "g.39926509A>C",
         "RPS16",
         "21768506",
         "c.80T>G",
         "ENST00000251453:c.80T>G",
         "substitution",
         "vcrs_2",
         "11981"
        ],
        [
         "5118182",
         "ENST00000398606",
         "c.555T>C",
         "p.S185=",
         "COSV66992478",
         "11",
         "+",
         "g.67353970T>C",
         "GSTP1",
         "48379878",
         "c.804T>C",
         "ENST00000398606:c.804T>C",
         "substitution",
         "vcrs_3",
         "11300"
        ],
        [
         "3224064",
         "ENST00000343262",
         "c.99C>T",
         "p.I33=",
         "COSV51909983",
         "16",
         "-",
         "g.2014528G>A",
         "RPS2",
         "31955826",
         "c.156C>T",
         "ENST00000343262:c.156C>T",
         "substitution",
         "vcrs_4",
         "6296"
        ],
        [
         "5239178",
         "ENST00000322723",
         "c.783_785del",
         "p.D261del",
         "COSV59544931",
         "2",
         null,
         "g.232325411_232325413del",
         "NCL",
         "26628001",
         "c.1024_1026del",
         "ENST00000322723:c.1024_1026del",
         "deletion",
         "vcrs_5",
         "2501"
        ],
        [
         "5383497",
         "ENST00000263574",
         "c.678_683del",
         "p.E226_E227del",
         "COSV53838522",
         "11",
         null,
         "g.129991670_129991675del",
         "APLP2",
         "20079349",
         "c.750_755del",
         "ENST00000263574:c.750_755del",
         "deletion",
         "vcrs_6",
         "254"
        ],
        [
         "5245332",
         "ENST00000367034",
         "c.30_32del",
         "p.L10del",
         "COSV58458382",
         "6",
         null,
         "g.160211649_160211651del",
         "MRPL18",
         "39042486",
         "c.152_154del",
         "ENST00000367034:c.152_154del",
         "deletion",
         "vcrs_7",
         "216"
        ],
        [
         "3000122",
         "ENST00000373477",
         "c.1217_1218insCGTAC",
         "p.Q408Rfs*15",
         "COSV65114720",
         "1",
         "-",
         "g.33245802_33245803insGTACG",
         "YARS",
         "39378578",
         "c.2126_2127insCGTAC",
         "ENST00000373477:c.2126_2127insCGTAC",
         "insertion",
         "vcrs_9",
         "145"
        ],
        [
         "3767335",
         "ENST00000314222",
         "c.357_358insCCCCC",
         "p.A120Pfs*50",
         "COSV56542491",
         "11",
         "-",
         "g.2950240_2950241insGGGGG",
         "PHLDA2",
         "27287821",
         "c.448_449insCCCCC",
         "ENST00000314222:c.448_449insCCCCC",
         "insertion",
         "vcrs_8",
         "145"
        ],
        [
         "5246024",
         "ENST00000434724",
         "c.39_47del",
         "p.G14_G16del",
         "COSV70453045",
         "16",
         null,
         "g.12009554_12009562del",
         "GSPT1",
         "57130237",
         "c.239_247del",
         "ENST00000434724:c.239_247del",
         "deletion",
         "vcrs_10",
         "115"
        ],
        [
         "5382296",
         "ENST00000529010",
         "c.1207_1209del",
         "p.Q403del",
         "COSV60376359",
         "11",
         null,
         "g.17352482_17352484del",
         "NUCB2",
         "81560910",
         "c.1426_1428del",
         "ENST00000529010:c.1426_1428del",
         "deletion",
         "vcrs_11",
         "108"
        ],
        [
         "548620",
         "ENST00000262710",
         "c.1930_1931insATTCAC",
         "p.S643_R644insHS",
         "COSV52972535",
         "14",
         "-",
         "g.23548792_23548793insTGTGAA",
         "ACIN1",
         "23586862",
         "c.2258_2259insATTCAC",
         "ENST00000262710:c.2258_2259insATTCAC",
         "insertion",
         "vcrs_12",
         "82"
        ],
        [
         "1823895",
         "ENST00000372797",
         "c.936_937insG",
         "p.R313Afs*11",
         "COSV61223036",
         "1",
         "+",
         "g.40535489_40535490insG",
         "CAP1",
         "40250283",
         "c.1497_1498insG",
         "ENST00000372797:c.1497_1498insG",
         "insertion",
         "vcrs_13",
         "27"
        ],
        [
         "455948",
         "ENST00000435765",
         "c.851_852insGTT",
         "p.F285dup",
         "COSV53007999",
         "7",
         "+",
         "g.103007866_103007867insGTT",
         "PSMC2",
         "55802712",
         "c.1262_1263insGTT",
         "ENST00000435765:c.1262_1263insGTT",
         "insertion",
         "vcrs_14",
         "26"
        ]
       ],
       "shape": {
        "columns": 14,
        "rows": 15
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5118134</th>\n",
       "      <td>ENST00000398606</td>\n",
       "      <td>c.313A&gt;G</td>\n",
       "      <td>p.I105V</td>\n",
       "      <td>COSV66992376</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>g.67352689A&gt;G</td>\n",
       "      <td>GSTP1</td>\n",
       "      <td>48379894</td>\n",
       "      <td>c.562A&gt;G</td>\n",
       "      <td>ENST00000398606:c.562A&gt;G</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_0</td>\n",
       "      <td>13934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1545555</th>\n",
       "      <td>ENST00000449260</td>\n",
       "      <td>c.1977C&gt;T</td>\n",
       "      <td>p.N659=</td>\n",
       "      <td>COSV59384225</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.56348028G&gt;A</td>\n",
       "      <td>PMEL</td>\n",
       "      <td>60961346</td>\n",
       "      <td>c.1988C&gt;T</td>\n",
       "      <td>ENST00000449260:c.1988C&gt;T</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_1</td>\n",
       "      <td>12589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4199446</th>\n",
       "      <td>ENST00000251453</td>\n",
       "      <td>c.27T&gt;G</td>\n",
       "      <td>p.S9=</td>\n",
       "      <td>COSV52238647</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>g.39926509A&gt;C</td>\n",
       "      <td>RPS16</td>\n",
       "      <td>21768506</td>\n",
       "      <td>c.80T&gt;G</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_2</td>\n",
       "      <td>11981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118182</th>\n",
       "      <td>ENST00000398606</td>\n",
       "      <td>c.555T&gt;C</td>\n",
       "      <td>p.S185=</td>\n",
       "      <td>COSV66992478</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>g.67353970T&gt;C</td>\n",
       "      <td>GSTP1</td>\n",
       "      <td>48379878</td>\n",
       "      <td>c.804T&gt;C</td>\n",
       "      <td>ENST00000398606:c.804T&gt;C</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_3</td>\n",
       "      <td>11300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3224064</th>\n",
       "      <td>ENST00000343262</td>\n",
       "      <td>c.99C&gt;T</td>\n",
       "      <td>p.I33=</td>\n",
       "      <td>COSV51909983</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>g.2014528G&gt;A</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>31955826</td>\n",
       "      <td>c.156C&gt;T</td>\n",
       "      <td>ENST00000343262:c.156C&gt;T</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_4</td>\n",
       "      <td>6296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5239178</th>\n",
       "      <td>ENST00000322723</td>\n",
       "      <td>c.783_785del</td>\n",
       "      <td>p.D261del</td>\n",
       "      <td>COSV59544931</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.232325411_232325413del</td>\n",
       "      <td>NCL</td>\n",
       "      <td>26628001</td>\n",
       "      <td>c.1024_1026del</td>\n",
       "      <td>ENST00000322723:c.1024_1026del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_5</td>\n",
       "      <td>2501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5383497</th>\n",
       "      <td>ENST00000263574</td>\n",
       "      <td>c.678_683del</td>\n",
       "      <td>p.E226_E227del</td>\n",
       "      <td>COSV53838522</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.129991670_129991675del</td>\n",
       "      <td>APLP2</td>\n",
       "      <td>20079349</td>\n",
       "      <td>c.750_755del</td>\n",
       "      <td>ENST00000263574:c.750_755del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_6</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5245332</th>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>c.30_32del</td>\n",
       "      <td>p.L10del</td>\n",
       "      <td>COSV58458382</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.160211649_160211651del</td>\n",
       "      <td>MRPL18</td>\n",
       "      <td>39042486</td>\n",
       "      <td>c.152_154del</td>\n",
       "      <td>ENST00000367034:c.152_154del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_7</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000122</th>\n",
       "      <td>ENST00000373477</td>\n",
       "      <td>c.1217_1218insCGTAC</td>\n",
       "      <td>p.Q408Rfs*15</td>\n",
       "      <td>COSV65114720</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.33245802_33245803insGTACG</td>\n",
       "      <td>YARS</td>\n",
       "      <td>39378578</td>\n",
       "      <td>c.2126_2127insCGTAC</td>\n",
       "      <td>ENST00000373477:c.2126_2127insCGTAC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_9</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3767335</th>\n",
       "      <td>ENST00000314222</td>\n",
       "      <td>c.357_358insCCCCC</td>\n",
       "      <td>p.A120Pfs*50</td>\n",
       "      <td>COSV56542491</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>g.2950240_2950241insGGGGG</td>\n",
       "      <td>PHLDA2</td>\n",
       "      <td>27287821</td>\n",
       "      <td>c.448_449insCCCCC</td>\n",
       "      <td>ENST00000314222:c.448_449insCCCCC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_8</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5246024</th>\n",
       "      <td>ENST00000434724</td>\n",
       "      <td>c.39_47del</td>\n",
       "      <td>p.G14_G16del</td>\n",
       "      <td>COSV70453045</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.12009554_12009562del</td>\n",
       "      <td>GSPT1</td>\n",
       "      <td>57130237</td>\n",
       "      <td>c.239_247del</td>\n",
       "      <td>ENST00000434724:c.239_247del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_10</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5382296</th>\n",
       "      <td>ENST00000529010</td>\n",
       "      <td>c.1207_1209del</td>\n",
       "      <td>p.Q403del</td>\n",
       "      <td>COSV60376359</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.17352482_17352484del</td>\n",
       "      <td>NUCB2</td>\n",
       "      <td>81560910</td>\n",
       "      <td>c.1426_1428del</td>\n",
       "      <td>ENST00000529010:c.1426_1428del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_11</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548620</th>\n",
       "      <td>ENST00000262710</td>\n",
       "      <td>c.1930_1931insATTCAC</td>\n",
       "      <td>p.S643_R644insHS</td>\n",
       "      <td>COSV52972535</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>g.23548792_23548793insTGTGAA</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>23586862</td>\n",
       "      <td>c.2258_2259insATTCAC</td>\n",
       "      <td>ENST00000262710:c.2258_2259insATTCAC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_12</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1823895</th>\n",
       "      <td>ENST00000372797</td>\n",
       "      <td>c.936_937insG</td>\n",
       "      <td>p.R313Afs*11</td>\n",
       "      <td>COSV61223036</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>g.40535489_40535490insG</td>\n",
       "      <td>CAP1</td>\n",
       "      <td>40250283</td>\n",
       "      <td>c.1497_1498insG</td>\n",
       "      <td>ENST00000372797:c.1497_1498insG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_13</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455948</th>\n",
       "      <td>ENST00000435765</td>\n",
       "      <td>c.851_852insGTT</td>\n",
       "      <td>p.F285dup</td>\n",
       "      <td>COSV53007999</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.103007866_103007867insGTT</td>\n",
       "      <td>PSMC2</td>\n",
       "      <td>55802712</td>\n",
       "      <td>c.1262_1263insGTT</td>\n",
       "      <td>ENST00000435765:c.1262_1263insGTT</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_14</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  seq_ID              mutation       mutation_aa  \\\n",
       "5118134  ENST00000398606              c.313A>G           p.I105V   \n",
       "1545555  ENST00000449260             c.1977C>T           p.N659=   \n",
       "4199446  ENST00000251453               c.27T>G             p.S9=   \n",
       "5118182  ENST00000398606              c.555T>C           p.S185=   \n",
       "3224064  ENST00000343262               c.99C>T            p.I33=   \n",
       "5239178  ENST00000322723          c.783_785del         p.D261del   \n",
       "5383497  ENST00000263574          c.678_683del    p.E226_E227del   \n",
       "5245332  ENST00000367034            c.30_32del          p.L10del   \n",
       "3000122  ENST00000373477   c.1217_1218insCGTAC      p.Q408Rfs*15   \n",
       "3767335  ENST00000314222     c.357_358insCCCCC      p.A120Pfs*50   \n",
       "5246024  ENST00000434724            c.39_47del      p.G14_G16del   \n",
       "5382296  ENST00000529010        c.1207_1209del         p.Q403del   \n",
       "548620   ENST00000262710  c.1930_1931insATTCAC  p.S643_R644insHS   \n",
       "1823895  ENST00000372797         c.936_937insG      p.R313Afs*11   \n",
       "455948   ENST00000435765       c.851_852insGTT         p.F285dup   \n",
       "\n",
       "        GENOMIC_MUTATION_ID chromosome strand               mutation_genome  \\\n",
       "5118134        COSV66992376         11      +                 g.67352689A>G   \n",
       "1545555        COSV59384225         12      -                 g.56348028G>A   \n",
       "4199446        COSV52238647         19      -                 g.39926509A>C   \n",
       "5118182        COSV66992478         11      +                 g.67353970T>C   \n",
       "3224064        COSV51909983         16      -                  g.2014528G>A   \n",
       "5239178        COSV59544931          2    NaN      g.232325411_232325413del   \n",
       "5383497        COSV53838522         11    NaN      g.129991670_129991675del   \n",
       "5245332        COSV58458382          6    NaN      g.160211649_160211651del   \n",
       "3000122        COSV65114720          1      -   g.33245802_33245803insGTACG   \n",
       "3767335        COSV56542491         11      -     g.2950240_2950241insGGGGG   \n",
       "5246024        COSV70453045         16    NaN        g.12009554_12009562del   \n",
       "5382296        COSV60376359         11    NaN        g.17352482_17352484del   \n",
       "548620         COSV52972535         14      -  g.23548792_23548793insTGTGAA   \n",
       "1823895        COSV61223036          1      +       g.40535489_40535490insG   \n",
       "455948         COSV53007999          7      +   g.103007866_103007867insGTT   \n",
       "\n",
       "        gene_name  mutation_id         mutation_cdna  \\\n",
       "5118134     GSTP1     48379894              c.562A>G   \n",
       "1545555      PMEL     60961346             c.1988C>T   \n",
       "4199446     RPS16     21768506               c.80T>G   \n",
       "5118182     GSTP1     48379878              c.804T>C   \n",
       "3224064      RPS2     31955826              c.156C>T   \n",
       "5239178       NCL     26628001        c.1024_1026del   \n",
       "5383497     APLP2     20079349          c.750_755del   \n",
       "5245332    MRPL18     39042486          c.152_154del   \n",
       "3000122      YARS     39378578   c.2126_2127insCGTAC   \n",
       "3767335    PHLDA2     27287821     c.448_449insCCCCC   \n",
       "5246024     GSPT1     57130237          c.239_247del   \n",
       "5382296     NUCB2     81560910        c.1426_1428del   \n",
       "548620      ACIN1     23586862  c.2258_2259insATTCAC   \n",
       "1823895      CAP1     40250283       c.1497_1498insG   \n",
       "455948      PSMC2     55802712     c.1262_1263insGTT   \n",
       "\n",
       "                                       header  variant_type       id  counts  \n",
       "5118134              ENST00000398606:c.562A>G  substitution   vcrs_0   13934  \n",
       "1545555             ENST00000449260:c.1988C>T  substitution   vcrs_1   12589  \n",
       "4199446               ENST00000251453:c.80T>G  substitution   vcrs_2   11981  \n",
       "5118182              ENST00000398606:c.804T>C  substitution   vcrs_3   11300  \n",
       "3224064              ENST00000343262:c.156C>T  substitution   vcrs_4    6296  \n",
       "5239178        ENST00000322723:c.1024_1026del      deletion   vcrs_5    2501  \n",
       "5383497          ENST00000263574:c.750_755del      deletion   vcrs_6     254  \n",
       "5245332          ENST00000367034:c.152_154del      deletion   vcrs_7     216  \n",
       "3000122   ENST00000373477:c.2126_2127insCGTAC     insertion   vcrs_9     145  \n",
       "3767335     ENST00000314222:c.448_449insCCCCC     insertion   vcrs_8     145  \n",
       "5246024          ENST00000434724:c.239_247del      deletion  vcrs_10     115  \n",
       "5382296        ENST00000529010:c.1426_1428del      deletion  vcrs_11     108  \n",
       "548620   ENST00000262710:c.2258_2259insATTCAC     insertion  vcrs_12      82  \n",
       "1823895       ENST00000372797:c.1497_1498insG     insertion  vcrs_13      27  \n",
       "455948      ENST00000435765:c.1262_1263insGTT     insertion  vcrs_14      26  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df_subset = cosmic_df[cosmic_df[\"header\"].isin(top15)]\n",
    "cosmic_df_subset[\"id\"] = cosmic_df_subset[\"header\"].map(vcrs_header_to_id)\n",
    "cosmic_df_subset[\"counts\"] = cosmic_df_subset[\"header\"].map(lambda x: top15_dict[x])\n",
    "cosmic_df_subset = cosmic_df_subset.sort_values(by=\"counts\", ascending=False)\n",
    "cosmic_df_subset.to_csv(os.path.join(notebook_out_dir, \"cosmic_df_subset.csv\"), index=False)\n",
    "cosmic_df_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "487178cb-9825-4c23-9551-baf7a6934f9c",
   "metadata": {},
   "source": [
    "### Align (1) pulled out reads and (2) the VCRSs to the human genome to generate bam files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28187021-d85a-4bd0-9892-7fd5e20e46e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcrs_0 (ENST00000398606:c.562A>G)\n",
      "\n",
      "13934 reads; of these:\n",
      "  13934 (100.00%) were unpaired; of these:\n",
      "    6542 (46.95%) aligned 0 times\n",
      "    7379 (52.96%) aligned exactly 1 time\n",
      "    13 (0.09%) aligned >1 times\n",
      "53.05% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_1 (ENST00000449260:c.1988C>T)\n",
      "\n",
      "12589 reads; of these:\n",
      "  12589 (100.00%) were unpaired; of these:\n",
      "    39 (0.31%) aligned 0 times\n",
      "    12550 (99.69%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "99.69% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_10 (ENST00000434724:c.239_247del)\n",
      "\n",
      "115 reads; of these:\n",
      "  115 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    115 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_11 (ENST00000529010:c.1426_1428del)\n",
      "\n",
      "108 reads; of these:\n",
      "  108 (100.00%) were unpaired; of these:\n",
      "    102 (94.44%) aligned 0 times\n",
      "    6 (5.56%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "5.56% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_12 (ENST00000262710:c.2258_2259insATTCAC)\n",
      "\n",
      "82 reads; of these:\n",
      "  82 (100.00%) were unpaired; of these:\n",
      "    38 (46.34%) aligned 0 times\n",
      "    44 (53.66%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "53.66% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_13 (ENST00000372797:c.1497_1498insG)\n",
      "\n",
      "27 reads; of these:\n",
      "  27 (100.00%) were unpaired; of these:\n",
      "    1 (3.70%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    26 (96.30%) aligned >1 times\n",
      "96.30% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_14 (ENST00000435765:c.1262_1263insGTT)\n",
      "\n",
      "26 reads; of these:\n",
      "  26 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    26 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_2 (ENST00000251453:c.80T>G)\n",
      "\n",
      "11981 reads; of these:\n",
      "  11981 (100.00%) were unpaired; of these:\n",
      "    53 (0.44%) aligned 0 times\n",
      "    1151 (9.61%) aligned exactly 1 time\n",
      "    10777 (89.95%) aligned >1 times\n",
      "99.56% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_3 (ENST00000398606:c.804T>C)\n",
      "\n",
      "11300 reads; of these:\n",
      "  11300 (100.00%) were unpaired; of these:\n",
      "    45 (0.40%) aligned 0 times\n",
      "    7958 (70.42%) aligned exactly 1 time\n",
      "    3297 (29.18%) aligned >1 times\n",
      "99.60% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_4 (ENST00000343262:c.156C>T)\n",
      "\n",
      "6296 reads; of these:\n",
      "  6296 (100.00%) were unpaired; of these:\n",
      "    16 (0.25%) aligned 0 times\n",
      "    4 (0.06%) aligned exactly 1 time\n",
      "    6276 (99.68%) aligned >1 times\n",
      "99.75% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_5 (ENST00000322723:c.1024_1026del)\n",
      "\n",
      "2501 reads; of these:\n",
      "  2501 (100.00%) were unpaired; of these:\n",
      "    1682 (67.25%) aligned 0 times\n",
      "    816 (32.63%) aligned exactly 1 time\n",
      "    3 (0.12%) aligned >1 times\n",
      "32.75% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_6 (ENST00000263574:c.750_755del)\n",
      "\n",
      "254 reads; of these:\n",
      "  254 (100.00%) were unpaired; of these:\n",
      "    96 (37.80%) aligned 0 times\n",
      "    158 (62.20%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "62.20% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_7 (ENST00000367034:c.152_154del)\n",
      "\n",
      "216 reads; of these:\n",
      "  216 (100.00%) were unpaired; of these:\n",
      "    209 (96.76%) aligned 0 times\n",
      "    7 (3.24%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "3.24% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_8 (ENST00000314222:c.448_449insCCCCC)\n",
      "\n",
      "145 reads; of these:\n",
      "  145 (100.00%) were unpaired; of these:\n",
      "    1 (0.69%) aligned 0 times\n",
      "    144 (99.31%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "99.31% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_9 (ENST00000373477:c.2126_2127insCGTAC)\n",
      "\n",
      "145 reads; of these:\n",
      "  145 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    145 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n"
     ]
    }
   ],
   "source": [
    "for folder in sorted(glob.glob(aligned_reads_parent_dir + \"/*/\")):\n",
    "    variant = folder.split(\"/\")[-2]\n",
    "    variant_header = id_to_vcrs_header[variant]\n",
    "    print(f\"{variant} ({variant_header})\\n\")\n",
    "\n",
    "    outfolder = f\"{bowtie_read_alignments}/{variant}\"\n",
    "    os.makedirs(outfolder, exist_ok=True)\n",
    "    \n",
    "    # Align reads to human ref using bowtie2\n",
    "    variant_sam = variant + \"_reads.sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 3 \\\n",
    "        -x $bowtie_reference_prefix \\\n",
    "        -p $threads \\\n",
    "        -q $aligned_reads_parent_dir/$variant/1.fastq \\\n",
    "        -S $outfolder/$variant_sam\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \"_reads.bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder/$variant_sam \\\n",
    "        > $outfolder/$variant_bam\n",
    "    \n",
    "    # Sort bam file\n",
    "    variant_bam_sorted_prefix = \"final_\" + variant + \"_sorted_reads\"\n",
    "    variant_bam_sorted = variant_bam_sorted_prefix + \".bam\"\n",
    "    !$samtools sort \\\n",
    "        -o $outfolder/$variant_bam_sorted_prefix \\\n",
    "        $outfolder/$variant_bam\n",
    "    \n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder/$variant_bam_sorted\n",
    "\n",
    "\n",
    "    # repeat but use vcrs as the input instead of reads\n",
    "    variant_reference_sequence = vcrs_fasta_indexed[variant_header].seq.strip()\n",
    "\n",
    "    # Align vcrs to human ref using bowtie2\n",
    "    variant_sam = variant + \"_vcrs.sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 3 \\\n",
    "        -x $bowtie_reference_prefix \\\n",
    "        -p $threads \\\n",
    "        -c $variant_reference_sequence \\\n",
    "        -S $outfolder/$variant_sam\n",
    "    # use k 3 to get multiple alignments, not just the best one\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \"_vcrs.bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder/$variant_sam \\\n",
    "        > $outfolder/$variant_bam\n",
    "\n",
    "    # Sort bam file\n",
    "    variant_bam_sorted_prefix = \"final_\" + variant + \"_sorted_vcrs\"\n",
    "    variant_bam_sorted = variant_bam_sorted_prefix + \".bam\"\n",
    "    !$samtools sort \\\n",
    "        $outfolder/$variant_bam \\\n",
    "        $outfolder/$variant_bam_sorted_prefix\n",
    "\n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder/$variant_bam_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb02e32",
   "metadata": {},
   "source": [
    "### Load the files in bowtie_read_alignments into NCBI Genome workbench (or another genome viewer) to visualize the alignments (the ones with \"final_\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RLSRWP_2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
