{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b8a4b538-0583-45e9-91d1-eec859bae4c1",
   "metadata": {},
   "source": [
    "# Extract cancer variant reads for alignment visualization\n",
    "\n",
    "## In this notebook, we use a single RNA-seq fastq file from a melanoma cancer cell line from the the CCLE project. Learn more about the project here:\n",
    "- sequencing data (ENA): https://www.ebi.ac.uk/ena/browser/view/PRJNA523380\n",
    "- paper: https://www.nature.com/articles/s41586-019-1186-3\n",
    "\n",
    "### Requirements: kb, samtools, and bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9cae3de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import varseek as vk\n",
    "except ImportError:\n",
    "    print(\"varseek not found, installing...\")\n",
    "    !pip install -U -q varseek\n",
    "try:\n",
    "    import RLSRWP_2025\n",
    "except ImportError:\n",
    "    print(\"RLSRWP_2025 not found, installing...\")\n",
    "    !pip install -q git+https://github.com/pachterlab/RLSRWP_2025.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "400e7db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gget\n",
    "import pyfastx\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "# pd.set_option('display.max_columns', None)\n",
    "import varseek as vk\n",
    "from varseek.utils import make_bus_df, add_variant_type\n",
    "\n",
    "RLSRWP_2025_dir = os.path.dirname(os.path.abspath(\"\"))  # if this notebook resides in RLSRWP_2025/notebooks/0_data_download.ipynb, then this retrieves RLSRWP_2025\n",
    "threads = \"2\"\n",
    "use_head = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "317e1ef9",
   "metadata": {},
   "source": [
    "### File path definitions and imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b81b187a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_count_out_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"varseek_count_out_alignment_visualization_full\")\n",
    "kb_count_out_dir = os.path.join(vk_count_out_dir, \"kb_count_out_vcrs\")\n",
    "adata_path = os.path.join(kb_count_out_dir, \"counts_unfiltered\", \"adata.h5ad\")\n",
    "aligned_reads_parent_dir = os.path.join(vk_count_out_dir, \"pseudoaligned_reads_to_vcrs_reference\")\n",
    "bowtie_read_alignments = os.path.join(vk_count_out_dir, \"bowtie_read_alignments\")\n",
    "\n",
    "# vk ref out directory and files - downloaded if not already present\n",
    "vk_ref_out_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"vk_ref_out\")\n",
    "vcrs_index = os.path.join(vk_ref_out_dir, \"vcrs_index.idx\")\n",
    "vcrs_t2g = os.path.join(vk_ref_out_dir, \"vcrs_t2g_filtered.txt\")\n",
    "vcrs_fasta = os.path.join(vk_ref_out_dir, \"vcrs_filtered.fa\")\n",
    "\n",
    "# fastq directories - fastqs_dir downloaded if not already present, and fastqs_processed_dir created with fastp if not already present\n",
    "fastqs_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"ccle_data_base\", \"RNASeq_MELHO_SKIN\")\n",
    "fastq_file = os.path.join(fastqs_dir, \"SRR8615233_1.fastq.gz\")\n",
    "technology = \"bulk\"\n",
    "\n",
    "# cosmic directory\n",
    "cosmic_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"reference\", \"cosmic\")\n",
    "cosmic_csv = os.path.join(cosmic_dir, \"CancerMutationCensus_AllData_Tsv_v101_GRCh37\", \"CancerMutationCensus_AllData_v101_GRCh37_mutation_workflow.csv\")\n",
    "\n",
    "# reference genome directory\n",
    "reference_genome_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"reference\", \"ensembl_grch37_release93\")\n",
    "\n",
    "# kb count to reference genome directory and files - created if not already present - only used if qc_against_gene_matrix=True\n",
    "qc_against_gene_matrix = False\n",
    "kb_count_reference_genome_dir = os.path.join(RLSRWP_2025_dir, \"data\", \"kb_count_reference_genome\")\n",
    "reference_genome_index = os.path.join(reference_genome_dir, \"index.idx\")  # either already exists or will be created\n",
    "reference_genome_t2g = os.path.join(reference_genome_dir, \"t2g.txt\")  # either already exists or will be created\n",
    "reference_genome_fasta = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.dna.primary_assembly.fa\")  # if reference_genome_index/reference_genome_t2g do not exist, then I need to supply the reference genome fasta and gtf\n",
    "reference_genome_gtf = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.87.gtf\")  # if reference_genome_index/reference_genome_t2g do not exist, then I need to supply the reference genome fasta and gtf\n",
    "\n",
    "# for bowtie2 - also uses reference_genome_fasta above (I could use cdna instead of genome too if desired)\n",
    "bowtie_reference_dir = os.path.join(reference_genome_dir, \"bowtie_index_genome\")\n",
    "bowtie_reference_prefix = os.path.join(bowtie_reference_dir, \"index\")\n",
    "\n",
    "# general\n",
    "w = \"47\"  # used during creation of the index, so cannot be altered\n",
    "k = \"51\"  # used during creation of the index, so cannot be altered\n",
    "strand = \"unstranded\"\n",
    "parity = \"single\"  # although the original data is paired, we will only be using a single file, so we will run in single-end mode\n",
    "\n",
    "# software\n",
    "bustools = \"/Users/joeyrich/miniconda3/envs/RLSRWP_2025/lib/python3.10/site-packages/kb_python/bins/darwin/m1/bustools/bustools\"\n",
    "bowtie2 = \"bowtie2\"\n",
    "bowtie2_build = \"bowtie2-build\"\n",
    "samtools = \"samtools\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f50ee94",
   "metadata": {},
   "source": [
    "### Download VCRS reference files with varseek ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ebc5dd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(vcrs_index) or not os.path.exists(vcrs_t2g) or not os.path.exists(vcrs_fasta):\n",
    "    vk.ref(variants=\"cosmic_cmc\", sequences=\"cdna\", w=w, k=k, dlist_reference_source=\"t2t\", index_out=vcrs_index, t2g_out=vcrs_t2g, fasta_out=vcrs_fasta, download=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ffe24bf",
   "metadata": {},
   "source": [
    "### Download COSMIC if not already downloaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "184711a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/1725634623.py:13: DtypeWarning: Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  cosmic_df = pd.read_csv(cosmic_csv)\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "b1c09fa7-7b3f-4f5e-9019-dfb1c932fe39",
       "rows": [
        [
         "0",
         "ENST00000396153",
         "c.1468C>T",
         "p.Q490*",
         "COSV105286190",
         "1",
         "-",
         "g.51001067G>A",
         "FAF1",
         "48293736",
         "c.1920C>T",
         "ENST00000396153:c.1920C>T"
        ],
        [
         "1",
         "ENST00000445907",
         "c.162C>A",
         "p.L54=",
         "COSV57765075",
         "7",
         "+",
         "g.136699774C>A",
         "CHRM2",
         "60222372",
         "c.690C>A",
         "ENST00000445907:c.690C>A"
        ],
        [
         "2",
         "ENST00000445907",
         "c.617A>T",
         "p.Y206F",
         "COSV108138517",
         "7",
         "+",
         "g.136700229A>T",
         "CHRM2",
         "60229878",
         "c.1145A>T",
         "ENST00000445907:c.1145A>T"
        ],
        [
         "3",
         "ENST00000445907",
         "c.93A>T",
         "p.G31=",
         "COSV100282253",
         "7",
         "+",
         "g.136699705A>T",
         "CHRM2",
         "60227007",
         "c.621A>T",
         "ENST00000445907:c.621A>T"
        ],
        [
         "4",
         "ENST00000445907",
         "c.956A>C",
         "p.N319T",
         "COSV108138508",
         "7",
         "+",
         "g.136700568A>C",
         "CHRM2",
         "60213402",
         "c.1484A>C",
         "ENST00000445907:c.1484A>C"
        ]
       ],
       "shape": {
        "columns": 11,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000396153</td>\n",
       "      <td>c.1468C&gt;T</td>\n",
       "      <td>p.Q490*</td>\n",
       "      <td>COSV105286190</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.51001067G&gt;A</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>48293736</td>\n",
       "      <td>c.1920C&gt;T</td>\n",
       "      <td>ENST00000396153:c.1920C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.162C&gt;A</td>\n",
       "      <td>p.L54=</td>\n",
       "      <td>COSV57765075</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699774C&gt;A</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60222372</td>\n",
       "      <td>c.690C&gt;A</td>\n",
       "      <td>ENST00000445907:c.690C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.617A&gt;T</td>\n",
       "      <td>p.Y206F</td>\n",
       "      <td>COSV108138517</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700229A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60229878</td>\n",
       "      <td>c.1145A&gt;T</td>\n",
       "      <td>ENST00000445907:c.1145A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.93A&gt;T</td>\n",
       "      <td>p.G31=</td>\n",
       "      <td>COSV100282253</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699705A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60227007</td>\n",
       "      <td>c.621A&gt;T</td>\n",
       "      <td>ENST00000445907:c.621A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.956A&gt;C</td>\n",
       "      <td>p.N319T</td>\n",
       "      <td>COSV108138508</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700568A&gt;C</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60213402</td>\n",
       "      <td>c.1484A&gt;C</td>\n",
       "      <td>ENST00000445907:c.1484A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seq_ID   mutation mutation_aa GENOMIC_MUTATION_ID chromosome  \\\n",
       "0  ENST00000396153  c.1468C>T     p.Q490*       COSV105286190          1   \n",
       "1  ENST00000445907   c.162C>A      p.L54=        COSV57765075          7   \n",
       "2  ENST00000445907   c.617A>T     p.Y206F       COSV108138517          7   \n",
       "3  ENST00000445907    c.93A>T      p.G31=       COSV100282253          7   \n",
       "4  ENST00000445907   c.956A>C     p.N319T       COSV108138508          7   \n",
       "\n",
       "  strand mutation_genome gene_name  mutation_id mutation_cdna  \\\n",
       "0      -   g.51001067G>A      FAF1     48293736     c.1920C>T   \n",
       "1      +  g.136699774C>A     CHRM2     60222372      c.690C>A   \n",
       "2      +  g.136700229A>T     CHRM2     60229878     c.1145A>T   \n",
       "3      +  g.136699705A>T     CHRM2     60227007      c.621A>T   \n",
       "4      +  g.136700568A>C     CHRM2     60213402     c.1484A>C   \n",
       "\n",
       "                      header  \n",
       "0  ENST00000396153:c.1920C>T  \n",
       "1   ENST00000445907:c.690C>A  \n",
       "2  ENST00000445907:c.1145A>T  \n",
       "3   ENST00000445907:c.621A>T  \n",
       "4  ENST00000445907:c.1484A>C  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if cosmic_csv and not os.path.isfile(cosmic_csv):\n",
    "    gget.cosmic(\n",
    "        None,\n",
    "        grch_version=37,\n",
    "        cosmic_version=101,\n",
    "        out=cosmic_dir,\n",
    "        mutation_class=\"cancer\",\n",
    "        download_cosmic=True,\n",
    "        keep_genome_info=True,\n",
    "        remove_duplicates=True\n",
    "    )\n",
    "\n",
    "cosmic_df = pd.read_csv(cosmic_csv)\n",
    "\n",
    "if \"mutation_cdna\" not in cosmic_df.columns:\n",
    "    reference_cdna_path = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.cdna.all.fa\")\n",
    "    reference_cds_path = os.path.join(reference_genome_dir, \"Homo_sapiens.GRCh37.cds.all.fa\")\n",
    "    if not os.path.exists(reference_cdna_path):\n",
    "        reference_cdna_dir = os.path.dirname(reference_cdna_path) if os.path.dirname(reference_cdna_path) else \".\"\n",
    "        !gget ref -w cdna -r 93 --out_dir {reference_cdna_dir} -d human_grch37\n",
    "        !gunzip {reference_cdna_path}.gz\n",
    "    if not os.path.exists(reference_cds_path):\n",
    "        reference_cds_dir = os.path.dirname(reference_cds_path) if os.path.dirname(reference_cds_path) else \".\"\n",
    "        !gget ref -w cdna -r 93 --out_dir {reference_cds_dir} -d human_grch37\n",
    "        !gunzip {reference_cds_path}.gz\n",
    "    print(\"Converting CDS to cDNA in COSMIC\")\n",
    "    cosmic_df, _ = vk.utils.convert_mutation_cds_locations_to_cdna(input_csv_path=cosmic_df, output_csv_path=cosmic_csv, cds_fasta_path=reference_cds_path, cdna_fasta_path=reference_cdna_path, verbose=True)\n",
    "\n",
    "if \"variant_type\" not in cosmic_df.columns:\n",
    "    add_variant_type(cosmic_df, \"mutation_cdna\")\n",
    "\n",
    "if \"header\" not in cosmic_df.columns:\n",
    "    cosmic_df[\"header\"] = cosmic_df[\"seq_ID\"] + \":\" + cosmic_df[\"mutation_cdna\"]\n",
    "    cosmic_df.to_csv(cosmic_csv, index=False)\n",
    "\n",
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "492c02ce",
   "metadata": {},
   "source": [
    "### Make bowtie2 index files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "17451c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(bowtie_reference_dir) or len(os.listdir(bowtie_reference_dir)) == 0:\n",
    "    os.makedirs(bowtie_reference_dir, exist_ok=True)\n",
    "    !{bowtie2_build} --threads {threads} {reference_genome_fasta} {bowtie_reference_prefix}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e3028e5",
   "metadata": {},
   "source": [
    "### Download fastq file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e2bae66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-03-21 09:14:25--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\n",
      "           => ‘/Users/joeyrich/Desktop/local/RLSRWP_2025/data/ccle_data_base/RNASeq_MELHO_SKIN/SRR8615233_1.fastq.gz’\n",
      "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.165\n",
      "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.165|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR861/003/SRR8615233 ... done.\n",
      "==> SIZE SRR8615233_1.fastq.gz ... 3728501472\n",
      "==> PASV ... done.    ==> RETR SRR8615233_1.fastq.gz ... done.\n",
      "Length: 3728501472 (3.5G) (unauthoritative)\n",
      "\n",
      "SRR8615233_1.fastq.  42%[=======>            ]   1.48G  --.-KB/s    in 31m 33s \n",
      "\n",
      "2025-03-21 09:46:01 (820 KB/s) - Data connection: Operation timed out; Control connection closed.\n",
      "Retrying.\n",
      "\n",
      "--2025-03-21 10:01:02--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\n",
      "  (try: 2) => ‘/Users/joeyrich/Desktop/local/RLSRWP_2025/data/ccle_data_base/RNASeq_MELHO_SKIN/SRR8615233_1.fastq.gz’\n",
      "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.165|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR861/003/SRR8615233 ... done.\n",
      "==> SIZE SRR8615233_1.fastq.gz ... 3728501472\n",
      "==> PASV ... done.    ==> REST 1589315168 ... done.    \n",
      "==> RETR SRR8615233_1.fastq.gz ... done.\n",
      "Length: 3728501472 (3.5G), 2139186304 (2.0G) remaining (unauthoritative)\n",
      "\n",
      "SRR8615233_1.fastq. 100%[++++++++===========>]   3.47G  24.1MB/s    in 92s     \n",
      "\n",
      "2025-03-21 10:02:36 (22.3 MB/s) - ‘/Users/joeyrich/Desktop/local/RLSRWP_2025/data/ccle_data_base/RNASeq_MELHO_SKIN/SRR8615233_1.fastq.gz’ saved [3728501472]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if not os.path.isfile(fastq_file):\n",
    "    fastq_file_link = \"ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\"  # [\"ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_1.fastq.gz\", \"ftp.sra.ebi.ac.uk/vol1/fastq/SRR861/003/SRR8615233/SRR8615233_2.fastq.gz\"]\n",
    "    os.makedirs(fastqs_dir, exist_ok=True)\n",
    "    !wget -c --tries=5 --retry-connrefused -O {fastq_file} {fastq_file_link}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0dc6d8e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "if use_head:\n",
    "    fastq_file_head = fastq_file.replace(\".fastq.gz\", \"_head.fastq\")\n",
    "    if not os.path.isfile(fastq_file_head):\n",
    "        !zcat {fastq_file} | head -1000000 > $fastq_file_head\n",
    "    fastq_file = fastq_file_head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "907dad2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not fastq_file.endswith(\".gz\"):\n",
    "    !head {fastq_file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202970af",
   "metadata": {},
   "source": [
    "### Make pyfastx index files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a971147c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fastq_indexed = pyfastx.Fastq(fastq_file, build_index=True)\n",
    "vcrs_fasta_indexed = pyfastx.Fasta(vcrs_fasta, build_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4fc8333",
   "metadata": {},
   "source": [
    "### Make kb reference files if qc_against_gene_matrix=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5fc75df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "kb_count_reference_genome_adata = os.path.join(kb_count_reference_genome_dir, \"counts_unfiltered\", \"adata.h5ad\")\n",
    "if qc_against_gene_matrix and not os.path.exists(kb_count_reference_genome_adata):  # check if kb count was run\n",
    "    os.makedirs(kb_count_reference_genome_dir, exist_ok=True)\n",
    "    if not os.path.exists(reference_genome_index) or not os.path.exists(reference_genome_t2g):  # check if kb ref was run\n",
    "        if not os.path.exists(reference_genome_fasta) or not os.path.exists(reference_genome_gtf):\n",
    "            reference_genome_out_dir = os.path.dirname(reference_genome_fasta) if reference_genome_fasta else \".\"\n",
    "            # using grch37, ensembl 93 to agree with COSMIC\n",
    "            !gget ref -w dna,gtf -r 93 --out_dir {reference_genome_out_dir} -d human_grch37 && gunzip {reference_genome_fasta}.gz && gunzip {reference_genome_gtf}.gz\n",
    "        reference_genome_f1 = os.path.join(kb_count_reference_genome_dir, \"f1.fa\")\n",
    "        !kb ref -t {threads} -i {reference_genome_index} -g {reference_genome_t2g} -f1 {reference_genome_f1} {reference_genome_fasta} {reference_genome_gtf}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70df8866",
   "metadata": {},
   "source": [
    "### Perform variant screening with varseek count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ca71a00e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10:04:58 - INFO - Setting length_required to 51 if fastqpp is run\n",
      "10:04:58 - INFO - Skipping vk fastqpp because disable_fastqpp=True\n",
      "10:04:58 - INFO - Running kb count with command: kb count -t 2 -k 51 -i /Users/joeyrich/Desktop/local/RLSRWP_2025/data/vk_ref_out/vcrs_index.idx -g /Users/joeyrich/Desktop/local/RLSRWP_2025/data/vk_ref_out/vcrs_t2g_filtered.txt -x BULK --h5ad -o /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs --overwrite --strand unstranded --parity single --num /Users/joeyrich/Desktop/local/RLSRWP_2025/data/ccle_data_base/RNASeq_MELHO_SKIN/SRR8615233_1.fastq.gz\n",
      "[2025-03-21 10:05:05,199]    INFO [count] Using index /Users/joeyrich/Desktop/local/RLSRWP_2025/data/vk_ref_out/vcrs_index.idx to generate BUS file to /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs from\n",
      "[2025-03-21 10:05:05,200]    INFO [count]         /Users/joeyrich/Desktop/local/RLSRWP_2025/data/ccle_data_base/RNASeq_MELHO_SKIN/SRR8615233_1.fastq.gz\n",
      "[2025-03-21 10:15:23,355]    INFO [count] Sorting BUS file /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/output.bus to /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/tmp/output.s.bus\n",
      "[2025-03-21 10:15:25,097]    INFO [count] Inspecting BUS file /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/tmp/output.s.bus\n",
      "[2025-03-21 10:15:26,209]    INFO [count] Generating count matrix /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/counts_unfiltered/cells_x_genes from BUS file /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/tmp/output.s.bus\n",
      "[2025-03-21 10:15:50,712]    INFO [count] Writing gene names to file /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/counts_unfiltered/cells_x_genes.genes.names.txt\n",
      "[2025-03-21 10:15:54,605] WARNING [count] 5329695 gene IDs do not have corresponding valid gene names. These genes will use their gene IDs instead.\n",
      "[2025-03-21 10:15:55,371]    INFO [count] Reading matrix /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/counts_unfiltered/cells_x_genes.mtx\n",
      "[2025-03-21 10:16:25,321]    INFO [count] Writing matrix to h5ad /Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_vcrs/counts_unfiltered/adata.h5ad\n",
      "10:16:27 - INFO - Skipping kb count for reference genome because the reference genome adata object was not needed and/or the file '/Users/joeyrich/Desktop/local/RLSRWP_2025/data/varseek_count_out_alignment_visualization_full/kb_count_out_reference_genome/counts_unfiltered/adata.h5ad' already exists. Note that even setting overwrite=True will still not overwrite this particular directory\n",
      "10:16:27 - INFO - Skipping vk clean because disable_clean=True\n",
      "10:16:27 - INFO - Skipping vk summarize because disable_summarize=True\n",
      "10:16:27 - INFO - Total runtime for vk count: 11m, 33.73s\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(adata_path):\n",
    "    vk_count_output_dict = vk.count(\n",
    "        fastq_file,\n",
    "        index=vcrs_index,\n",
    "        t2g=vcrs_t2g,\n",
    "        technology=technology,\n",
    "        out=vk_count_out_dir,\n",
    "        k=k,\n",
    "        strand=strand,\n",
    "        parity=parity,\n",
    "        threads=threads,\n",
    "        disable_fastqpp=True,\n",
    "        disable_clean=True,\n",
    "        disable_summarize=True\n",
    "        # quality_control_fastqs=True, cut_front=True, cut_tail=True  # equivalent to the fastp step above\n",
    "        # qc_against_gene_matrix=qc_against_gene_matrix, reference_genome_index=reference_genome_index, reference_genome_t2g=reference_genome_t2g,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6038ec84",
   "metadata": {},
   "source": [
    "### Load in adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "0dc1900e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1 × 5329695"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = anndata.read_h5ad(adata_path)\n",
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fef39c0",
   "metadata": {},
   "source": [
    "### View top non-mitochondrial 15 variants by total reads aligned - 5 substitution, 5 deletion, 5 insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ec5f4dbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "efea39e8-2063-4379-a288-e7e4865fa52e",
       "rows": [
        [
         "0",
         "ENST00000396153",
         "c.1468C>T",
         "p.Q490*",
         "COSV105286190",
         "1",
         "-",
         "g.51001067G>A",
         "FAF1",
         "48293736",
         "c.1920C>T",
         "ENST00000396153:c.1920C>T"
        ],
        [
         "1",
         "ENST00000445907",
         "c.162C>A",
         "p.L54=",
         "COSV57765075",
         "7",
         "+",
         "g.136699774C>A",
         "CHRM2",
         "60222372",
         "c.690C>A",
         "ENST00000445907:c.690C>A"
        ],
        [
         "2",
         "ENST00000445907",
         "c.617A>T",
         "p.Y206F",
         "COSV108138517",
         "7",
         "+",
         "g.136700229A>T",
         "CHRM2",
         "60229878",
         "c.1145A>T",
         "ENST00000445907:c.1145A>T"
        ],
        [
         "3",
         "ENST00000445907",
         "c.93A>T",
         "p.G31=",
         "COSV100282253",
         "7",
         "+",
         "g.136699705A>T",
         "CHRM2",
         "60227007",
         "c.621A>T",
         "ENST00000445907:c.621A>T"
        ],
        [
         "4",
         "ENST00000445907",
         "c.956A>C",
         "p.N319T",
         "COSV108138508",
         "7",
         "+",
         "g.136700568A>C",
         "CHRM2",
         "60213402",
         "c.1484A>C",
         "ENST00000445907:c.1484A>C"
        ]
       ],
       "shape": {
        "columns": 11,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000396153</td>\n",
       "      <td>c.1468C&gt;T</td>\n",
       "      <td>p.Q490*</td>\n",
       "      <td>COSV105286190</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.51001067G&gt;A</td>\n",
       "      <td>FAF1</td>\n",
       "      <td>48293736</td>\n",
       "      <td>c.1920C&gt;T</td>\n",
       "      <td>ENST00000396153:c.1920C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.162C&gt;A</td>\n",
       "      <td>p.L54=</td>\n",
       "      <td>COSV57765075</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699774C&gt;A</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60222372</td>\n",
       "      <td>c.690C&gt;A</td>\n",
       "      <td>ENST00000445907:c.690C&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.617A&gt;T</td>\n",
       "      <td>p.Y206F</td>\n",
       "      <td>COSV108138517</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700229A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60229878</td>\n",
       "      <td>c.1145A&gt;T</td>\n",
       "      <td>ENST00000445907:c.1145A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.93A&gt;T</td>\n",
       "      <td>p.G31=</td>\n",
       "      <td>COSV100282253</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136699705A&gt;T</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60227007</td>\n",
       "      <td>c.621A&gt;T</td>\n",
       "      <td>ENST00000445907:c.621A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000445907</td>\n",
       "      <td>c.956A&gt;C</td>\n",
       "      <td>p.N319T</td>\n",
       "      <td>COSV108138508</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.136700568A&gt;C</td>\n",
       "      <td>CHRM2</td>\n",
       "      <td>60213402</td>\n",
       "      <td>c.1484A&gt;C</td>\n",
       "      <td>ENST00000445907:c.1484A&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            seq_ID   mutation mutation_aa GENOMIC_MUTATION_ID chromosome  \\\n",
       "0  ENST00000396153  c.1468C>T     p.Q490*       COSV105286190          1   \n",
       "1  ENST00000445907   c.162C>A      p.L54=        COSV57765075          7   \n",
       "2  ENST00000445907   c.617A>T     p.Y206F       COSV108138517          7   \n",
       "3  ENST00000445907    c.93A>T      p.G31=       COSV100282253          7   \n",
       "4  ENST00000445907   c.956A>C     p.N319T       COSV108138508          7   \n",
       "\n",
       "  strand mutation_genome gene_name  mutation_id mutation_cdna  \\\n",
       "0      -   g.51001067G>A      FAF1     48293736     c.1920C>T   \n",
       "1      +  g.136699774C>A     CHRM2     60222372      c.690C>A   \n",
       "2      +  g.136700229A>T     CHRM2     60229878     c.1145A>T   \n",
       "3      +  g.136699705A>T     CHRM2     60227007      c.621A>T   \n",
       "4      +  g.136700568A>C     CHRM2     60213402     c.1484A>C   \n",
       "\n",
       "                      header  \n",
       "0  ENST00000396153:c.1920C>T  \n",
       "1   ENST00000445907:c.690C>A  \n",
       "2  ENST00000445907:c.1145A>T  \n",
       "3   ENST00000445907:c.621A>T  \n",
       "4  ENST00000445907:c.1484A>C  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "dd33e1aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map cosmic_df['gene_name'] to adata.var using 'header'\n",
    "adata.var['header'] = adata.var_names\n",
    "cosmic_df_gene_map = cosmic_df.drop_duplicates('header').set_index('header')[['gene_name', 'variant_type']]\n",
    "adata.var = adata.var.merge(cosmic_df_gene_map, on='header', how='left').set_index(adata.var.index)\n",
    "adata.var.index = adata.var['header']\n",
    "\n",
    "# Filter out columns where gene_name starts with \"mt-\" (case-insensitive)\n",
    "mask = ~adata.var['gene_name'].astype(str).str.upper().str.startswith('MT-', na=False)\n",
    "adata = adata[:, mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "6bb6edfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Variant",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Counts",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Gene_Name",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "a4c9a94b-912a-4f64-86db-5e0d0897eef3",
       "rows": [
        [
         "0",
         "ENST00000398606:c.562A>G",
         "13934",
         "GSTP1"
        ],
        [
         "1",
         "ENST00000449260:c.1988C>T",
         "12589",
         "PMEL"
        ],
        [
         "2",
         "ENST00000251453:c.80T>G",
         "11981",
         "RPS16"
        ],
        [
         "3",
         "ENST00000398606:c.804T>C",
         "11300",
         "GSTP1"
        ],
        [
         "4",
         "ENST00000343262:c.156C>T",
         "6296",
         "RPS2"
        ],
        [
         "5",
         "ENST00000322723:c.1024_1026del",
         "2501",
         "NCL"
        ],
        [
         "6",
         "ENST00000263574:c.750_755del",
         "254",
         "APLP2"
        ],
        [
         "7",
         "ENST00000367034:c.152_154del",
         "216",
         "MRPL18"
        ],
        [
         "8",
         "ENST00000314222:c.448_449insCCCCC",
         "145",
         "PHLDA2"
        ],
        [
         "9",
         "ENST00000373477:c.2126_2127insCGTAC",
         "145",
         "YARS"
        ],
        [
         "10",
         "ENST00000434724:c.239_247del",
         "115",
         "GSPT1"
        ],
        [
         "11",
         "ENST00000529010:c.1426_1428del",
         "108",
         "NUCB2"
        ],
        [
         "12",
         "ENST00000262710:c.2258_2259insATTCAC",
         "82",
         "ACIN1"
        ],
        [
         "13",
         "ENST00000372797:c.1497_1498insG",
         "27",
         "CAP1"
        ],
        [
         "14",
         "ENST00000435765:c.1262_1263insGTT",
         "26",
         "PSMC2"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 15
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Counts</th>\n",
       "      <th>Gene_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000398606:c.562A&gt;G</td>\n",
       "      <td>13934</td>\n",
       "      <td>GSTP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000449260:c.1988C&gt;T</td>\n",
       "      <td>12589</td>\n",
       "      <td>PMEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "      <td>11981</td>\n",
       "      <td>RPS16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000398606:c.804T&gt;C</td>\n",
       "      <td>11300</td>\n",
       "      <td>GSTP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000343262:c.156C&gt;T</td>\n",
       "      <td>6296</td>\n",
       "      <td>RPS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENST00000322723:c.1024_1026del</td>\n",
       "      <td>2501</td>\n",
       "      <td>NCL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENST00000263574:c.750_755del</td>\n",
       "      <td>254</td>\n",
       "      <td>APLP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENST00000367034:c.152_154del</td>\n",
       "      <td>216</td>\n",
       "      <td>MRPL18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENST00000314222:c.448_449insCCCCC</td>\n",
       "      <td>145</td>\n",
       "      <td>PHLDA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENST00000373477:c.2126_2127insCGTAC</td>\n",
       "      <td>145</td>\n",
       "      <td>YARS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENST00000434724:c.239_247del</td>\n",
       "      <td>115</td>\n",
       "      <td>GSPT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENST00000529010:c.1426_1428del</td>\n",
       "      <td>108</td>\n",
       "      <td>NUCB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENST00000262710:c.2258_2259insATTCAC</td>\n",
       "      <td>82</td>\n",
       "      <td>ACIN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENST00000372797:c.1497_1498insG</td>\n",
       "      <td>27</td>\n",
       "      <td>CAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENST00000435765:c.1262_1263insGTT</td>\n",
       "      <td>26</td>\n",
       "      <td>PSMC2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 Variant  Counts Gene_Name\n",
       "0               ENST00000398606:c.562A>G   13934     GSTP1\n",
       "1              ENST00000449260:c.1988C>T   12589      PMEL\n",
       "2                ENST00000251453:c.80T>G   11981     RPS16\n",
       "3               ENST00000398606:c.804T>C   11300     GSTP1\n",
       "4               ENST00000343262:c.156C>T    6296      RPS2\n",
       "5         ENST00000322723:c.1024_1026del    2501       NCL\n",
       "6           ENST00000263574:c.750_755del     254     APLP2\n",
       "7           ENST00000367034:c.152_154del     216    MRPL18\n",
       "8      ENST00000314222:c.448_449insCCCCC     145    PHLDA2\n",
       "9    ENST00000373477:c.2126_2127insCGTAC     145      YARS\n",
       "10          ENST00000434724:c.239_247del     115     GSPT1\n",
       "11        ENST00000529010:c.1426_1428del     108     NUCB2\n",
       "12  ENST00000262710:c.2258_2259insATTCAC      82     ACIN1\n",
       "13       ENST00000372797:c.1497_1498insG      27      CAP1\n",
       "14     ENST00000435765:c.1262_1263insGTT      26     PSMC2"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_to_select_per_mutation_type = 5\n",
    "top15, top15_counts, top15_gene_names = [], [], []\n",
    "\n",
    "for mutation_type in [\"substitution\", \"insertion\", \"deletion\"]:\n",
    "    adata_tmp = adata[:, adata.var['variant_type'] == mutation_type].copy()  # select only mutation_type\n",
    "    variants_list = list(adata_tmp.var.index[np.argsort(np.array(adata_tmp.X.todense())[0])][::-1][:number_to_select_per_mutation_type])\n",
    "    top15.extend(variants_list)\n",
    "    top15_counts.extend(list(np.array(adata_tmp.X.todense())[0][np.argsort(np.array(adata_tmp.X.todense())[0])][::-1][:number_to_select_per_mutation_type].astype(int)))\n",
    "    top15_gene_names.extend(cosmic_df.set_index('header').loc[variants_list, 'gene_name'].tolist())\n",
    "\n",
    "top15_dict = dict(zip(top15, top15_counts))\n",
    "\n",
    "top15_df = pd.DataFrame({\n",
    "    'Variant': top15,\n",
    "    'Counts': top15_counts,\n",
    "    'Gene_Name': top15_gene_names\n",
    "})\n",
    "\n",
    "top15 = top15_df.sort_values(by=\"Counts\", ascending=False)[\"Variant\"].tolist()  # sort descending order\n",
    "top15_df = top15_df.sort_values(by=\"Counts\", ascending=False).reset_index(drop=True)\n",
    "top15_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baee92d7",
   "metadata": {},
   "source": [
    "### Map the reads to the VCRS to which they aligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "2cf1c7a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/3798617008.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  filtered_bus_df[\"variant_names_str\"] = filtered_bus_df[\"variant_names\"].apply(lambda x: x[0])  # cast to string\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reads in filtered bus file: 59719\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "barcode",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "UMI",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "EC",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "read_index",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "fastq_header",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "file_index",
         "rawType": "category",
         "type": "unknown"
        },
        {
         "name": "variant_names",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "counted_in_count_matrix",
         "rawType": "bool",
         "type": "boolean"
        },
        {
         "name": "variant_names_str",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "bc4848d0-a6ef-494d-a8a3-a0e6b68ae26c",
       "rows": [
        [
         "9665",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917225",
         "SRR8615233.917226",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9666",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917227",
         "SRR8615233.917228",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9667",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917230",
         "SRR8615233.917231",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9668",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917231",
         "SRR8615233.917232",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ],
        [
         "9669",
         "AAAAAAAAAAAAAAAA",
         "T",
         "1702",
         "917232",
         "SRR8615233.917233",
         "0",
         "['ENST00000251453:c.80T>G']",
         "True",
         "ENST00000251453:c.80T>G"
        ]
       ],
       "shape": {
        "columns": 9,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>UMI</th>\n",
       "      <th>EC</th>\n",
       "      <th>read_index</th>\n",
       "      <th>fastq_header</th>\n",
       "      <th>file_index</th>\n",
       "      <th>variant_names</th>\n",
       "      <th>counted_in_count_matrix</th>\n",
       "      <th>variant_names_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9665</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917225</td>\n",
       "      <td>SRR8615233.917226</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9666</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917227</td>\n",
       "      <td>SRR8615233.917228</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9667</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917230</td>\n",
       "      <td>SRR8615233.917231</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9668</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917231</td>\n",
       "      <td>SRR8615233.917232</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9669</th>\n",
       "      <td>AAAAAAAAAAAAAAAA</td>\n",
       "      <td>T</td>\n",
       "      <td>1702</td>\n",
       "      <td>917232</td>\n",
       "      <td>SRR8615233.917233</td>\n",
       "      <td>0</td>\n",
       "      <td>[ENST00000251453:c.80T&gt;G]</td>\n",
       "      <td>True</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               barcode UMI    EC read_index       fastq_header file_index  \\\n",
       "9665  AAAAAAAAAAAAAAAA   T  1702     917225  SRR8615233.917226          0   \n",
       "9666  AAAAAAAAAAAAAAAA   T  1702     917227  SRR8615233.917228          0   \n",
       "9667  AAAAAAAAAAAAAAAA   T  1702     917230  SRR8615233.917231          0   \n",
       "9668  AAAAAAAAAAAAAAAA   T  1702     917231  SRR8615233.917232          0   \n",
       "9669  AAAAAAAAAAAAAAAA   T  1702     917232  SRR8615233.917233          0   \n",
       "\n",
       "                  variant_names  counted_in_count_matrix  \\\n",
       "9665  [ENST00000251453:c.80T>G]                     True   \n",
       "9666  [ENST00000251453:c.80T>G]                     True   \n",
       "9667  [ENST00000251453:c.80T>G]                     True   \n",
       "9668  [ENST00000251453:c.80T>G]                     True   \n",
       "9669  [ENST00000251453:c.80T>G]                     True   \n",
       "\n",
       "            variant_names_str  \n",
       "9665  ENST00000251453:c.80T>G  \n",
       "9666  ENST00000251453:c.80T>G  \n",
       "9667  ENST00000251453:c.80T>G  \n",
       "9668  ENST00000251453:c.80T>G  \n",
       "9669  ENST00000251453:c.80T>G  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bus_df = make_bus_df(kb_count_out_dir, fastq_file, technology=technology, parity=parity, bustools=bustools)\n",
    "bus_df = bus_df.drop(columns=[\"transcript_names\"]).rename(columns={\"gene_names\": \"variant_names\"})\n",
    "filtered_bus_df = bus_df[bus_df[\"variant_names\"].apply(lambda x: len(x) == 1)]  # remove multi-mapping reads\n",
    "filtered_bus_df[\"variant_names_str\"] = filtered_bus_df[\"variant_names\"].apply(lambda x: x[0])  # cast to string\n",
    "filtered_bus_df = filtered_bus_df[(filtered_bus_df[\"variant_names_str\"].isin(top15)) & (filtered_bus_df[\"counted_in_count_matrix\"])]\n",
    "print(f\"Number of reads in filtered bus file: {len(filtered_bus_df)}\")\n",
    "filtered_bus_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "32c55d55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENST00000398606:c.562A>G': 'vcrs_0',\n",
       " 'ENST00000449260:c.1988C>T': 'vcrs_1',\n",
       " 'ENST00000251453:c.80T>G': 'vcrs_2',\n",
       " 'ENST00000398606:c.804T>C': 'vcrs_3',\n",
       " 'ENST00000343262:c.156C>T': 'vcrs_4',\n",
       " 'ENST00000322723:c.1024_1026del': 'vcrs_5',\n",
       " 'ENST00000263574:c.750_755del': 'vcrs_6',\n",
       " 'ENST00000367034:c.152_154del': 'vcrs_7',\n",
       " 'ENST00000314222:c.448_449insCCCCC': 'vcrs_8',\n",
       " 'ENST00000373477:c.2126_2127insCGTAC': 'vcrs_9',\n",
       " 'ENST00000434724:c.239_247del': 'vcrs_10',\n",
       " 'ENST00000529010:c.1426_1428del': 'vcrs_11',\n",
       " 'ENST00000262710:c.2258_2259insATTCAC': 'vcrs_12',\n",
       " 'ENST00000372797:c.1497_1498insG': 'vcrs_13',\n",
       " 'ENST00000435765:c.1262_1263insGTT': 'vcrs_14'}"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make IDs since the gene names in HGVSC format do not make for good folder names or bowtie2 header names\n",
    "vcrs_header_to_id = {vcrs: f\"vcrs_{i}\" for i, vcrs in enumerate(list(top15))}\n",
    "filtered_bus_df[\"vcrs_ids\"] = filtered_bus_df[\"variant_names_str\"].map(vcrs_header_to_id)\n",
    "\n",
    "vcrs_header_to_id_file = os.path.join(vk_count_out_dir, \"vcrs_header_to_id.txt\")\n",
    "with open(vcrs_header_to_id_file, \"w\") as f:\n",
    "    for key, value in vcrs_header_to_id.items():\n",
    "        f.write(f\"{key}\\t{value}\\n\")  # Tab-separated\n",
    "\n",
    "id_to_vcrs_header = {v: k for k, v in vcrs_header_to_id.items()}\n",
    "vcrs_header_to_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "b504ecc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing vcrs_0\n",
      "Processing vcrs_1\n",
      "Processing vcrs_10\n",
      "Processing vcrs_11\n",
      "Processing vcrs_12\n",
      "Processing vcrs_13\n",
      "Processing vcrs_14\n",
      "Processing vcrs_2\n",
      "Processing vcrs_3\n",
      "Processing vcrs_4\n",
      "Processing vcrs_5\n",
      "Processing vcrs_6\n",
      "Processing vcrs_7\n",
      "Processing vcrs_8\n",
      "Processing vcrs_9\n"
     ]
    }
   ],
   "source": [
    "for vcrs_id in sorted(filtered_bus_df[\"vcrs_ids\"].unique()):  # Get unique gene names\n",
    "    print(f\"Processing {vcrs_id}\")\n",
    "    temp_df = filtered_bus_df[filtered_bus_df[\"vcrs_ids\"] == vcrs_id]  # Filter\n",
    "    fastq_headers = temp_df[\"fastq_header\"].tolist()  # Get values as a list\n",
    "\n",
    "    gene_dir = os.path.join(aligned_reads_parent_dir, vcrs_id)\n",
    "    os.makedirs(gene_dir, exist_ok=True)\n",
    "    \n",
    "    aligned_reads_file = os.path.join(gene_dir, \"1.fastq\")\n",
    "    with open(aligned_reads_file, \"w\") as f:\n",
    "        for header in fastq_headers:\n",
    "            sequence = fastq_indexed[header].seq\n",
    "            qualities = fastq_indexed[header].qual\n",
    "            f.write(f\"@{header}\\n{sequence}\\n+\\n{qualities}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f1ecf2",
   "metadata": {},
   "source": [
    "### View COSMIC metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "107a9752",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/2740002178.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cosmic_df_subset[\"id\"] = cosmic_df_subset[\"header\"].map(vcrs_header_to_id)\n",
      "/var/folders/wr/wdr46lxn7fbc_kty6kwbfhxc0000gn/T/ipykernel_28337/2740002178.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cosmic_df_subset[\"counts\"] = cosmic_df_subset[\"header\"].map(lambda x: top15_dict[x])\n"
     ]
    },
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "seq_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_aa",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "GENOMIC_MUTATION_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "strand",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "mutation_cdna",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "header",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "counts",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "d2d0fb6d-3b42-4d13-9c63-c9d1475cd228",
       "rows": [
        [
         "5118134",
         "ENST00000398606",
         "c.313A>G",
         "p.I105V",
         "COSV66992376",
         "11",
         "+",
         "g.67352689A>G",
         "GSTP1",
         "48379894",
         "c.562A>G",
         "ENST00000398606:c.562A>G",
         "substitution",
         "vcrs_0",
         "13934"
        ],
        [
         "1545555",
         "ENST00000449260",
         "c.1977C>T",
         "p.N659=",
         "COSV59384225",
         "12",
         "-",
         "g.56348028G>A",
         "PMEL",
         "60961346",
         "c.1988C>T",
         "ENST00000449260:c.1988C>T",
         "substitution",
         "vcrs_1",
         "12589"
        ],
        [
         "4199446",
         "ENST00000251453",
         "c.27T>G",
         "p.S9=",
         "COSV52238647",
         "19",
         "-",
         "g.39926509A>C",
         "RPS16",
         "21768506",
         "c.80T>G",
         "ENST00000251453:c.80T>G",
         "substitution",
         "vcrs_2",
         "11981"
        ],
        [
         "5118182",
         "ENST00000398606",
         "c.555T>C",
         "p.S185=",
         "COSV66992478",
         "11",
         "+",
         "g.67353970T>C",
         "GSTP1",
         "48379878",
         "c.804T>C",
         "ENST00000398606:c.804T>C",
         "substitution",
         "vcrs_3",
         "11300"
        ],
        [
         "3224064",
         "ENST00000343262",
         "c.99C>T",
         "p.I33=",
         "COSV51909983",
         "16",
         "-",
         "g.2014528G>A",
         "RPS2",
         "31955826",
         "c.156C>T",
         "ENST00000343262:c.156C>T",
         "substitution",
         "vcrs_4",
         "6296"
        ],
        [
         "5239178",
         "ENST00000322723",
         "c.783_785del",
         "p.D261del",
         "COSV59544931",
         "2",
         null,
         "g.232325411_232325413del",
         "NCL",
         "26628001",
         "c.1024_1026del",
         "ENST00000322723:c.1024_1026del",
         "deletion",
         "vcrs_5",
         "2501"
        ],
        [
         "5383497",
         "ENST00000263574",
         "c.678_683del",
         "p.E226_E227del",
         "COSV53838522",
         "11",
         null,
         "g.129991670_129991675del",
         "APLP2",
         "20079349",
         "c.750_755del",
         "ENST00000263574:c.750_755del",
         "deletion",
         "vcrs_6",
         "254"
        ],
        [
         "5245332",
         "ENST00000367034",
         "c.30_32del",
         "p.L10del",
         "COSV58458382",
         "6",
         null,
         "g.160211649_160211651del",
         "MRPL18",
         "39042486",
         "c.152_154del",
         "ENST00000367034:c.152_154del",
         "deletion",
         "vcrs_7",
         "216"
        ],
        [
         "3000122",
         "ENST00000373477",
         "c.1217_1218insCGTAC",
         "p.Q408Rfs*15",
         "COSV65114720",
         "1",
         "-",
         "g.33245802_33245803insGTACG",
         "YARS",
         "39378578",
         "c.2126_2127insCGTAC",
         "ENST00000373477:c.2126_2127insCGTAC",
         "insertion",
         "vcrs_9",
         "145"
        ],
        [
         "3767335",
         "ENST00000314222",
         "c.357_358insCCCCC",
         "p.A120Pfs*50",
         "COSV56542491",
         "11",
         "-",
         "g.2950240_2950241insGGGGG",
         "PHLDA2",
         "27287821",
         "c.448_449insCCCCC",
         "ENST00000314222:c.448_449insCCCCC",
         "insertion",
         "vcrs_8",
         "145"
        ],
        [
         "5246024",
         "ENST00000434724",
         "c.39_47del",
         "p.G14_G16del",
         "COSV70453045",
         "16",
         null,
         "g.12009554_12009562del",
         "GSPT1",
         "57130237",
         "c.239_247del",
         "ENST00000434724:c.239_247del",
         "deletion",
         "vcrs_10",
         "115"
        ],
        [
         "5382296",
         "ENST00000529010",
         "c.1207_1209del",
         "p.Q403del",
         "COSV60376359",
         "11",
         null,
         "g.17352482_17352484del",
         "NUCB2",
         "81560910",
         "c.1426_1428del",
         "ENST00000529010:c.1426_1428del",
         "deletion",
         "vcrs_11",
         "108"
        ],
        [
         "548620",
         "ENST00000262710",
         "c.1930_1931insATTCAC",
         "p.S643_R644insHS",
         "COSV52972535",
         "14",
         "-",
         "g.23548792_23548793insTGTGAA",
         "ACIN1",
         "23586862",
         "c.2258_2259insATTCAC",
         "ENST00000262710:c.2258_2259insATTCAC",
         "insertion",
         "vcrs_12",
         "82"
        ],
        [
         "1823895",
         "ENST00000372797",
         "c.936_937insG",
         "p.R313Afs*11",
         "COSV61223036",
         "1",
         "+",
         "g.40535489_40535490insG",
         "CAP1",
         "40250283",
         "c.1497_1498insG",
         "ENST00000372797:c.1497_1498insG",
         "insertion",
         "vcrs_13",
         "27"
        ],
        [
         "455948",
         "ENST00000435765",
         "c.851_852insGTT",
         "p.F285dup",
         "COSV53007999",
         "7",
         "+",
         "g.103007866_103007867insGTT",
         "PSMC2",
         "55802712",
         "c.1262_1263insGTT",
         "ENST00000435765:c.1262_1263insGTT",
         "insertion",
         "vcrs_14",
         "26"
        ]
       ],
       "shape": {
        "columns": 14,
        "rows": 15
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>mutation</th>\n",
       "      <th>mutation_aa</th>\n",
       "      <th>GENOMIC_MUTATION_ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>mutation_cdna</th>\n",
       "      <th>header</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5118134</th>\n",
       "      <td>ENST00000398606</td>\n",
       "      <td>c.313A&gt;G</td>\n",
       "      <td>p.I105V</td>\n",
       "      <td>COSV66992376</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>g.67352689A&gt;G</td>\n",
       "      <td>GSTP1</td>\n",
       "      <td>48379894</td>\n",
       "      <td>c.562A&gt;G</td>\n",
       "      <td>ENST00000398606:c.562A&gt;G</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_0</td>\n",
       "      <td>13934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1545555</th>\n",
       "      <td>ENST00000449260</td>\n",
       "      <td>c.1977C&gt;T</td>\n",
       "      <td>p.N659=</td>\n",
       "      <td>COSV59384225</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>g.56348028G&gt;A</td>\n",
       "      <td>PMEL</td>\n",
       "      <td>60961346</td>\n",
       "      <td>c.1988C&gt;T</td>\n",
       "      <td>ENST00000449260:c.1988C&gt;T</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_1</td>\n",
       "      <td>12589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4199446</th>\n",
       "      <td>ENST00000251453</td>\n",
       "      <td>c.27T&gt;G</td>\n",
       "      <td>p.S9=</td>\n",
       "      <td>COSV52238647</td>\n",
       "      <td>19</td>\n",
       "      <td>-</td>\n",
       "      <td>g.39926509A&gt;C</td>\n",
       "      <td>RPS16</td>\n",
       "      <td>21768506</td>\n",
       "      <td>c.80T&gt;G</td>\n",
       "      <td>ENST00000251453:c.80T&gt;G</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_2</td>\n",
       "      <td>11981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5118182</th>\n",
       "      <td>ENST00000398606</td>\n",
       "      <td>c.555T&gt;C</td>\n",
       "      <td>p.S185=</td>\n",
       "      <td>COSV66992478</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "      <td>g.67353970T&gt;C</td>\n",
       "      <td>GSTP1</td>\n",
       "      <td>48379878</td>\n",
       "      <td>c.804T&gt;C</td>\n",
       "      <td>ENST00000398606:c.804T&gt;C</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_3</td>\n",
       "      <td>11300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3224064</th>\n",
       "      <td>ENST00000343262</td>\n",
       "      <td>c.99C&gt;T</td>\n",
       "      <td>p.I33=</td>\n",
       "      <td>COSV51909983</td>\n",
       "      <td>16</td>\n",
       "      <td>-</td>\n",
       "      <td>g.2014528G&gt;A</td>\n",
       "      <td>RPS2</td>\n",
       "      <td>31955826</td>\n",
       "      <td>c.156C&gt;T</td>\n",
       "      <td>ENST00000343262:c.156C&gt;T</td>\n",
       "      <td>substitution</td>\n",
       "      <td>vcrs_4</td>\n",
       "      <td>6296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5239178</th>\n",
       "      <td>ENST00000322723</td>\n",
       "      <td>c.783_785del</td>\n",
       "      <td>p.D261del</td>\n",
       "      <td>COSV59544931</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.232325411_232325413del</td>\n",
       "      <td>NCL</td>\n",
       "      <td>26628001</td>\n",
       "      <td>c.1024_1026del</td>\n",
       "      <td>ENST00000322723:c.1024_1026del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_5</td>\n",
       "      <td>2501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5383497</th>\n",
       "      <td>ENST00000263574</td>\n",
       "      <td>c.678_683del</td>\n",
       "      <td>p.E226_E227del</td>\n",
       "      <td>COSV53838522</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.129991670_129991675del</td>\n",
       "      <td>APLP2</td>\n",
       "      <td>20079349</td>\n",
       "      <td>c.750_755del</td>\n",
       "      <td>ENST00000263574:c.750_755del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_6</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5245332</th>\n",
       "      <td>ENST00000367034</td>\n",
       "      <td>c.30_32del</td>\n",
       "      <td>p.L10del</td>\n",
       "      <td>COSV58458382</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.160211649_160211651del</td>\n",
       "      <td>MRPL18</td>\n",
       "      <td>39042486</td>\n",
       "      <td>c.152_154del</td>\n",
       "      <td>ENST00000367034:c.152_154del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_7</td>\n",
       "      <td>216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3000122</th>\n",
       "      <td>ENST00000373477</td>\n",
       "      <td>c.1217_1218insCGTAC</td>\n",
       "      <td>p.Q408Rfs*15</td>\n",
       "      <td>COSV65114720</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>g.33245802_33245803insGTACG</td>\n",
       "      <td>YARS</td>\n",
       "      <td>39378578</td>\n",
       "      <td>c.2126_2127insCGTAC</td>\n",
       "      <td>ENST00000373477:c.2126_2127insCGTAC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_9</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3767335</th>\n",
       "      <td>ENST00000314222</td>\n",
       "      <td>c.357_358insCCCCC</td>\n",
       "      <td>p.A120Pfs*50</td>\n",
       "      <td>COSV56542491</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>g.2950240_2950241insGGGGG</td>\n",
       "      <td>PHLDA2</td>\n",
       "      <td>27287821</td>\n",
       "      <td>c.448_449insCCCCC</td>\n",
       "      <td>ENST00000314222:c.448_449insCCCCC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_8</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5246024</th>\n",
       "      <td>ENST00000434724</td>\n",
       "      <td>c.39_47del</td>\n",
       "      <td>p.G14_G16del</td>\n",
       "      <td>COSV70453045</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.12009554_12009562del</td>\n",
       "      <td>GSPT1</td>\n",
       "      <td>57130237</td>\n",
       "      <td>c.239_247del</td>\n",
       "      <td>ENST00000434724:c.239_247del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_10</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5382296</th>\n",
       "      <td>ENST00000529010</td>\n",
       "      <td>c.1207_1209del</td>\n",
       "      <td>p.Q403del</td>\n",
       "      <td>COSV60376359</td>\n",
       "      <td>11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>g.17352482_17352484del</td>\n",
       "      <td>NUCB2</td>\n",
       "      <td>81560910</td>\n",
       "      <td>c.1426_1428del</td>\n",
       "      <td>ENST00000529010:c.1426_1428del</td>\n",
       "      <td>deletion</td>\n",
       "      <td>vcrs_11</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548620</th>\n",
       "      <td>ENST00000262710</td>\n",
       "      <td>c.1930_1931insATTCAC</td>\n",
       "      <td>p.S643_R644insHS</td>\n",
       "      <td>COSV52972535</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "      <td>g.23548792_23548793insTGTGAA</td>\n",
       "      <td>ACIN1</td>\n",
       "      <td>23586862</td>\n",
       "      <td>c.2258_2259insATTCAC</td>\n",
       "      <td>ENST00000262710:c.2258_2259insATTCAC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_12</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1823895</th>\n",
       "      <td>ENST00000372797</td>\n",
       "      <td>c.936_937insG</td>\n",
       "      <td>p.R313Afs*11</td>\n",
       "      <td>COSV61223036</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>g.40535489_40535490insG</td>\n",
       "      <td>CAP1</td>\n",
       "      <td>40250283</td>\n",
       "      <td>c.1497_1498insG</td>\n",
       "      <td>ENST00000372797:c.1497_1498insG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_13</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455948</th>\n",
       "      <td>ENST00000435765</td>\n",
       "      <td>c.851_852insGTT</td>\n",
       "      <td>p.F285dup</td>\n",
       "      <td>COSV53007999</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "      <td>g.103007866_103007867insGTT</td>\n",
       "      <td>PSMC2</td>\n",
       "      <td>55802712</td>\n",
       "      <td>c.1262_1263insGTT</td>\n",
       "      <td>ENST00000435765:c.1262_1263insGTT</td>\n",
       "      <td>insertion</td>\n",
       "      <td>vcrs_14</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  seq_ID              mutation       mutation_aa  \\\n",
       "5118134  ENST00000398606              c.313A>G           p.I105V   \n",
       "1545555  ENST00000449260             c.1977C>T           p.N659=   \n",
       "4199446  ENST00000251453               c.27T>G             p.S9=   \n",
       "5118182  ENST00000398606              c.555T>C           p.S185=   \n",
       "3224064  ENST00000343262               c.99C>T            p.I33=   \n",
       "5239178  ENST00000322723          c.783_785del         p.D261del   \n",
       "5383497  ENST00000263574          c.678_683del    p.E226_E227del   \n",
       "5245332  ENST00000367034            c.30_32del          p.L10del   \n",
       "3000122  ENST00000373477   c.1217_1218insCGTAC      p.Q408Rfs*15   \n",
       "3767335  ENST00000314222     c.357_358insCCCCC      p.A120Pfs*50   \n",
       "5246024  ENST00000434724            c.39_47del      p.G14_G16del   \n",
       "5382296  ENST00000529010        c.1207_1209del         p.Q403del   \n",
       "548620   ENST00000262710  c.1930_1931insATTCAC  p.S643_R644insHS   \n",
       "1823895  ENST00000372797         c.936_937insG      p.R313Afs*11   \n",
       "455948   ENST00000435765       c.851_852insGTT         p.F285dup   \n",
       "\n",
       "        GENOMIC_MUTATION_ID chromosome strand               mutation_genome  \\\n",
       "5118134        COSV66992376         11      +                 g.67352689A>G   \n",
       "1545555        COSV59384225         12      -                 g.56348028G>A   \n",
       "4199446        COSV52238647         19      -                 g.39926509A>C   \n",
       "5118182        COSV66992478         11      +                 g.67353970T>C   \n",
       "3224064        COSV51909983         16      -                  g.2014528G>A   \n",
       "5239178        COSV59544931          2    NaN      g.232325411_232325413del   \n",
       "5383497        COSV53838522         11    NaN      g.129991670_129991675del   \n",
       "5245332        COSV58458382          6    NaN      g.160211649_160211651del   \n",
       "3000122        COSV65114720          1      -   g.33245802_33245803insGTACG   \n",
       "3767335        COSV56542491         11      -     g.2950240_2950241insGGGGG   \n",
       "5246024        COSV70453045         16    NaN        g.12009554_12009562del   \n",
       "5382296        COSV60376359         11    NaN        g.17352482_17352484del   \n",
       "548620         COSV52972535         14      -  g.23548792_23548793insTGTGAA   \n",
       "1823895        COSV61223036          1      +       g.40535489_40535490insG   \n",
       "455948         COSV53007999          7      +   g.103007866_103007867insGTT   \n",
       "\n",
       "        gene_name  mutation_id         mutation_cdna  \\\n",
       "5118134     GSTP1     48379894              c.562A>G   \n",
       "1545555      PMEL     60961346             c.1988C>T   \n",
       "4199446     RPS16     21768506               c.80T>G   \n",
       "5118182     GSTP1     48379878              c.804T>C   \n",
       "3224064      RPS2     31955826              c.156C>T   \n",
       "5239178       NCL     26628001        c.1024_1026del   \n",
       "5383497     APLP2     20079349          c.750_755del   \n",
       "5245332    MRPL18     39042486          c.152_154del   \n",
       "3000122      YARS     39378578   c.2126_2127insCGTAC   \n",
       "3767335    PHLDA2     27287821     c.448_449insCCCCC   \n",
       "5246024     GSPT1     57130237          c.239_247del   \n",
       "5382296     NUCB2     81560910        c.1426_1428del   \n",
       "548620      ACIN1     23586862  c.2258_2259insATTCAC   \n",
       "1823895      CAP1     40250283       c.1497_1498insG   \n",
       "455948      PSMC2     55802712     c.1262_1263insGTT   \n",
       "\n",
       "                                       header  variant_type       id  counts  \n",
       "5118134              ENST00000398606:c.562A>G  substitution   vcrs_0   13934  \n",
       "1545555             ENST00000449260:c.1988C>T  substitution   vcrs_1   12589  \n",
       "4199446               ENST00000251453:c.80T>G  substitution   vcrs_2   11981  \n",
       "5118182              ENST00000398606:c.804T>C  substitution   vcrs_3   11300  \n",
       "3224064              ENST00000343262:c.156C>T  substitution   vcrs_4    6296  \n",
       "5239178        ENST00000322723:c.1024_1026del      deletion   vcrs_5    2501  \n",
       "5383497          ENST00000263574:c.750_755del      deletion   vcrs_6     254  \n",
       "5245332          ENST00000367034:c.152_154del      deletion   vcrs_7     216  \n",
       "3000122   ENST00000373477:c.2126_2127insCGTAC     insertion   vcrs_9     145  \n",
       "3767335     ENST00000314222:c.448_449insCCCCC     insertion   vcrs_8     145  \n",
       "5246024          ENST00000434724:c.239_247del      deletion  vcrs_10     115  \n",
       "5382296        ENST00000529010:c.1426_1428del      deletion  vcrs_11     108  \n",
       "548620   ENST00000262710:c.2258_2259insATTCAC     insertion  vcrs_12      82  \n",
       "1823895       ENST00000372797:c.1497_1498insG     insertion  vcrs_13      27  \n",
       "455948      ENST00000435765:c.1262_1263insGTT     insertion  vcrs_14      26  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df_subset = cosmic_df[cosmic_df[\"header\"].isin(top15)]\n",
    "cosmic_df_subset[\"id\"] = cosmic_df_subset[\"header\"].map(vcrs_header_to_id)\n",
    "cosmic_df_subset[\"counts\"] = cosmic_df_subset[\"header\"].map(lambda x: top15_dict[x])\n",
    "cosmic_df_subset = cosmic_df_subset.sort_values(by=\"counts\", ascending=False)\n",
    "cosmic_df_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "487178cb-9825-4c23-9551-baf7a6934f9c",
   "metadata": {},
   "source": [
    "### Align (1) pulled out reads and (2) the VCRSs to the human genome to generate bam files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "28187021-d85a-4bd0-9892-7fd5e20e46e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcrs_0 (ENST00000398606:c.562A>G)\n",
      "\n",
      "13934 reads; of these:\n",
      "  13934 (100.00%) were unpaired; of these:\n",
      "    6542 (46.95%) aligned 0 times\n",
      "    7379 (52.96%) aligned exactly 1 time\n",
      "    13 (0.09%) aligned >1 times\n",
      "53.05% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_1 (ENST00000449260:c.1988C>T)\n",
      "\n",
      "12589 reads; of these:\n",
      "  12589 (100.00%) were unpaired; of these:\n",
      "    39 (0.31%) aligned 0 times\n",
      "    12550 (99.69%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "99.69% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_10 (ENST00000434724:c.239_247del)\n",
      "\n",
      "115 reads; of these:\n",
      "  115 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    115 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_11 (ENST00000529010:c.1426_1428del)\n",
      "\n",
      "108 reads; of these:\n",
      "  108 (100.00%) were unpaired; of these:\n",
      "    102 (94.44%) aligned 0 times\n",
      "    6 (5.56%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "5.56% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_12 (ENST00000262710:c.2258_2259insATTCAC)\n",
      "\n",
      "82 reads; of these:\n",
      "  82 (100.00%) were unpaired; of these:\n",
      "    38 (46.34%) aligned 0 times\n",
      "    44 (53.66%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "53.66% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_13 (ENST00000372797:c.1497_1498insG)\n",
      "\n",
      "27 reads; of these:\n",
      "  27 (100.00%) were unpaired; of these:\n",
      "    1 (3.70%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    26 (96.30%) aligned >1 times\n",
      "96.30% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_14 (ENST00000435765:c.1262_1263insGTT)\n",
      "\n",
      "26 reads; of these:\n",
      "  26 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    26 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_2 (ENST00000251453:c.80T>G)\n",
      "\n",
      "11981 reads; of these:\n",
      "  11981 (100.00%) were unpaired; of these:\n",
      "    53 (0.44%) aligned 0 times\n",
      "    1151 (9.61%) aligned exactly 1 time\n",
      "    10777 (89.95%) aligned >1 times\n",
      "99.56% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_3 (ENST00000398606:c.804T>C)\n",
      "\n",
      "11300 reads; of these:\n",
      "  11300 (100.00%) were unpaired; of these:\n",
      "    45 (0.40%) aligned 0 times\n",
      "    7958 (70.42%) aligned exactly 1 time\n",
      "    3297 (29.18%) aligned >1 times\n",
      "99.60% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_4 (ENST00000343262:c.156C>T)\n",
      "\n",
      "6296 reads; of these:\n",
      "  6296 (100.00%) were unpaired; of these:\n",
      "    16 (0.25%) aligned 0 times\n",
      "    4 (0.06%) aligned exactly 1 time\n",
      "    6276 (99.68%) aligned >1 times\n",
      "99.75% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    1 (100.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_5 (ENST00000322723:c.1024_1026del)\n",
      "\n",
      "2501 reads; of these:\n",
      "  2501 (100.00%) were unpaired; of these:\n",
      "    1682 (67.25%) aligned 0 times\n",
      "    816 (32.63%) aligned exactly 1 time\n",
      "    3 (0.12%) aligned >1 times\n",
      "32.75% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_6 (ENST00000263574:c.750_755del)\n",
      "\n",
      "254 reads; of these:\n",
      "  254 (100.00%) were unpaired; of these:\n",
      "    96 (37.80%) aligned 0 times\n",
      "    158 (62.20%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "62.20% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_7 (ENST00000367034:c.152_154del)\n",
      "\n",
      "216 reads; of these:\n",
      "  216 (100.00%) were unpaired; of these:\n",
      "    209 (96.76%) aligned 0 times\n",
      "    7 (3.24%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "3.24% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    1 (100.00%) aligned 0 times\n",
      "    0 (0.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "0.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_8 (ENST00000314222:c.448_449insCCCCC)\n",
      "\n",
      "145 reads; of these:\n",
      "  145 (100.00%) were unpaired; of these:\n",
      "    1 (0.69%) aligned 0 times\n",
      "    144 (99.31%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "99.31% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "vcrs_9 (ENST00000373477:c.2126_2127insCGTAC)\n",
      "\n",
      "145 reads; of these:\n",
      "  145 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    145 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n",
      "1 reads; of these:\n",
      "  1 (100.00%) were unpaired; of these:\n",
      "    0 (0.00%) aligned 0 times\n",
      "    1 (100.00%) aligned exactly 1 time\n",
      "    0 (0.00%) aligned >1 times\n",
      "100.00% overall alignment rate\n",
      "[samopen] SAM header is present: 84 sequences.\n"
     ]
    }
   ],
   "source": [
    "for folder in sorted(glob.glob(aligned_reads_parent_dir + \"/*/\")):\n",
    "    variant = folder.split(\"/\")[-2]\n",
    "    variant_header = id_to_vcrs_header[variant]\n",
    "    print(f\"{variant} ({variant_header})\\n\")\n",
    "\n",
    "    outfolder = f\"{bowtie_read_alignments}/{variant}\"\n",
    "    os.makedirs(outfolder, exist_ok=True)\n",
    "    \n",
    "    # Align reads to human ref using bowtie2\n",
    "    variant_sam = variant + \"_reads.sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 3 \\\n",
    "        -x $bowtie_reference_prefix \\\n",
    "        -p $threads \\\n",
    "        -q $aligned_reads_parent_dir/$variant/1.fastq \\\n",
    "        -S $outfolder/$variant_sam\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \"_reads.bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder/$variant_sam \\\n",
    "        > $outfolder/$variant_bam\n",
    "    \n",
    "    # Sort bam file\n",
    "    variant_bam_sorted_prefix = \"final_\" + variant + \"_sorted_reads\"\n",
    "    variant_bam_sorted = variant_bam_sorted_prefix + \".bam\"\n",
    "    !$samtools sort \\\n",
    "        $outfolder/$variant_bam \\\n",
    "        $outfolder/$variant_bam_sorted_prefix\n",
    "    \n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder/$variant_bam_sorted\n",
    "\n",
    "\n",
    "    # repeat but use vcrs as the input instead of reads\n",
    "    variant_reference_sequence = vcrs_fasta_indexed[variant_header].seq.strip()\n",
    "\n",
    "    # Align vcrs to human ref using bowtie2\n",
    "    variant_sam = variant + \"_vcrs.sam\"\n",
    "    !$bowtie2 \\\n",
    "        --very-sensitive \\\n",
    "        -k 3 \\\n",
    "        -x $bowtie_reference_prefix \\\n",
    "        -p $threads \\\n",
    "        -c $variant_reference_sequence \\\n",
    "        -S $outfolder/$variant_sam\n",
    "    # use k 3 to get multiple alignments, not just the best one\n",
    "\n",
    "    # Convert sam to bam\n",
    "    variant_bam = variant + \"_vcrs.bam\"\n",
    "    !$samtools view \\\n",
    "        -bS -F4 $outfolder/$variant_sam \\\n",
    "        > $outfolder/$variant_bam\n",
    "\n",
    "    # Sort bam file\n",
    "    variant_bam_sorted_prefix = \"final_\" + variant + \"_sorted_vcrs\"\n",
    "    variant_bam_sorted = variant_bam_sorted_prefix + \".bam\"\n",
    "    !$samtools sort \\\n",
    "        $outfolder/$variant_bam \\\n",
    "        $outfolder/$variant_bam_sorted_prefix\n",
    "\n",
    "    # Create an index for the sorted bam file (creates a .bai file)\n",
    "    !$samtools index $outfolder/$variant_bam_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb02e32",
   "metadata": {},
   "source": [
    "### Load the files in bowtie_read_alignments into NCBI Genome workbench (or another genome viewer) to visualize the alignments (the ones with \"final_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "520bec19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000263574:c.750_755del is at the 24.76th percentile.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArMAAAHWCAYAAABkNgFvAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAX9xJREFUeJzt3QeYE1X79/Gb3qQtHWk+otjBhmLDgoAFUR8VsVBELI8FG/pHRVRUVBRRUbGC2EAULKhIsaCCBRCxoiBSlCq9l533+p28EybZXTbZzZId9vu5rmHJZDI5c2aS3HPmPmeKeZ7nGQAAABBCxdNdAAAAACCvCGYBAAAQWgSzAAAACC2CWQAAAIQWwSwAAABCi2AWAAAAoUUwCwAAgNAimAUAAEBoEcwCAAAgtAhmgTxo1KiRdenSJd3F2O3179/f/vOf/1iJEiWsWbNmOS6nfaF9gh3++usvK1asmD3yyCPpLkqhMHToUFcfqpfdeX9rOws7fVbPPPPMdBcDuxGCWRR5/o/c1KlTs33+xBNPtIMOOijf7/Phhx/a3Xffne/1FBXjxo2zW2+91Y499lgbMmSIPfDAA1YUabvfeeeddBcDSMovv/zivu9215MHFC4l010AIIxmzZplxYsXTzqYfeqppwhoE/TJJ5+4On7xxRetdOnSVlQpmD3vvPPs7LPPTndRgKSC2Xvuucc1BnDVBAWNllkgD8qUKWOlSpWyMFm/fr2FydKlS61cuXJFOpAFwva5BdKBYBZIQc7s1q1bXSvEPvvsY2XLlrVq1arZcccdZ+PHj3fPa1m1yopSGvwp+IN18803W/369V2g3KRJE5fr6HlezPtu3LjRrr/+eqtevbpVrFjRzjrrLPv777/duoItvvq/5ql15KKLLrKqVau68sjMmTNdeZSLqrLWrl3bLrvsMvv3339j3stfx++//26XXHKJVa5c2WrUqGG9e/d25VqwYIG1b9/eKlWq5Nbx6KOPJlR327Zts759+9ree+/ttlV1efvtt9vmzZujy+h9lVqgevHrKtlcwETrVOu+9tpr3aV8pZNo2QMPPNDGjh2bZZ2fffaZHXHEEa7eVP5nn302Wk/xXn31VTv88MNdQJ6RkWEXXnihq7OgP/74w/773/+6+tM669Wr55ZbvXp1tGzajpdffjlaD3nJ1X7uueei9X3kkUfad999l21L+PHHH28VKlSwKlWquH3766+/pvyY0ElKt27drFatWm6bmzZt6rYvp3zfxx57zBo2bOjqsWXLlvbTTz/lqezxOnfu7D5H+uzGa926tTte4o+RkSNH2gEHHODK0qJFC/vxxx/d8zoOGjdu7LZHLZHZXVr/5ptvrG3btq7Oypcv77blq6++yrZ+s/vc5tVvv/3mWvZ1DKp8On7fe++9bFOtVJ6bbrrJ7VPV5TnnnGPLli2LWTYzM9OVs27dum47TjrpJFfe4Hei1nf++ee7/+t5/9jV5yfoyy+/tObNm7ty6fto2LBh+dpWFGEeUMQNGTJE0Y03YcIEb9myZVmmY445xjvwwANjXtOwYUOvc+fO0ce33367V6xYMa979+7e888/7z366KNex44dvQcffNA9P3nyZO/UU0917/PKK69EJ8nMzPROPvlk9/rLL7/cGzRokNeuXTu37A033BDzvhdccIGbf+mll3pPPfWUe9y0aVM3r0+fPtHl9H/NO+CAA7z27dt7Tz/9tFteHnnkEe/444/37r33Xu+5557zevTo4ZUrV85r3ry5K0v8Opo1a+a2Res444wz3LwBAwZ4TZo08a6++mo3/9hjj3XzP//881zrW/WmZc877zxXpk6dOrnHZ599dnQZ1Y3KWKZMmWhdzZkzZ6fr1D7xJVOnmqc6rFOnjte3b19v4MCB3n/+8x+vfPny3vLly6PLTZ8+3ZWnUaNGbr/ef//9Xt26daP1H3Tfffe59+7QoYOrn3vuucerXr26e+3KlSvdMps3b/b22msvtw4t/8ILL7jljjzySO+vv/6K1oPeU3Xh14OOpUTMnTvXlevQQw/1Gjdu7D300EPeww8/7MpRr149b8uWLdFlx48f75UsWdLbd9993TJ+eatWrerWk6pjYsOGDd7+++/vlSpVyrvxxhu9J554wm2bllO9x5f94IMPdnWmsqtMGRkZXo0aNbzFixcnXXb/c+7P0+v0+P3334+pt0WLFnklSpRwn4/gMXLIIYd49evXd/teU+XKlb0GDRq4Y0ufM33m77zzTq906dLeSSedFLPOiRMnuvktWrRwyz322GNufZr3zTffJPS5TXR/azt9P/30kyun1qc6VFlPOOEEd2yOGjUqS93oWNHn5sknn/RuvvlmVw/6jgm69dZb3bL6PGl9+s7T8aQ6978T9Vm9/vrr3XL6bvSPXX+/6bOqY6VWrVruea3nsMMOc+VSmYFkEcyiyPO/yHc25RbMKqDRj/rOXHPNNVmCHnnnnXfcfAU0QQr29OU+e/Zs93jatGnZBmNdunTJMZhVwBFPAUW8N954wy0/adKkLOu44oorovO2bdvmfrhULj9QFwVoCoiDdZKdGTNmuHUqwAy65ZZb3PxPPvkkOk/rqlChwk7Xl1Mwm2idipZTUBGc98MPP7j5+lH36cdbAe7ff/8dnffHH3+4QCq4XxWIKghQsBv0448/umX9+d9//7173ciRI3e6baqD3Op1Z8FNtWrVvBUrVkTnv/vuu1mCOAWnNWvW9P7999+YOihevLg72UjVMaGAVa9/9dVXo/MUVCvI22OPPbw1a9bElF2vX7hwYXRZBX6ar0A42bLHB7Pbt2935dYJR5CCcm3Ln3/+GZ2n1+mkIhgcP/vss25+7dq1o+WWXr16xbyPTqz22Wcfr02bNjEni/oc6mRGJ7mJfG7zEsyecsop7oRg06ZN0Xkqg07QVab4umnVqlVMGVXPOpZXrVrlHisY1TEcPPGUu+++270+uK91XGvep59+mqWs+qzGf98sXbrU1bGCaCBZpBkA/5/SAJQWED8dcsghub5WlzZ//vlnd9k4WeoYpqGnlD4QpEvk+h396KOP3GP/svf//ve/mOWuu+66HNd91VVXZZmnS6S+TZs22fLly+3oo492j6dPn55l+csvvzz6f5VTlylVLl0qDm6/Lsv++eefuW6r6FJm/LbKBx98YKmQaJ36WrVq5S7D+7TPdanc357t27fbhAkTXCcsXV716dLyaaedFrOuUaNGuUuxF1xwgatbf9Jld6WhfPrpp245XW6Wjz/+2DZs2GAFpUOHDu5ytU+X48XftkWLFtmMGTPcJWJdig7WwamnnhrdZ6k4JrQu1UPHjh2j85R7rv20bt06+/zzz2PeR/W95557Rh/rkvRRRx0VLVNeyu5T58KLL77YXXJfu3ZtdP5rr71mxxxzjO21114xy59yyikxHZlUDlGaiFJ+4uf7263y6XtBaQNK5fGPB6WPaJ2TJk1yx0tun9tkrVixwqVf6DjU9vnvqzK0adPGlUkpSkFXXHFFTMqMjhUd+/PmzXOPJ06c6NKEkvkOyonSNfxjUZTakMh3CJCdIh3M6kukXbt27sdJH+C8DH+jL3Dlde27774uH01fvPfff3+BlBcFSz+UCmrip2AgkJN7773XVq1a5Y6Dgw8+2Hr27OlyUxOhHwodg8EfRNl///2jz/t/9QMc/yOrgCon8cv6P3I9evRwOYsKbPUj4i/n52oGNWjQIOaxgjDluCnfMH7+ypUrc91WbUN8mRXgKPjxtzW/Eq3TnLZRtN/97VGep/KVs6vr+HkKEvS9oMBVdRuclMepdYnqXEH9Cy+84OpSAYZOqLLbB/kRv23+8exvm18XwRzRYH35gVcqjgm9l+olfiSQnPaLlo2nz5ifk5qXsgd16tTJ7dfRo0dHRymZNm2aXXrppVmWzW6bRTnZ2c33t9s/wVWObvzxoH2vXPH4fZ7d5zZZs2fPdseh8pnj37dPnz5uGf9YTPZYiT/mdSKRyPfkzt7Lf7/cvkOA7BTpobn0JafOB+r8cu655+ZpHQoKNB6mAloFMQoUNKFoOeGEE2zOnDn27rvvuuNBP1LquDJ48OCYVqxdLdgK61NLzeTJk13ArRsR7LHHHq5lSJ1T4luI/Ja3ROZJfOeqnGTXYSqd8rs9QapDbZ9af7Nbr+rbpw5SalX0jxu1UPbr18++/vpr1xmssG3bztZZEO9T0NQ6qE566qynwFZ/NXqGPiOJbl9u2+1/pnQDkJxu/BE8JnL63CbLf99bbrnFnShlJz4o3ZX7MIzHCwqvIh3M6vJg/CXCIJ0x33HHHfbGG2+4Vjf1dH7ooYdcb1VRK8szzzzjetf6LQOpOKNGOKl1omvXrm7SJVMFuOr16wezOQVw6qmtS9i6FBhsSVQvZP95/69+oObOnRvTYqUWmESp1UOXCjXywl133RWdn5f0iLzwt0Hv57fGyZIlS9xnzN/WVLxPInWaqJo1a7qWx+zqOn6e0hX0g6zvArUi5kYnwZruvPNOd5Khm0ToJOi+++7bJYG/XxdqlYyn+lJrq3q2p+q9dMVCx0CwdTan/ZLdcamRFPzL/akou4JYtZArZeH111+3M844I+lWxp3x01eUtqIrPbuKRgfw0zhS9b5+feuYD/7WKXUhvkW1sJ2wYvdWpNMMcqOhWKZMmWLDhw93X8AaakStV/4X7Pvvv+++MMaMGeM+2PqCVeBCy2zREz+slVpa1OoRHG7K/1FV0BZ0+umnu7y0QYMGxcxXy65+EPwTLr915emnn45Z7sknn0y6NSS+9WPgwIG2K2hbs3u/AQMGuL8KJFL1PonUaTL1poBAqUj//PNPdL5+1OPzb3WVR8vrhCG+nvXYP1bWrFnj8g+DFNQqyIs/buKPmVSqU6eOazHU8FjB99FJulqL/X2WClrX4sWLbcSIEdF5qgMdw/rMaLiqINV3MK/z22+/dUNc+fsvFWVX/q6OCV1lU76mhhxLJbX8KqDV1Tud5MaLH/oqVXQCpoYXDRumQD0V76sc35IlS7pGnKD4z9nOvu+AglCkW2Z3Zv78+W6cS/31O3zoco064fi31tQXn3KINPagxsfTj+eNN97oxvRT4j2KDl2u1A+HfrjUQqtb47711lvuhMin50SXkhWYKuDRmKLK29ZYjLoKoFxApb7oh1iXnm+44YZoy45er84mCgQVEKnTljrMqKUq0ZYQtQ6pxfjhhx9242sqx1vvpdbeXUHbptxBjXuqHzkFLwpQFIyos4/qIRUSrdNkqJVd61DL6dVXXx0NlnXFRp18fFq3WlV79erl3lvbpdZh1bFyM9XJRt8l+o7Q8aGTZLXgKqh75ZVX3HGh/ezTflcrswJ+fRfpxNnvZJQqugSuAFFjp6oTl/JIFWAq/zOVd6zTtiu4UmqFclPVAKDPicY31XEdn+OsE0KNs6r6VoCvZTSGs25znKqyK4dUjRT6HlfedqpOqHw6OVHakcqo8Yt15UafOwXp6gyoz6QaRgqCcrBVfzpJ6t69u2t80VUQNdIsXLjQfvjhh6TWpzx7Bf1Kj9EY16o3rUMndGoFD34H6SRDx7KuZionWH1KTj75ZBdkAymX9PgHuylVxejRo6OPx4wZ4+ZpWJzgpGFJ/HH3NL6elpk1a1b0df7wSb/99ltatgPJ84el+e6777J9vmXLlrkOzaUhoDROa5UqVdxwQvvtt58bgik4lqeGMLruuuvcOJka+if48Vu7dq0bBkdjjmoMTg2b079//5hhcmT9+vVuiC+Nt6mhjDREjo4/rSs4LJI/xI/GyY2noY7OOeccV1aNQXn++ed7//zzT47De8WvI6chs7Krp+xs3brVjQWqYYm0rRq7U8MZBYcP2tn7JDI0VzJ1qm1UncaL38f+eKEai1NDee29995ubFgNJVS2bNksr3/77be94447LvrdoWNC7+N/X2jop8suu8ytR6/XPtX4pBrvOEjfJRobVMdV/PBHiQzVpG2OF7+vRe+rsWH1PpUqVXJDkf3yyy8xy6TimFiyZInXtWtXNy6p6lFDRwWHk4ovu8Zl1THij7erYbfiJVL2+KG5gt58880sQ47ldozkVL8aiiq7Idc0FNu5557rhkrTtuj40m+JjqlEPrd5GZrLH/NVQ5RpCDF9Dvbcc0/vzDPP9N56661cvwP9bQkOr6Xvsd69e7v1qb41Lu2vv/7qtuuqq66Keb3G3NaYzRreK7gebXt2QxnqeNEEJKuY/kl9iBw+OqNUq4l//3NdBtOwLRpuKT5RXZfD1PtaPULVQhu8g4xaBXRXFLXgaGgYoKCpVfDQQw91nVd0zGLX0ndGXodlQ/bUoq0WaLW6qhW7oKnFXvtRI9wEh4tCYnSVRXnGuiKhqyHArkaaQQ4UHOgyooYuyenLTZcbdWlQvdj9y5b+Jd9UdWQBgnSyFN/TWZdedSlT6QPYtfWvAFZjmSp1AuH1/PPPu0vw+b11bFH+DhK/czSwqxXpYFbJ+MGeyMppUyuXch6Vw6ZWLvV0VX6QglslzKsnuAbkVl6VOoQcdthhbmgvfZjVQ/eaa65xLbKJ9GIGkqVcV+UaKh9UHTGUq6ZJuYjx410i9RTwKN9Tf5Uvr44wGsopmMNZ0HSSnVvnHV09ih/uCVn5nXt1s47HH3+8UPbA37JlS66dipUfnIrhvBKhq5ZDhw51net0jH355ZduxJ/WrVu7Bh4gLbwizM8Hip/8nDTlO951113u3uDKNdK925VrOHPmzOg6dGtL5UEpf1H3mdatRYO3VQRSady4cS43UPed1zGpfEvdSlJ5qCh4+nwr3085j8rN1C1KlSe/K/m5kTub4vNhw2Zn+b6ppPfQd3e3bt0K7Wcop9+p4BSfJ1uQdLzrNrnKkdV3kG4J3KNHD5ejDqQLObMAECK6BbFaw3ZGLcf+OKMIN43fqqsxO6NREjRMGVBUEcwCAAAgtLhpAgAAAEKryHUAUyct3cFHg3MXxmR/AACAos7zPHdLct0sJnj76+wUuWBWgSy9vgEAAAq/BQsWWL169Xa6TJELZv3bJapydBtBJGm//cx0n291Nvjtt3SXBgAA7IbWrFnjGh/jb3OdnSIXzPqpBQpkCWbz4KSTzJYvN6teXZWY7tIAAIDdWCIpoUUumEU+vfZauksAAAAQxWgGAAAACC2CWQAAAIQWwSwAAABCi2AWyTn5ZN07MfIXAAAgzegAhuT8/rvZ33+brV6d7pIAAADQMgsAAIDwIpgFAABAaBHMAgAAILQIZgGgMGjUyGzgwMSXv/tus2bNCrJEABAKBLPIm5UrdY+5rFPbtjt+mPX4669jX3fDDWYnnrjj8YYNZr16me29t1nZsmY1api1bGn27rtmf/2V/XsEp6FDI+uZOdPs+OMj66hf3+zhh7OWeeRIs/32iyxz8MFmH34Y+7znmd11l1mdOmblypm1amX2xx+xy6xYYXbxxZFb+VapYtatm9m6dbHL5FaW55+PPF+1amTS+3z7bdby/vqr2VlnmVWubFahgtmRR5rNn7/j+SuvjNSbyqp6a9/e7LffYteRXZ0NH77j+S5dsl9GI1Zk58EHI89rP6aSArPsyqHt9mlfxz+vOk52H+bks89yPs6++y6yTE7HZPA4T6ScAICUIZhF3ilwXbQodnrjjR3P6wf8ttt2vo6rrjIbNcrsyScjgdjYsWbnnWf277+RQDC47ptvjgRZwXkdOpitWWPWurVZw4Zm06aZ9e8fCY6ee27H+0yebNaxYyT4/P57s7PPjkw//bRjGQWdTzxhNniw2TffRAKpNm3MNm3asYwC2Z9/Nhs/3mzMGLNJk8yuuGLH84mURUGTyvLpp2ZTpkS2U6/RKBG+OXPMjjsuEnxreQXIvXvHBkWHH242ZEgk6P3440ggp/Vs3x5bx1omWGfabt/jj8c+t2CBWUaG2fnnZ91XCuiefdbskEMs5W65JeuxdMABWcuhk4jgMvPmxT6fyD7MyTHHZC3D5Zeb7bWX2RFHxC47YULsctoXyZQTAJA6XhGzevVqT5utv8iDPfdUyOR55ct7Xvv2OS/XsKHnXX+955Uu7XkffLBjfo8entey5Y7HlSt73tChib13nz6e17Rp1vlPP+15Vat63ubNO+bddpvnNWmy4/EFF3jeGWfEvu6oozzvyisj/8/M9LzatT2vf/8dz69a5XllynjeG29EHv/yS2Tbv/tuxzIffeR5xYp53t9/J16WeNu2eV7Fip738ss75nXo4HmXXOIl5YcfIuWbPXvHPD0ePTrxdWhZbc9ff8XOX7vW8/bZx/PGj4/sP+3H3Hz5ZWTZcuU8r0oVz2vd2vNWrEisHDNmRMo+adKOeUOGRI6XnCSyD5OxZYvn1ajheffeu2Pe3LmRcn3/fc6vy62csmSJ5515pueVLet5jRp53quvRj4zjz22Y5mVKz2vWzfPq149cnycdFKkXnL7PABAEYvXaJlFwVGLllpelUaQmZn9MrVrRy73r12b9/dR6+YJJ5iVLr1jnlrjZs2KpEP4y+iSc5CW0XyZO9ds8eLYZXR5/6ijdiyjv0otCLbSafnixSOtgImWJZ5SLbZujbSIiurqgw/M9t038tqaNSPleOednOtg/fpIC6zqXC29QddcY1a9ulnz5mYvvRRpwc3Jiy9Gtkkty/HrOOOMrHUYTFcIpo/MmGF2yimR1lXVyZdfmrVrl7XVOCcvvBDZfqVjBCmlQ2XTNiqtQq3kvkT2YTLeey9yhaBr16zPKf1D+0Wt51ou3s7K6deXWsHVOv/WW2ZPP222dGnsMmqV1ryPPoq08h92WKROleoCAIjipglF2LJly2yNLosnoeLVV1vx9eutwtixVnbMGPP+f05jMZeeWczs9tsjk+/OOyNB1muvmV16adYV6vK7Lt1Xq2bWtGkkOFCawbHHJl4oBTAK4oJq1drxnPJS9defF1xG8/3lgq/LaRkFMEElS0aC0OAyuZUlnlIx6tbdEYQpgFEwpPzU++4ze+ihSPrFuedGgh/lFPsUBN16aySYbdIkkv4QDKTvvTdyt7by5c3GjTP73/8i677++qzl+OefSOD0+uux85VjO336jrzR7ChHNXjCosv9CvpVPl9OebjxlBKg4+X//i92vrZPwbjSHHTTjkceiaQGKFCsVy+xfZgMBfY6mdC6fXvsYfboo5HjUycxb78dSdvQiYYC3ETKqRuPqJ6VJ608aP+99t9/x/so+NfzOhbKlInM03r0Pgp+g6ktAFDEEcwW4UD2kq6X24q1G/L0+j7/LLYaFSvbg/vs5x5XqVDWHnnwAau2zz6xC6pjkvIh1SlH+a3x1Ir555+RDjTKa504MZLHec89kRzR3Z0CVgWLyov182H9oFAtejfeGPm/eq2rfpQLGgxmdSJw6qmRvEwFOxdcYPbVVzvWFazDQw+NBL3K480umH355UjLczCnVq2HPXpEguSddWLq1y/2sVpms8u7TcTo0ZGW+s6dY+e3aBGZfAoQFQAqj7dvX0uphQsjechvvhk7Xy3cN92047GCUZ0EqE79YDa3cirHWSdBwTxb5Uar7n0//BA56dBJXtDGjZF8agBAFMFsEaUWWQWyNVr81ypkxLVkJWCP9U9asY3rreR5t9j6FUvspylv26rq1a2af6k8SD/+aqELttIFlSoVuZysSa2Uao1Ui6L+H2xlzIlSFZYsiZ3nP9ZzO1sm+Lw/T62MwWX84Y+0TPyl4G3bIpd9c3uf4Hv4FHwqmFVnomCnKgVMCnZ0iT5IAZFa7IJ0GV2TTiKOPjrS8qtgUB3MsqNL7gqoNm/e0eInSj1Qa6Jaz4N1rsvb2mZd4vYpVUAd3wYNiqynRIms76ORBPJKKQZnnpm1hTW740YB+uzZie/DROlqggJJP0DdGdWpgv1Ey5kIBbLaBp3kxAsGvQAARjMo6hTIVqpZL+mpVNnyVqpMWff/XINhXZpVC+H99yeWG6sgTkFiIj3QRa1gCq6Ud+pTcKHLvf5lfS2jVt8gLeO3oCk1QMFQcBmlYCgX1l9Gf1etigR4vk8+ibSkKqBJtCz+ZXgFlUofiO8pr2BSLX7Ksw3S5en4XNagSNekSICZE7WYqhzBQFY+/zwSbGm0hyDlaP74Y+R1/qTyqkVY/88ukBUF5/H1nQjlvSqVIr4c2VFQrbL5gWsi+zARqkMFs506RQLR3KgegsFzbuVUK6yO7+BxpH2tY8unkwelRuikpnHj2EknOwCAKFpmkZTKq5Zb8cztVmrLZiu5batVWr3ciq9dadW2bLYSy5ZFWgmz+7FVjt9jj0XyMf3AT9RpSK2ICpDUEvbLL5Gc25NOigxvlIiLLoqkJSgAUmuuhttSqoLez6dL5bo8r3xHdWTSpf2pU3cMmeWPnapWYbVyKjBSAK5cVv+yu1pGNRxZ9+6Ry/0KWK+91uzCCyPLJVoW5cAq7UJ1ofF4/XxOBf2apGfPSFqG0jBUFwp6339/R0udUjNGjIgMxaVUDl0WVyuvWkRPPz2yjJZXq6RabJUioKD6gQciaR/xlLOp/XLQQbHzK1bMOk950tpXwfnq5KehxYYN2/FYY/kqR1edABWgK0hV6oGOD7XqqgU5PuBV67CCvtNOy1pGtdZrWxTQKfDTpX0NeaXhsxLdh4nQCYqCan+98akY2ha1tIqGlVOZ1ZqcaDl1YqPjSOMEP/NMJGBVuYOt2cqfVgCucuvER53hlM6gjoHnnJP1BAgAijCCWSTlznu7WMbKpba5dFkrs2WTPXbD/w+cRD+++qGOH7hf1MKllkgFe0HqYKMAQQGsevUr8NAlZgV7iVIArc5N6nGvPEQFS3p9sJOM8hYVPKpDmt5LwY460wQDMr8jlV6nIESd0RREBnNF1TFJAaxaLNUB6L//jYxrmkxZFMBs2RLp6BbUp09kTFpRwKKAWbmoym9VvaqzkcokKtMXX0TuGKVREnRJXoGv8mr9Tmqq86eeiuTdqrVRwdWAAZFgPEidlLRuBd15pZzd4A0dFHypHlTXGkVBgZqCZT/9YfnyrLmfauHWDQfU0z+7Fl9tp8rud6RT/Wp7g+kYiezD3Ciw1/GiFtTs6DhWcKogVMvopCK4LxMpp1p+FdzqBEv7TgF4ML9ZgblG+bjjjshoCjpRVKuz9nFu6RcAUMQU0/hcVsRyRStXrmyrV6+2Som2/O2G5syZYxdedpU1OuN/LlUgUf1vOtMFsyuq1rSeA8a4eWuWLrS/Pnjahr802PbWHakAAAB2UbxGziwAAABCi2AWQNGgfGE/Lzl+yi5HFwAQCuTMAiga1BFN4/BmJz9DiQEA0opgFkDRoDGQsxsHGQAQaqQZIM8qrFtlj13fxmqsiLuRQE7UO79du4IuFgAAKEIIZpFnZ74/xGYceoIty/j/Q0H5/v03cg96DS8UHAj+ssvMpk+PDCkFAACQAgSzyBvPs+O+eM++OCGb233qhgHB27P6NNi8xpkNjssKAACQDwSzyJNSWzfbtpKl7c+9D459QjcEUGtsdneZEqUZvPee2caNu6ScAABg90YHMCTlkZ5PWYnMbXbaB8Os4rpVsQHuH39EbuX5zTeR261mR7fh1H3ptYxuZQsAAJAPBLNIypI6Dd3fcpvW26oq1aPzS2VmWi3dNlX3oW/QIOdgtnz5yC1fdTtQAACAfCLNAHlSastm21qqTPTxNXNn2xbdyvaSS3J/scb03LChYAsIAACKBIJZ5Mm6ilWs/Po10cdHrlphe3z0kVnJkpHplFMiT1SvbtanT+yLV6wwq1FjF5cYAADsjkgzQFKOmjLWSm/Z5EYzqPvP3Oj8Ww84xJ68506rX79+ZMZ330WG4tIwXGqx9c2ZY7Zpk9mhh6ah9AAAYHdDMIuknDdykGWsXGqrK2VYhfWrXeus2mf/Llfetuy7747AdfnyyN/99zerUmXHChTc/uc/sQEuAABAHpFmgDzZXqKkzW+4nx357YTkXvjGG2bduxdUsQAAQBFDMIs8e/+sbtZqwggrlpmZ9UkNu+V5sa2yP/9sNmOG2dVX79JyAgCA3RdpBsizmU2Ps5pLFljGmhW2I3t2JxYtMhs2LDI0FwAAQAoQzCJfJrTuaGuWLkxs4VatCro4AACgiCHNAAAAAKFFMAsAAIDQIpgFAABAaBHMAgAAILToAIakrKmcEfMXAAAgnQhmkZS+fYaluwgAAABRpBkAAAAgtAhmAQAAEFoEswAAAAgtcmaRlEuH9rM91q+2dRUq2ytdeqW7OAAAoIgjmEVSDpn5lWWsXGorqtZMd1EAAABIMwAAAEB4EcwCAAAgtAhmAQAAEFoEswAAAAgtglkAAACEFsEsAAAAQotgFgAAAKFFMAsAAIDQ4qYJSMq3R7W28uvX2IYKldJdFAAAAIJZJGdkh+vTXQQAAIAo0gwAAAAQWgSzAAAACK20BrP9+vWzI4880ipWrGg1a9a0s88+22bNmpXr60aOHGn77beflS1b1g4++GD78MMPd0l5AQAAULikNZj9/PPP7ZprrrGvv/7axo8fb1u3brXWrVvb+vXrc3zN5MmTrWPHjtatWzf7/vvvXQCs6aefftqlZS+q7ut1vg26+iT3FwAAoEh3ABs7dmzM46FDh7oW2mnTptkJJ5yQ7Wsef/xxa9u2rfXs2dM97tu3rwuEBw0aZIMHD94l5S7KymzeaOU2rbeNmyukuygAAACFK2d29erV7m9GRkaOy0yZMsVatWoVM69NmzZufnY2b95sa9asiZkAAACweyg0wWxmZqbdcMMNduyxx9pBBx2U43KLFy+2WrVqxczTY83PKS+3cuXK0al+/fopLzsAAACKeDCr3FnlvQ4fPjyl6+3Vq5dr8fWnBQsWpHT9AAAAKOI3Tbj22mttzJgxNmnSJKtXr95Ol61du7YtWbIkZp4ea352ypQp4yYAAADsftLaMut5ngtkR48ebZ988onttddeub6mRYsWNnHixJh56gCm+QAAAChaSqY7teD111+3d99914016+e9Kre1XLly7v+dOnWyPffc0+W+So8ePaxly5b26KOP2hlnnOHSEqZOnWrPPfdcOjcFAAAARa1l9plnnnF5rCeeeKLVqVMnOo0YMSK6zPz5823RokXRx8ccc4wLgBW8Nm3a1N566y175513dtppDAAAALunkulOM8jNZ599lmXe+eef7yYAAAAUbYWiAxjC45VOt1nprZttSyk61QEAgPQjmEVSZjY7Pt1FAAAAKHzjzAIAAADJIpgFAABAaJFmgKQ0/OtXK7ltq20rWcrmNdo/3cUBAABFHMEsknLtEz0tY+VSW1G1pvUcMCbdxQEAAEUcaQYAAAAILYJZAAAAhBbBLAAAAEKLYBYAAAChRTALAACA0CKYBQAAQGgRzAIAACC0CGYBAAAQWgSzAAAACC3uAIak9L5/hJl5ZlYs3UUBAAAgmEVyNpWrkO4iAAAARJFmAAAAgNAimAUAAEBokWaApLT++DUru3G9SzcY1+bidBcHAAAUcQSzSMqpH79hGSuX2oqqNQlmAQBA2pFmAAAAgNAimAUAAEBoEcwCAAAgtAhmAQAAEFoEswAAAAgtglkAAACEFsEsAAAAQotgFgAAAKHFTROQlPkNm9jKjJq2tmLVdBcFAACAYBbJebLHo+kuAgAAQBRpBgAAAAgtglkAAACEFsEsAAAAQoucWSTlusdvtoprV7oOYOTPAgCAdCOYRVIazJtlGSuX2oqqNdNdFAAAANIMAAAAEF4EswAAAAgtglkAAACEFsEsAAAAQotgFgAAAKFFMAsAAIDQIpgFAABAaBHMAgAAILS4aQKSMr5NRyu7cb1tKlch3UUBAAAgmEVyxrW5ON1FAAAAiCLNAAAAAKFFMAsAAIDQIs0ASVG+rJlnZsXImwUAAGlHMIuk9L2jg2WsXGorqta0ngPGpLs4AACgiCPNAAAAAKFFMAsAAIDQIpgFAABAaBHMAgAAILQIZgEAABBaBLMAAAAILYJZAAAAhBbBLAAAAEKLYBYAAAChxR3AkJRB1/e3ktu22raSpdJdFAAAAIJZJGdeo/3TXQQAAIAo0gwAAAAQWgSzAAAACC3SDJCUQ2Z8YaW3brYtpcrYzGbHp7s4AACgiCOYRVIuHfaQZaxcaiuq1rSeBLMAACDNSDMAAABAaBHMAgAAILQIZgEAABBaaQ1mJ02aZO3atbO6detasWLF7J133tnp8p999plbLn5avHjxLiszAAAACo+0BrPr16+3pk2b2lNPPZXU62bNmmWLFi2KTjVr1iywMgIAAKDwSutoBqeddpqbkqXgtUqVKgVSJgAAAIRHKHNmmzVrZnXq1LFTTz3Vvvrqq50uu3nzZluzZk3MBAAAgN1DqIJZBbCDBw+2t99+203169e3E0880aZPn57ja/r162eVK1eOTnoNAAAAdg+humlCkyZN3OQ75phjbM6cOfbYY4/ZK6+8ku1revXqZTfddFP0sVpmCWjzbnOZcraxbAX3FwAAIN1CFcxmp3nz5vbll1/m+HyZMmXchNS4s9/IdBcBAAAgnGkG2ZkxY4ZLPwAAAEDRk9aW2XXr1tns2bOjj+fOneuC04yMDGvQoIFLEfj7779t2LBh7vmBAwfaXnvtZQceeKBt2rTJXnjhBfvkk09s3LhxadwKAAAAFMlgdurUqXbSSSdFH/u5rZ07d7ahQ4e6MWTnz58ffX7Lli128803uwC3fPnydsghh9iECRNi1gEAAICiI63BrEYi8Dwvx+cV0AbdeuutbkL6nD/iCSu/fo1tqFDJRna4Pt3FAQAARVzoc2axazX/Zpyd8MV77i8AAEC6EcwCAAAgtAhmAQAAEFoEswAAAAgtglkAAAAUrWD2zz//TH1JAAAAgF0RzDZu3NiN7frqq6+6mxcAAAAAoQlmp0+f7m5YoJsc1K5d26688kr79ttvU186AAAAINXBbLNmzezxxx+3f/75x1566SV3p67jjjvODjroIBswYIAtW7YsL6sFAAAAdl0HsJIlS9q5555rI0eOtIceeshmz55tt9xyi9WvX986derkglzsXmYecqxNPeJk9xcAACDUt7OdOnWqa5kdPny4VahQwQWy3bp1s4ULF9o999xj7du3J/1gN/NKl17pLgIAAED+glmlEgwZMsRmzZplp59+ug0bNsz9LV480tC711572dChQ61Ro0Z5WT0AAABQcMHsM888Y5dddpl16dLF6tSpk+0yNWvWtBdffDEvqwcAAAAKLpj9448/cl2mdOnS1rlz57ysHgAAACi4YFYpBnvssYedf/75MfPVEWzDhg0Esbux3vd0skqrV9iayhnWt8+wdBcHAAAUcXkazaBfv35WvXr1bFMLHnjggVSUC4WUAtmMlUvdXwAAgFAGs/Pnz3edvOI1bNjQPQcAAAAU2mBWLbAzZ87MMv+HH36watWqpaJcAAAAQMEEsx07drTrr7/ePv30U9u+fbubPvnkE+vRo4ddeOGFeVklAAAAsGs6gPXt29f++usvO+WUU9xdwCQzM9Pd9YucWQAAABTqYFbDbo0YMcIFtUotKFeunB188MEuZxYAAAAIxe1s9913XzcBAAAAoQlmlSOr29VOnDjRli5d6lIMgpQ/CwAAABTKYFYdvRTMnnHGGXbQQQdZsWLFUl8yAAAAoCCC2eHDh9ubb75pp59+el5ejhB76/xrrfSWTbaldNl0FwUAACDvHcAaN26c+tIgV8uWLbM1a9bkez3z5s2zbVu3Jf26b1q0zfd7AwAApDWYvfnmm+3xxx+3QYMGkWKwiwPZS7pebivWbsj3ujZt3GAL/15kDbZuTUnZAAAAQhPMfvnll+6GCR999JEdeOCBVqpUqZjnR40alaryIUAtsgpka7T4r1XIqJWvdS2d85PNW/CSbd9GMAsAAIpYMFulShU755xzUl8aJESBbKWa9fK1jnX/Ls7T62otmmclMrfZ9uIlbUkdxhUGAAAhDGaHDBmS+pIgFG7pf41lrFxqK6rWtJ4DxqS7OAAAoIgrntcXbtu2zSZMmGDPPvusrV271s37559/bN26daksHwAAAJDalln1hG/btq3Nnz/fNm/ebKeeeqpVrFjRHnroIfd48ODBeVktAAAAUPAts7ppwhFHHGErV660cuXKRecrj1Z3BQMAAAAKbcvsF198YZMnT3bjzQY1atTI/v7771SVDQAAAEh9y2xmZqZt3749y/yFCxe6dAMAAACg0LbMtm7d2gYOHGjPPfece6wbJ6jjV58+fbjFbRG1dcsWl0udCpUqVbIaNWqkZF0AAGD3lqdg9tFHH7U2bdrYAQccYJs2bbKLLrrI/vjjD6tevbq98cYbqS8lCrXN61bbX3P/tBtuv9vKlCmT7/VlVCxvrw55gYAWAAAUTDBbr149++GHH2z48OE2c+ZM1yrbrVs3u/jii2M6hKFo2Lp5o2UWK2nVjz7XqtXN340U1q9YYsumvO3udkYwCwAACiSYdS8sWdIuueSSvL4cu6HyVWvk+85ksiwlpQEAAEVBnoLZYcOG7fT5Tp065bU8KOTuu2uoFc/cbpnFS6S7KAAAAHkLZjXObNDWrVttw4YNbqiu8uXLE8zuxlZXqZ7uIgAAAORvaC7dLCE4KWd21qxZdtxxx9EBDAAAAIU7mM3OPvvsYw8++GCWVlsAAACg0HUAy3ZlJUvaP//8k8pVopA54bPRVnbzBttUprxNOvGcdBcHAAAUcXkKZt97772Yx57n2aJFi2zQoEF27LHHpqpsKITavfeiZaxcaiuq1iSYBQAA4Qxmzz777JjHugOYxgQ9+eST3Q0VAAAAgEIbzGZmZqa+JAAAAEC6OoABAAAAoWiZvemmmxJedsCAAXl5CwAAAKBggtnvv//eTbpZQpMmTdy833//3UqUKGGHHXZYTC4tAAAAUKiC2Xbt2lnFihXt5ZdftqpVq7p5unlC165d7fjjj7ebb7451eUEAAAAUpMzqxEL+vXrFw1kRf+/7777GM0AAAAAhTuYXbNmjS1btizLfM1bu3ZtKsoFAAAAFEyawTnnnONSCtQK27x5czfvm2++sZ49e9q5556bl1UiJJbUqm8by1WwNZUy0l0UAACAvAWzgwcPtltuucUuuugi1wnMrahkSevWrZv1798/1WVEIfLIbc+kuwgAAAD5C2bLly9vTz/9tAtc58yZ4+btvffeVqFChbysDgAAANj1N01YtGiRm/bZZx8XyHqel5/VAQAAAAUfzP777792yimn2L777munn366C2hFaQYMywUAAIBCHczeeOONVqpUKZs/f75LOfB16NDBxo4dm8ryoZDp/mxvu/GR69xfAACAUObMjhs3zj7++GOrV69ezHylG8ybNy9VZUMhtO+s7y1j5VJbUbVmuosCAACQt5bZ9evXx7TI+lasWGFlypRJRbkAAACAgglmdcvaYcOGRR8XK1bMMjMz7eGHH7aTTjopL6sEAAAAdk2agYJWdQCbOnWqbdmyxW699Vb7+eefXcvsV199lZdVAgAAALumZfaggw6y33//3Y477jhr3769SzvQnb++//57N94sAAAAUChbZnXHr7Zt27q7gN1xxx0FUyoAAACgIFpmNSTXzJkzk30ZAAAAUDjSDC655BJ78cUXU18aAAAAoKA7gG3bts1eeuklmzBhgh1++OHuVrZBAwYMyMtqAQAAgIJrmf3zzz/dEFw//fSTHXbYYVaxYkXXEUwdv/xpxowZCa9v0qRJ1q5dO6tbt64b3uudd97J9TWfffaZe2+NZ9u4cWMbOnRoMpuAfPrihPY2rnVH9xcAACBULbO6w9eiRYvs008/jd6+9oknnrBatWrl6c01CkLTpk3tsssuc6Mh5Gbu3Ll2xhln2FVXXWWvvfaaTZw40S6//HKrU6eOtWnTJk9lQHLeO7t7uosAAACQt2DW87yYxx999JELSPPqtNNOc1OiNILCXnvtZY8++qh7vP/++9uXX35pjz32GMEsAABAEZSnDmA5BbcFbcqUKdaqVauYeQpiNT8nmzdvtjVr1sRMAAAAKILBrPJaNcXP21UWL16cJaVBjxWgbty4MdvX9OvXzypXrhyd6tevv4tKCwAAgEKXZtClSxfX+Uo2bdrk8lfjRzMYNWqUFRa9evWym266KfpYgS8Bbd71v+lMy1i51FZUrWk9B4xJd3EAAEARl1Qw27lz5yzjze5KtWvXtiVLlsTM0+NKlSpZuXLlsn2NAm8/+AYAAEARDmaHDBli6dSiRQv78MMPY+aNHz/ezQcAAEDRk68OYPm1bt06Ny6tPzatht7S/+fPnx9NEejUqVN0eaU0aKzbW2+91X777Td7+umn7c0337Qbb7wxbdsAAACAIhrMTp061Q499FA3iXJb9f+77rrLPdaYtn5gKxqW64MPPnCtsRqfVkN0vfDCCwzLBQAAUETl6Xa2qXLiiSfudHiv7O7updfoTmMAAABAWltmAQAAgPwgmAUAAEBoEcwCAAAgtAhmAQAAEFpp7QCG8Hnhinus5NYttq1U6XQXBQAAgGAWyZm13+HpLgIAAEAUaQYAAAAILYJZAAAAhBZpBkhKk9+mRXNmSTkAAADpRjCLpFz+XB/LWLnUVlStaT0HjEl3cQAAQBFHmgEAAABCi2AWAAAAoUUwCwAAgNAimAUAAEBoEcwCAAAgtAhmAQAAEFoEswAAAAgtglkAAACEFsEsAAAAQos7gCEp3PULAAAUJrTMAgAAILQIZgEAABBaBLMAAAAILXJmkZSz3nneym1cZxvL7WHvnd093cUBAABFHMEsknL8pHctY+VSW1G1JsEsAABIO9IMAAAAEFoEswAAAAgtglkAAACEFsEsAAAAQotgFgAAAKFFMAsAAIDQIpgFAABAaBHMAgAAILS4aQKS8nuTQ22PtatsXcUq6S4KAAAAwSyS8/yVfdNdBAAAgCjSDAAAABBaBLMAAAAILYJZAAAAhBY5s0jKLQ9dbZXWrLA1lTLskdueSXdxAABAEUcwi6TUWrLAMlYutXIb16e7KAAAAKQZAAAAILwIZgEAABBaBLMAAAAILYJZAAAAhBbBLAAAAEKLYBYAAAChRTALAACA0CKYBQAAQGhx0wQk5f2zulnZzRtsU5ny6S4KAAAAwSySM+nEc9JdBAAAgCjSDAAAABBaBLMAAAAILdIMkJTKq5Zb8cztllm8hK2uUj3dxQEAAEUcwSyScue9XSxj5VJbUbWm9RwwJt3FAQAARRxpBgAAAAgtglkAAACEFsEsAAAAQotgFgAAAKFFMAsAAIDQIpgFAABAaBHMAgAAILQIZgEAABBaBLMAAAAILe4AhqQ80vMpK5G5zbYX59ABAADpR0SCpCyp0zDdRQAAAIgizQAAAAChRTALAACA0CLNAEk5aspYK71lk20pXda+adE23cUBAABFXKFomX3qqaesUaNGVrZsWTvqqKPs22+/zXHZoUOHWrFixWImvQ67xnkjB1mXoQ+4vwAAAFbUg9kRI0bYTTfdZH369LHp06db06ZNrU2bNrZ06dIcX1OpUiVbtGhRdJo3b94uLTMAAAAKh7QHswMGDLDu3btb165d7YADDrDBgwdb+fLl7aWXXsrxNWqNrV27dnSqVavWLi0zAAAACoe0BrNbtmyxadOmWatWrXYUqHhx93jKlCk5vm7dunXWsGFDq1+/vrVv395+/vnnHJfdvHmzrVmzJmYCAADA7iGtwezy5ctt+/btWVpW9Xjx4sXZvqZJkyau1fbdd9+1V1991TIzM+2YY46xhQsXZrt8v379rHLlytFJATAAAAB2D2lPM0hWixYtrFOnTtasWTNr2bKljRo1ymrUqGHPPvtstsv36tXLVq9eHZ0WLFiwy8sMAACA3XBorurVq1uJEiVsyZIlMfP1WLmwiShVqpQdeuihNnv27GyfL1OmjJsAAACw+0lry2zp0qXt8MMPt4kTJ0bnKW1Aj9UCmwilKfz4449Wp06dAiwpAAAACqO03zRBw3J17tzZjjjiCGvevLkNHDjQ1q9f70Y3EKUU7Lnnni73Ve699147+uijrXHjxrZq1Srr37+/G5rr8ssvT/OWAAAAoMgFsx06dLBly5bZXXfd5Tp9KRd27Nix0U5h8+fPdyMc+FauXOmG8tKyVatWdS27kydPdsN6oeCtqZwR8xcAAKBIB7Ny7bXXuik7n332Wczjxx57zE1Ij759hqW7CAAAAOEdzQAAAADwEcwCAAAgtAhmAQAAEFqFImcW4XHp0H62x/rVtq5CZXulS690FwcAABRxBLNIyiEzv7KMlUttRdWa6S4KAAAAaQYAAAAIL4JZAAAAhBbBLAAAAEKLYBYAAAChRTALAACA0CKYBQAAQGgRzAIAACC0CGYBAAAQWtw0AUn59qjWVn79GttQoVK6iwIAAEAwi+SM7HB9uosAAAAQRTCLQmfrli02b968fK+nUqVKVqNGjZSUCQAAFE4EsyhUNq9bbX/N/dNuuP1uK1OmTL7WlVGxvL065AUCWgAAdmMEsyhUtm7eaJnFSlr1o8+1anUb5nk961cssWVT3rY1a9YQzAIAsBsjmEVS7ut1vlVZtdxWValud/YbWWDvU75qDatUs16+1rEsZaUBAACFFUNzISllNm+0cpvWu78AAADpRjALAACA0CKYBQAAQGgRzAIAACC0CGYBAAAQWgSzAAAACC2CWQAAAIQWwSwAAABCi2AWAAAAocUdwJCUVzrdZqW3brYtpcqkuygAAAAEs0jOzGbHp7sIAAAAUaQZAAAAILQIZgEAABBapBkgKQ3/+tVKbttq20qWsnmN9k93cQAAQBFHMIukXPtET8tYudRWVK1pPQeMSXdxAABAEUeaAQAAAEKLYBYAAAChRTALAACA0CKYBQAAQGjRAQwAdhPbt2+3rVu3prsYAJCQ0qVLW/Hi+W9XJZgFgJDzPM8WL15sq1atSndRACBhCmT32msvF9TmB8EsAIScH8jWrFnTypcvb8WKFUt3kQBgpzIzM+2ff/6xRYsWWYMGDfL1vUUwCwAhTy3wA9lq1aqluzgAkLAaNWq4gHbbtm1WqlQpyys6gAFAiPk5smqRBYAw8dMLdFKeH7TMIim97x+hDD0z4zImUJiQWgCgqH5vEcwiKZvKVUh3EQAAAKJIMwAAAEBoEcwCAADsJho1amQDBw5MePm7777bmjVrlutyvXv3tiuuuCLh9W7ZssWVZerUqVbQCGaRlNYfv2ZnvfOc+wsA+dGlSxeXMxc/tW3bNrqMfgw17+uvv4557Q033GAnnnhi9PGGDRusV69etvfee1vZsmVdL+mWLVvau+++a3/99Ve27xOchg4d6tYzc+ZMO/7449066tevbw8//HCWco8cOdL2228/t8zBBx9sH374YZZxf++66y6rU6eOlStXzlq1amV//PFHzDIrVqywiy++2CpVqmRVqlSxbt262bp162KWya0szz//vHu+atWqbtL7fPvtt1nK++uvv9pZZ51llStXtgoVKtiRRx5p8+fPjz5/5ZVXunpTWVVv7du3t99++y1mHdnV2fDhw3PdlwceeKBl58EHH3TPaz+mkgKz7Mqh7fZpX8c/rzpOdh/m5LPPPsvxOPvuu+/cMjkdk8HjPJFy7srh/x5//HG74447Et6f6tx1yy232G233Vbg5SOYRVJO/fgNa//uC+4vAOSXAleNMxmc3ngj9vtFP+C5/SBeddVVNmrUKHvyySddIDZ27Fg777zz7N9//3WBYHD9N998swuygvM6dOhga9assdatW1vDhg1t2rRp1r9/fxccPffcc9H3mTx5snXs2NEFn99//72dffbZbvrpp5+iyyjofOKJJ2zw4MH2zTffuECqTZs2tmnTpugyCmR//vlnGz9+vI0ZM8YmTZoU0+qVSFkUNKksn376qU2ZMsVtp17z999/R5eZM2eOHXfccS741vIKkNXCFgyKDj/8cBsyZIgLej/++GMXyGk98T3MtUywzrTdPgU6wecWLFhgGRkZdv7552fZVwronn32WTvkkEMs1RQ8xR9PBxxwQJZy6CQiuMy8efNink9kH+bkmGOOyVKGyy+/3N0c4IgjjohZdsKECTHLaV8kU85d5YUXXnDbpeMxmf2p4/zLL790x3pBogMYACBtypQpY7Vr197pMgryFFSoBfT000/Pdpn33nvPBVT+82rRDQYGwffYY489rGTJklneVy1hujT60ksvuVYlBbwzZsywAQMGRANNvYcC8J49e7rHffv2dQHpoEGDXBkVCOoS75133ulaOGXYsGFWq1Yte+edd+zCCy90QaOCbQUBfnCjIFxlf+SRR6xu3br22muv5VoWLRMfcLz99ts2ceJE69Spk5unljStN9iqq1bY+Pr1qd7uu+8+a9q0qWs9DC6rFuSc9pVafTX5tK0rV660rl27xiyn1mcFOGpV1vsk4quvvnLboVZnHS/Nmzd3rcJqjY6nfavJ98MPP9gvv/zi9k2QWhFz2pZE9uHOaH8F163h83SF4LrrrsvSe19jQ+/s+N9ZOWXp0qXuxEpBsZbLrk41DrWCfJVh8+bN7ph77LHH3D5OlOr76quvzjI/t/2pfXTssce61+uzUlBomQWA3dWAAWb16uU+nXVW1tdqXiKv1XsUMLVoqeVVaQS6a1B29EOuYHft2rV5fh+1bp5wwgkxt9ZUa9ysWbNcYOYvo0vOQVpG82Xu3LnukmxwGQV5Rx11VHQZ/VVgGGyl0/K6tadaARMtSzylWihwUouoqK4++OAD23fffd1rdWMNlUMBWU7Wr1/vWmBV52rpDbrmmmusevXqLphUkK2gLycvvvii26b4ljyt44wzzshSh8F0hWD6iAL4U045xbWuqk7UyteuXbuExyVVgK/tVzpGfBCmsmkbFbAGWw4T2YfJ0ImWrhDEB/ai9A/tF7Wea7l4OyunX18LFixwrfNvvfWWPf300y7ADVKrtOZ99NFHrpX/sMMOc3WqVJdEaDmdEMS3KieyP0XHyxdffGEFiZZZANhdrVljFrjknKO4oMVZtiyx1+o98kGX2IMtaXL77be7KUitZAqy1Bp56aWXZlmPLr+rhUgtXWpxUnCgNAO1CiVKAYyCuCC1xvnPqZVJf/15wWU0318u+LqcllEAE6SWYgWhwWVyK0s8pWKoVdcPLBTAKBhSPqNazR566CHXInzuuee64Ec5xT4FQbfeeqsLZps0aeJam4OB9L333msnn3yyuznHuHHj7H//+59b9/XXX5+lHLqjkwKn119/PWa+WuemT58ezRvNjnJUgycsalFWEKXy+XLKw42nlAAdL//3f/8XM1/bp2Bcl8VXr17tWsN1CV2BYr169RLah8lQYK+TCa3bp2P+0UcfdcenTmLUoq60DZ1oKMBNpJy///67q+dvv/3W5UH777X//vtH30fBv57XsaBWbdF69D4KfhPp0KX8ap246NhKdn+KXlfQ6REEswCwu6pUyWzPPXNfrkaN7Ocl8lq9Rz6cdNJJ9swzz8TM81sWY4tTw10qVacc5bfGUyvmn3/+6TrQKK9Vl9qVEnDPPfe4HNHdnQJWBRfKi/XzYf2gUC16N954o/u/eq2rfnTZPRjM6kTg1FNPdXmZCnYuuOACd3nfX1ewDg899FAX9CqPN7tg9uWXX3Ytz8GcWrUe9ujRwwXJO+vE1K9fv5jHapnNLu82EaNHj3Yt9Z07d46Z36JFCzf5FCAqAFTeZ6ovhS9cuNDlIb/55psx89XCfdNNN0UfKxjVSYDq1A9mcyun0lV0EnR4IJ1GudGq+2CahU464m91vXHjRpdPnQgtK8H9luj+FHWg01WDgkQwCwC7K/1YBn4wk5LNJc+CoI41jRs3TmhZ/firhS7YSheke7vrcrImtVKqNVItivp/sJUxJ0pVWLJkScw8/7Gft5jTMsHn/XlqZQwu4w9/pGXiLwXr3vS6nJvb+wTfw6fgU8Gs8iaDnXAUMCnY0SX6IAVEarHLLud1n332saOPPtq1/CoYVAez7OiSuwIq5WD6LX6iFjy1Jqr1PFjnurytbdYlbp9SBdTxTfnGWk+JEiWyDYTySikGZ555ZpYW1uyOGwXos2fPTngfJkpXExRI+gHqzqhOFRwmWs5ErFu3zm2DTnLiBYPendFxJEpv0UllsvtTx7X/uoJCziwAIBR0aVYthPfff39CubEK4hQkJtIDXdQKph9j5Z36FFzocq9/WV/LqNU3SMv4LWhKDVAwFFxGIxMoF9ZfRn/VKUcBge+TTz5xLakKaBIti38ZXkGl0gficxoVTKrFT3m2Qbo8nV2v9GBAqkkBSU7UYqpyBANZ+fzzz12wpU5JQcrR/PHHH93r/EnlVYuw/p9dICsKzuPrOxHKe1UqRXw5sqMgTGXzA9dE9mEiVIcKZtUZT4FoblQPweA5t3KqFVbH97TAcaR9rWPLp2BTqRE6qdFJY3Dyg9TcqBOgRlVQ3mxe9qdG+lAQXpBomQUApI0Cpvg8RP3w5vRDqxw/9cRWPqYf+Ik6DakVUT+oagnTD6/ybpXGoB/iRFx00UUuLUEBkFpz9SOsVAW9n0+XVnV5XvmO6viiS/saFN4fMssfa1OtwmrlVGCkAFx5g/5ld7WMakSE7t27u8v9ClivvfZa10vez0tMpCzKgVXahepCoxD49Rjs0a9RF5SWoTQM1YWC3vfffz/aUqfUjBEjRrihuNR6psviauVVi6g/MoSWV6ukWmx1SVlB9QMPPODSPuIpZ1P75aCDDoqZX7FixSzz1CqvfRWcr05+GlpMowf4jzWWr3J01QlQAbqCVKUe6BhRK6BakOMDXrUOK+g77bTTspRRrfXaFgV0Cvx0aV85nRo+K9F9mAidoCio9tcbn4qhbfGDPA0rpzKrNTnRcurERsfRlVde6VJ19LlRuYOt2cqfVgCucuvER53hlM6gjoHnnHNOtp264imnV+tRa76//YnuT1Hnr4IcycCVsUDXDgDATii4UtARnNR5Kydq4dIPY3xrqzrYKEBQUKZgUcMgaV58ruLO6DK7OjcpAFEeosajVbAY7CSjvEUFjwpe1dFMnWjUmSb4A66OVHp/vU4to7rUq+0M5haqY5Ja1tTCpaBR2xwcQzaRsiiA0fBd6ugWrD+lHfgUsChgViCjoNAfvsuvY5VJwYbKoKBJga8CFeXV+p3UVOdPPfWUC4p0mV05mxoirE+fPjH1p05KWnciraE5Uc5u8IYOCr5UD8r9VK94lUFDTClwk+XLl2fJ/VQLt4ZZU0//7Fp8dblcJxI6TrTdanXV9gbTMRLZh7lRYK/jRfs5OzqOtW8V/GubdFIRHPEgkXKq5bdu3bruBEsd+1TeYOdCBeYa5UMnM1q36lMnTQqKc0u/CFIArRO3nEYTyYlGf9BxoWO0IBXzdja2xm5IB4O+JFS5iZ6tFxb6wF542VXW6Iz/WaWaO3pF5sU/v061r4Y9bCdcdZ/Varhvwq+77vGbreLalba2YlV7ssej+VpXKssVb83ShfbH6IE24P4+O72cligdKwWd8wPkhYI6BTxqPUrX3YEA7N48z3NBtzoS5pRHnR2dHOmkL350kkS+v5KJ10gzQFL8ALaw27xutf0190+74fa7s+R05UVGxfL26pAXCGgBAEVOsWLF3JUD5ckmSlcNdDXAH0mjIBHMYre0dfNGyyxW0qoffa5Vq5u/ltn1K5bYsilvu7NEglkAKNqUL6wpOxpJQ2O/7o6aNWuW1GgOygnW+NC7AsEsdmvlq9bId0qGLEtJaQAAYaeOaBqHNzv5GUoMeUcwCwAAkCDd1CO7G3sgfQhmkZTsOoABSL8i1pcXwG7AS9H3FsEsktJg3izLWLnUVlSNva84gPTwB2PX7SK5xAkgTNRJTHK6aUaiCGYBIMT0I6DbUvq3Ry1fvrzreQwAhZnGrF22bJn7zvLHDc4rgtldQDtLPeHzS4Mcb9u6LSVlArD78O8l7we0ABAGurtYgwYN8n0CTjC7CwLZS7pebivWbsj3ujZt3GAL/15kDQL36gYA/RDozk+6849ujQoAYaDhuxTQ5hfBbAFTi6wC2Rot/msVMhK/dVx2ls75yeYteMm2b+PHalfbumWLaxlPBe4mhoJMOchv7hkAhE2hCGZ1z+f+/fvb4sWL3W3PnnzySXcP5pyMHDnSevfubX/99Zfts88+9tBDD7n7FhdmCmTzO97pun8Xp6w8SBx3EwMAoPBKezA7YsQIu+mmm2zw4MHuvr8DBw60Nm3a2KxZs9wls3iTJ0929wXu16+fnXnmmfb666/b2WefbdOnT7eDDjooLduA3Rt3EwMAoPBKezA7YMAA6969u3Xt2tU9VlD7wQcf2EsvvWT/93//l2X5xx9/3Nq2bWs9e/Z0j/v27Wvjx4+3QYMGudcChf1uYv+kMGVBw5oo56gwrYs0CgBAkQlm9eM5bdo069WrV3SeEoFbtWplU6ZMyfY1mq+W3CC15L7zzjvZLr9582Y3+VavXu3+pmJ0gUSsXbvWtm/bZqsW/WVbN+WvE9iapQvNy8y0NYsXWMl8jryT13Wt2bbVHTT6++/83wtNuQpqPale18p//rS5c2bbdbfeme+UBeXx/rNwge1Zv6GVLFWy0KxrjzIl7e47e3GHHADYzVSpUmWXfbf7cVpCN1bw0ujvv/9WCb3JkyfHzO/Zs6fXvHnzbF9TqlQp7/XXX4+Z99RTT3k1a9bMdvk+ffq492BiYmJiYmJiYrJQTQsWLMg1nkx7mkFBU6tvsCVXg/SuWLHCqlWrluu4ZjorqF+/vi1YsMBdOkVqUK+pR50WDOo19ajT1KNOCwb1mt46VYusrm7XrVs31/WmNZitXr26G0ZmyZIlMfP12B8EPJ7mJ7O8LuXGX85VM3kyVOEcyKlHvaYedVowqNfUo05TjzotGNRr+uq0cuXKCa0v/yPV5oM6mxx++OE2ceLEmJZTPW7RokW2r9H84PKiDmA5LQ8AAIDdV9rTDJQC0LlzZzviiCPc2LIammv9+vXR0Q06depke+65pxuKS3r06GEtW7a0Rx991M444wwbPny4TZ061Z577rk0bwkAAACKXDDboUMHd8vXu+66y900oVmzZjZ27FirVStyt6z58+fH3OrsmGOOcWPL3nnnnXb77be7myZoJIOCGGNW6Ql9+vRJyUD52IF6TT3qtGBQr6lHnaYedVowqNfw1Gkx9QJL6RoBAACAXSStObMAAABAfhDMAgAAILQIZgEAABBaBLMAAAAILYLZnXjqqaesUaNGVrZsWTvqqKPs22+/TXeRQuvuu+92d1wLTvvtt1+6ixU6kyZNsnbt2rk7oqgONZJHkPpzamSQOnXqWLly5axVq1b2xx9/pK28u0OddunSJcux27Zt27SVNww0lOKRRx5pFStWtJo1a9rZZ59ts2bNillm06ZNds0117i7Me6xxx723//+N8sNcZB8vZ544olZjterrroqbWUu7J555hk75JBDooP4a8z6jz76KPo8x2nq67QgjlGC2RyMGDHCjYGrISSmT59uTZs2tTZt2tjSpUvTXbTQOvDAA23RokXR6csvv0x3kUJHYzDrWNSJVnYefvhhe+KJJ2zw4MH2zTffWIUKFdxxqy9k5K1ORcFr8Nh94403dmkZw+bzzz93AcDXX3/tbmqzdetWa926tatr34033mjvv/++jRw50i3/zz//2LnnnpvWcu8O9Srdu3ePOV71vYDs1atXzx588EGbNm2aG7P+5JNPtvbt29vPP//snuc4TX2dFsgxqqG5kFXz5s29a665Jvp4+/btXt26db1+/fqltVxh1adPH69p06bpLsZuRR/f0aNHRx9nZmZ6tWvX9vr37x+dt2rVKq9MmTLeG2+8kaZShrtOpXPnzl779u3TVqbdwdKlS13dfv7559HjslSpUt7IkSOjy/z6669umSlTpqSxpOGuV2nZsqXXo0ePtJYr7KpWreq98MILHKcFUKcFdYzSMpuNLVu2uDMKXaL16cYNejxlypS0li3MdLlbl3L/85//2MUXX+xuiIHUmTt3rrvxSPC41X2tlSLDcZs/n332mbus26RJE7v66qvt33//TXeRQmX16tXub0ZGhvur71e1KgaPVaUdNWjQgGM1H/Xqe+2116x69eruZkK9evWyDRs2pKmE4bJ9+3Z3V1G1dOvSOMdp6uu0oI7RtN8BrDBavny52wH+Xch8evzbb7+lrVxhpoBq6NChLhjQJYV77rnHjj/+ePvpp59c/hfyT4GsZHfc+s8heUox0GXFvfbay+bMmePuPHjaaae5H7MSJUqku3iFXmZmpt1www127LHHRu/UqOOxdOnSVqVKlZhlOVbzV69y0UUXWcOGDV3DwcyZM+22225zebWjRo1Ka3kLsx9//NEFWkrHUl7s6NGj7YADDrAZM2ZwnKa4TgvqGCWYxS6hH3+fEsMV3OpgfvPNN61bt25pLRuwMxdeeGH0/wcffLA7fvfee2/XWnvKKaektWxhoBxPnbSSI79r6vWKK66IOV7VGVTHqU7EdNwiKzWyKHBVS/dbb71lnTt3dvmxSH2dKqAtiGOUNINsqOlbLS7xPRb1uHbt2mkr1+5EZ7r77ruvzZ49O91F2W34xybHbcFSmoy+Izh2c3fttdfamDFj7NNPP3WdQnw6HpXOtWrVqpjlOVbzV6/ZUcOBcLzmTK2vjRs3tsMPP9yNGKEOoY8//jjHaQHUaUEdowSzOewE7YCJEyfGXNLR42DOB/Ju3bp17ixMZ2RIDV0G1xds8Lhds2aNG9WA4zZ1Fi5c6HJmOXZzpr50Crh0afGTTz5xx2aQvl9LlSoVc6zqMqPy6DlW816v2VHrmHC8Jk6/95s3b+Y4LYA6LahjlDSDHGhYLjWLH3HEEda8eXMbOHCgS2Du2rVruosWSrfccosby1OpBRraREOeqfW7Y8eO6S5a6E4Cgmev6vSlLwJ1AFGnBOXQ3XfffbbPPvu4H7revXu7vCSNR4nk61ST8rs1tqROFHQCduutt7oWBw15hpwvgb/++uv27rvvupx4P79QHRI1/rH+Kr1I37OqY41Fed1117kA4eijj0538UNbrzo+9fzpp5/uxkVVPqKGljrhhBNcegyyUucjpcHp+3Pt2rWu/pRC9PHHH3OcFkCdFtgxmtKxEXYzTz75pNegQQOvdOnSbqiur7/+Ot1FCq0OHTp4derUcXW55557usezZ89Od7FC59NPP3XDwsRPGj7KH56rd+/eXq1atdyQXKeccoo3a9asdBc7tHW6YcMGr3Xr1l6NGjXcED0NGzb0unfv7i1evDjdxS7UsqtPTUOGDIkus3HjRu9///ufG7KnfPny3jnnnOMtWrQoreUOe73Onz/fO+GEE7yMjAz3+W/cuLHXs2dPb/Xq1ekueqF12WWXuc+1fpv0Odd35rhx46LPc5ymtk4L6hgtpn/yHgoDAAAA6UPOLAAAAEKLYBYAAAChRTALAACA0CKYBQAAQGgRzAIAACC0CGYBAAAQWgSzAAAACC2CWQAAAIQWwSyApJ144onu1rm7sy5duuz2twEuVqyYvfPOOxY2f/31lyu7f0/3dCsKxwpQmBHMAtjtFbbgB+HEcQQUTgSzAIBCZcuWLekuAoAQIZgFkC8rV660Tp06WdWqVa18+fJ22mmn2R9//BF9fujQoValShX7+OOPbf/997c99tjD2rZta4sWLYous23bNrv++uvdctWqVbPbbrvNOnfuHHPpNjMz0/r162d77bWXlStXzpo2bWpvvfVWTDkuvvhiq1Gjhnt+n332sSFDhrjn9Bo59NBDXcua0iQS9cgjj1idOnVcua655hrbunVr0ts+ZswYa9KkiVvmvPPOsw0bNtjLL79sjRo1cq/Vtm/fvj3p9SpFQNtZtmxZa9OmjS1YsCCm7M8884ztvffeVrp0aff+r7zySo7befLJJ9u1114bM2/ZsmXutRMnTnSPVd777rvPlU37sWHDhvbee++55dq3b+/mHXLIITZ16tSY9Xz55Zd2/PHHu/1Sv359t73r16+PPq/19u3b1623UqVKdsUVV1iyfvrpJ1dPKkOtWrXs0ksvteXLl0ef1z7X+956662WkZFhtWvXtrvvvjtmHb/99psdd9xxrj4POOAAmzBhQkwqRm7H0c6OFQAFyAOAJLVs2dLr0aOH+/9ZZ53l7b///t6kSZO8GTNmeG3atPEaN27sbdmyxT0/ZMgQr1SpUl6rVq287777zps2bZpb/qKLLoqu77777vMyMjK8UaNGeb/++qt31VVXeZUqVfLat28fs8x+++3njR071pszZ45bb5kyZbzPPvvMPX/NNdd4zZo1c+8xd+5cb/z48d57773nnvv22289fd1NmDDBW7Rokffvv//muo2dO3d2ZVBZVKb333/fK1++vPfcc89Fl0l020899VRv+vTp3ueff+5Vq1bNa926tXfBBRd4P//8s1tv6dKlveHDhye93iOOOMKbPHmyN3XqVK958+beMcccE12H6lLLPPXUU96sWbO8Rx991CtRooT3ySefRJdRnYwePdr9/7XXXvOqVq3qbdq0Kfr8gAEDvEaNGnmZmZnuccOGDd1+Gjx4sPf77797V199taujtm3bem+++aZ7n7PPPtuV3X/N7NmzvQoVKniPPfaYe81XX33lHXrooV6XLl2i76P1aj2PPPKIW17Tzmj/quzff/+9e7xy5UqvRo0aXq9evdy+Ul2rzk866aSYY1bvcffdd7tyvPzyy16xYsW8cePGuee3bdvmNWnSxL1Odf7FF1+4Og3WUU7HUSLHCoCCQzALIM/BrIIC/bgrQPEtX77cK1eunAtu/MBLywQDFAVYtWrVij7W//v37x99rMCiQYMG0WBWAZaCAwVuQd26dfM6duzo/t+uXTuva9euCQU/iVCAoiBLZfGdf/75XocOHdz/87rtV155pduWtWvXRucpWNX8ZNf79ddfR5dREKV533zzjXuswLZ79+4x26Tyn3766dHHwUBt48aNLpgdMWJE9PlDDjnEBX8+1ccll1wSfayATuvo3bt3dN6UKVPcPD3n76MrrrgiphwKFIsXL+7e01+vguBExe/Pvn37uhOEoAULFrhlFGD7x+xxxx0Xs8yRRx7p3Xbbbe7/H330kVeyZMlouUUnRME6yuk4yu1YAVCwSDMAkGe//vqrlSxZ0o466qjoPF1i1SVtPefTpXJd7vbpUuzSpUvd/1evXm1Lliyx5s2bR58vUaKEHX744dHHs2fPdpfmTz31VHcZ2Z+GDRtmc+bMcctcffXVNnz4cGvWrJm7lDx58uR8b9+BBx7oypJdufO67boErsvqKn9wXrLr1TJHHnlk9PF+++3nUg/8ZfT32GOPjdkePQ6uI0iX1nVp/qWXXnKPp0+f7i7dq6d+kNIIguWWgw8+OMs8f3t++OEHlxYR3G9KiVDayNy5c6OvO+KIIyyv9B6ffvppzHuoPsQ/PuLLHr8/Z82a5VIglH7gCx6T+TlWABSskgW8fgCwUqVKxTxWvmGkYTAx69atc38/+OAD23PPPWOeK1OmjPurfMl58+bZhx9+aOPHj7dTTjnF5S0qjzGV5VYQlt91pGK9BeHyyy93JwMLFy50+cbKo1VebFCw7Cp3TvP87dG+u/LKK12+arwGDRpE/1+hQoU8l1vv0a5dO3vooYeyPKegMruyp7reC+s+BYoCWmYB5Jk6dKnz1jfffBOd9++//7pWLnWgSUTlypVda953330XnafOUGoZ9GldClrnz59vjRs3jpnUmuZT5y91HHv11Vdt4MCB9txzz7n56sTkr7cwbXt+1qtlgh2t9PyqVavc6/31fPXVVzHr1uOdlU0trGohff755+3111+3yy67zPLrsMMOs19++SXLftPk75dUvMfPP//sWrzj3yPRIFkt3+pAp6sEvuAxWVDHEYD8I5gFkGfqSa9e7N27d3c91nW595JLLnGtp5qfqOuuu86NVPDuu++6oKxHjx6uR7/fylexYkW75ZZb7MYbb3SjAOjSsYLdJ5980j2Wu+66y71eKQkKbDSCgB/Y1axZ0/WkHzt2rAtWlNpQWLY9r+tVS6DqTUHvtGnTXDrA0UcfHb003rNnT3d5XyMaaCSEAQMG2KhRo1w95tY6++CDD7qW83POOcfySyNTKOVDIyVofFaVRfspfuSE/FAL/IoVK6xjx44uANXxodEzunbtmnDgqRQWpYPoZGjmzJku8L/zzjvdc/5xWBDHEYD8I5gFkC+6HK381jPPPNNatGjhgiBd6o+/7JpbwKNAREMzaR1+XqXyOH0auql3794u6FWQquG9lHbgD5ekVrNevXq5vMgTTjjB5S8qh9bPL33iiSfs2Weftbp16+Yr2Ez1tud1vcrFVb1ddNFFLhdWdTZixIjo8xrW7PHHH3dpFsrn1LZrvbkNS6b9oPrS32D955X2x+eff26///67G55Lw1rpxEP7IVW0LgWfClxbt27tWph1hzrlEBcvntjPnI4XDcGllAXlIiuov+OOO9xzfj0U1HEEIH+KqRdYPtcBACmlXEMFrBdccIELYhFLLa4K1pRWUBB3uVILpVo4dfm+KFOArHFn1dof7MQHoHChAxiAtFPHrXHjxlnLli1t8+bNNmjQINfTXa2O2DU0wL9yc3VpXekKRTGQHT16tGvhVqqHAlilu6jVm0AWKNxIMwCQdroUrNZGXd5V8PDjjz+6uy/5Oa8FITiMU/z0xRdfWFFshVTPf7XIDh48OK1leeCBB3LcNxq1oqCsXbvW5d9qWC/lIOt4VH4vgMKNNAMARZJa3nKizlbq6IP0UGcuTdnRfokfng1A0UYwCwAAgNAizQAAAAChRTALAACA0CKYBQAAQGgRzAIAACC0CGYBAAAQWgSzAAAACC2CWQAAAFhY/T+LnNEUarbzygAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from varseek.utils import longest_homopolymer\n",
    "\n",
    "# Load the CSV and select relevant columns\n",
    "file_path = \"/Users/joeyrich/Desktop/local/RLSRWP_2025/data/vk_ref_out/variants_updated_filtered.csv\"\n",
    "df = pd.read_csv(file_path, usecols=[\"vcrs_header\", \"vcrs_sequence\"])\n",
    "\n",
    "(\n",
    "df[\"longest_homopolymer_length\"],\n",
    "    df[\"longest_homopolymer\"],\n",
    ") = zip(*df[\"vcrs_sequence\"].apply(lambda x: (longest_homopolymer(x) if pd.notna(x) else (np.nan, np.nan))))\n",
    "\n",
    "# Define the target variant\n",
    "target_header = \"ENST00000263574:c.750_755del\"\n",
    "\n",
    "# Ensure longest_homopolymer_length is numeric\n",
    "df = df.dropna(subset=[\"longest_homopolymer_length\"])\n",
    "df[\"longest_homopolymer_length\"] = pd.to_numeric(df[\"longest_homopolymer_length\"], errors=\"coerce\")\n",
    "\n",
    "# Get the target value\n",
    "target_value = df.loc[df[\"vcrs_header\"] == target_header, \"longest_homopolymer_length\"]\n",
    "\n",
    "if target_value.empty:\n",
    "    print(f\"Header {target_header} not found in dataset.\")\n",
    "else:\n",
    "    target_value = target_value.iloc[0]\n",
    "\n",
    "    # Compute percentile\n",
    "    percentile = (df[\"longest_homopolymer_length\"] < target_value).mean() * 100\n",
    "    print(f\"{target_header} is at the {percentile:.2f}th percentile.\")\n",
    "\n",
    "    # Plot histogram\n",
    "    plt.figure(figsize=(8, 5))\n",
    "    plt.hist(df[\"longest_homopolymer_length\"], bins=30, alpha=0.7, edgecolor=\"black\")\n",
    "    plt.axvline(target_value, color=\"red\", linestyle=\"dashed\", linewidth=2, label=f\"{target_header} ({target_value})\")\n",
    "    \n",
    "    # Annotate the target point\n",
    "    plt.text(target_value, plt.ylim()[1] * 0.9, f\"{target_header}\\n({target_value})\", color=\"red\")\n",
    "\n",
    "    # Labels and title\n",
    "    plt.xlabel(\"longest_homopolymer_length\")\n",
    "    plt.ylabel(\"Frequency\")\n",
    "    plt.title(\"Histogram of longest_homopolymer_length\")\n",
    "    plt.legend()\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "52273b41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Value 8 is at the 99.6286th percentile.\n"
     ]
    }
   ],
   "source": [
    "from scipy.stats import percentileofscore\n",
    "\n",
    "value = 8\n",
    "percentile = percentileofscore(df[\"longest_homopolymer_length\"].dropna(), value, kind=\"rank\")\n",
    "\n",
    "print(f\"Value {value} is at the {percentile:.4f}th percentile.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RLSRWP_2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
