{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['_missing', 'abdomen, nos', 'adrenal gland, nos', 'ampulla of vater', 'anal canal', 'anterior floor of mouth', 'anterior mediastinum', 'anterior wall of bladder', 'anus, nos', 'aortic body and other paraganglia', 'appendix', 'ascending colon', 'autonomic nervous system, nos', 'base of tongue, nos', 'biliary tract, nos', 'bladder neck', 'bladder, nos', 'blood', 'body of pancreas', 'body of stomach', 'bone marrow', 'bone, nos', 'bones of skull and face and associated joints', 'border of tongue', 'brain stem', 'brain, nos', 'breast, nos', 'cardia, nos', 'cecum', 'cerebellum, nos', 'cerebrum', 'cervix uteri', 'cheek mucosa', 'choroid', 'ciliary body', 'colon, nos', 'connective, subcutaneous and other soft tissues of abdomen', 'connective, subcutaneous and other soft tissues of head, face, and neck', 'connective, subcutaneous and other soft tissues of lower limb and hip', 'connective, subcutaneous and other soft tissues of pelvis', 'connective, subcutaneous and other soft tissues of thorax', 'connective, subcutaneous and other soft tissues of trunk, nos', 'connective, subcutaneous and other soft tissues of upper limb and shoulder', 'connective, subcutaneous and other soft tissues, nos', 'corpus uteri', 'cortex of adrenal gland', 'descending colon', 'dome of bladder', 'duodenum', 'endometrium', 'esophagus, nos', 'external upper lip', 'extrahepatic bile duct', 'eye, nos', 'fallopian tube', 'female genital tract, nos', 'floor of mouth, nos', 'frontal lobe', 'fundus of stomach', 'fundus uteri', 'gallbladder', 'gastric antrum', 'gastrointestinal tract, nos', 'gum, nos', 'hard palate', 'head of pancreas', 'head, face or neck, nos', 'heart', 'hematopoietic system, nos', 'hepatic flexure of colon', 'hypopharynx, nos', 'intestinal tract, nos', 'intra-abdominal lymph nodes', 'intrahepatic bile duct', 'intrathoracic lymph nodes', 'isthmus uteri', 'jejunum', 'kidney, nos', 'larynx, nos', 'lateral wall of bladder', 'lesser curvature of stomach, nos', 'lip, nos', 'liver', 'long bones of lower limb and associated joints', 'long bones of upper limb, scapula and associated joints', 'lower gum', 'lower lobe, lung', 'lower third of esophagus', 'lower-inner quadrant of breast', 'lower-outer quadrant of breast', 'lung, nos', 'lymph node, nos', 'lymph nodes of axilla or arm', 'lymph nodes of head, face and neck', 'lymph nodes of inguinal region or leg', 'main bronchus', 'major salivary gland, nos', 'mandible', 'mediastinum, nos', 'medulla of adrenal gland', 'middle lobe, lung', 'middle third of esophagus', 'mouth, nos', 'myometrium', 'nasal cavity', 'nervous system, nos', 'not reported', 'occipital lobe', 'orbit, nos', 'oropharynx, nos', 'other ill-defined sites', 'ovary', 'overlapping lesion of brain', 'overlapping lesion of breast', 'overlapping lesion of colon', 'overlapping lesion of connective, subcutaneous and other soft tissues', 'overlapping lesion of eye and adnexa', 'overlapping lesion of ill-defined sites', 'overlapping lesion of lip, oral cavity and pharynx', 'overlapping lesion of lung', 'overlapping lesion of nasopharynx', 'overlapping lesion of pancreas', 'overlapping lesion of retroperitoneum and peritoneum', 'palate, nos', 'pancreas, nos', 'parietal lobe', 'pelvic bones, sacrum, coccyx and associated joints', 'pelvic lymph nodes', 'pelvis, nos', 'penis, nos', 'peripheral nerves and autonomic nervous system of abdomen', 'peripheral nerves and autonomic nervous system of head, face, and neck', 'peripheral nerves and autonomic nervous system of pelvis', 'peripheral nerves and autonomic nervous system of thorax', 'peripheral nerves and autonomic nervous system of trunk, nos', 'peripheral nerves and autonomic nervous system of upper limb and shoulder', 'peritoneum, nos', 'pharynx, nos', 'pituitary gland', 'pleura, nos', 'posterior mediastinum', 'posterior wall of bladder', 'posterior wall of oropharynx', 'prostate gland', 'pylorus', 'rectosigmoid junction', 'rectum, nos', 'renal pelvis', 'retromolar area', 'retroperitoneum', 'rib, sternum, clavicle and associated joints', 'short bones of lower limb and associated joints', 'sigmoid colon', 'skin of lower limb and hip', 'skin of scalp and neck', 'skin of trunk', 'skin of upper limb and shoulder', 'skin, nos', 'small intestine, nos', 'specified parts of peritoneum', 'spermatic cord', 'spinal cord', 'spinal meninges', 'spleen', 'splenic flexure of colon', 'stomach, nos', 'submandibular gland', 'supraglottis', 'tail of pancreas', 'temporal lobe', 'testis, nos', 'thoracic esophagus', 'thorax, nos', 'thymus', 'thyroid gland', 'tongue, nos', 'tonsil, nos', 'trachea', 'transverse colon', 'trigone of bladder', 'unknown', 'unknown primary site', 'upper gum', 'upper lobe, lung', 'upper third of esophagus', 'upper-inner quadrant of breast', 'upper-outer quadrant of breast', 'ureter', 'ureteric orifice', 'urethra', 'urinary system, nos', 'uterus, nos', 'vagina, nos', 'ventral surface of tongue, nos', 'vertebral column', 'vulva, nos']\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import requests\n",
    "\n",
    "# I will use the mappings from \"files\" when filtering in my download command, but for any mappings of \"files\" that begin with \"case\", I might want to explore the full set of options with the \"cases\" endpoint (or go on the TCGA website in cohort builder)\n",
    "def return_all_mappings_for_endpoint(endpoint = \"files\"):  # can replace 'files' with 'cases' or any of the other endpoints here: https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/\n",
    "    # Define the endpoint\n",
    "    endpoint_full = f\"https://api.gdc.cancer.gov/{endpoint}/_mapping\"  # can replace 'files' with 'cases' or any of the other endpoints here: https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/\n",
    "\n",
    "    # Make the GET request\n",
    "    response = requests.get(endpoint_full)\n",
    "\n",
    "    # Check the response\n",
    "    if response.status_code == 200:\n",
    "        fields = response.json()[\"fields\"]\n",
    "        return fields\n",
    "    else:\n",
    "        print(f\"Error: {response.status_code}, {response.text}\")\n",
    "        return None\n",
    "\n",
    "\n",
    "def return_all_options_for_a_mapping(mapping, endpoint = \"files\"):  # can replace 'files' with 'cases' or any of the other endpoints here: https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/\n",
    "    # Define the endpoint\n",
    "    endpoint_full = f\"https://api.gdc.cancer.gov/{endpoint}\"  # can replace 'files' with 'cases' or any of the other endpoints here: https://docs.gdc.cancer.gov/API/Users_Guide/Getting_Started/\n",
    "\n",
    "    # Define the parameters\n",
    "    params = {\n",
    "        \"facets\": mapping,\n",
    "        \"size\": 0  # No actual data retrieval, just the facet counts\n",
    "    }\n",
    "\n",
    "    # Make the request\n",
    "    response = requests.get(endpoint_full, params=params)\n",
    "\n",
    "    # Check the response\n",
    "    if response.status_code == 200:\n",
    "        # Extract and display unique values\n",
    "        buckets = response.json()[\"data\"][\"aggregations\"][mapping][\"buckets\"]\n",
    "        mapping_values = sorted([bucket[\"key\"] for bucket in buckets])\n",
    "        return mapping_values\n",
    "    else:\n",
    "        print(f\"Error: {response.status_code}, {response.text}\")\n",
    "        return None\n",
    "\n",
    "all_cases_mappings = return_all_mappings_for_endpoint(endpoint = \"cases\")  # one of these values is diagnoses.tissue_or_organ_of_origin \n",
    "tissue_or_organ_of_origin_all_values = return_all_options_for_a_mapping(mapping = \"diagnoses.tissue_or_organ_of_origin\", endpoint = \"cases\")\n",
    "print(tissue_or_organ_of_origin_all_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kidney                                                         129\n",
      "Breast                                                         113\n",
      "Bronchus and lung                                              109\n",
      "Thyroid gland                                                   59\n",
      "Liver and intrahepatic bile ducts                               57\n",
      "Prostate gland                                                  52\n",
      "Stomach                                                         33\n",
      "Colon                                                           25\n",
      "Bladder                                                         19\n",
      "Other and ill-defined sites in lip, oral cavity and pharynx     13\n",
      "Esophagus                                                       13\n",
      "Other and unspecified parts of tongue                           13\n",
      "Larynx                                                          11\n",
      "Pancreas                                                         4\n",
      "Rectosigmoid junction                                            4\n",
      "Floor of mouth                                                   3\n",
      "Adrenal gland                                                    3\n",
      "Cervix uteri                                                     3\n",
      "Base of tongue                                                   2\n",
      "Thymus                                                           2\n",
      "Other and unspecified parts of mouth                             1\n",
      "Connective, subcutaneous and other soft tissues                  1\n",
      "Other and unspecified parts of biliary tract                     1\n",
      "Rectum                                                           1\n",
      "Retroperitoneum and peritoneum                                   1\n",
      "Name: primary_site, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', None)  # Show all rows\n",
    "pd.set_option('display.max_columns', None)  # Show all columns\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_colwidth', 50)\n",
    "\n",
    "tumor_normal_metadata = \"tcga/matched_tumor_normal_metadata_RNA-Seq.tsv\"\n",
    "tumor_normal_metadata_df = pd.read_csv(tumor_normal_metadata, sep = \"\\t\")\n",
    "\n",
    "# Count occurrences of each value in the 'primary_site' series\n",
    "counts = tumor_normal_metadata_df['primary_site'].value_counts()\n",
    "print(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                         primary_site                                  primary_diagnosis  count\n",
      "5                                              Breast                   Infiltrating duct carcinoma, NOS     90\n",
      "35                                             Kidney                     Clear cell adenocarcinoma, NOS     72\n",
      "49                                     Prostate gland                                Adenocarcinoma, NOS     51\n",
      "41                  Liver and intrahepatic bile ducts                      Hepatocellular carcinoma, NOS     49\n",
      "62                                      Thyroid gland                      Papillary adenocarcinoma, NOS     49\n",
      "19                                  Bronchus and lung                       Squamous cell carcinoma, NOS     46\n",
      "13                                  Bronchus and lung                                Adenocarcinoma, NOS     42\n",
      "36                                             Kidney                      Papillary adenocarcinoma, NOS     32\n",
      "37                                             Kidney             Renal cell carcinoma, chromophobe type     25\n",
      "25                                              Colon                                Adenocarcinoma, NOS     24\n",
      "54                                            Stomach                                Adenocarcinoma, NOS     18\n",
      "3                                             Bladder                        Transitional cell carcinoma     18\n",
      "42  Other and ill-defined sites in lip, oral cavit...                       Squamous cell carcinoma, NOS     13\n",
      "45              Other and unspecified parts of tongue                       Squamous cell carcinoma, NOS     12\n",
      "38                                             Larynx                       Squamous cell carcinoma, NOS     11\n",
      "4                                              Breast            Infiltrating duct and lobular carcinoma      9\n",
      "28                                          Esophagus                                Adenocarcinoma, NOS      9\n",
      "64                                      Thyroid gland            Papillary carcinoma, follicular variant      7\n",
      "8                                              Breast                             Lobular carcinoma, NOS      7\n",
      "39                  Liver and intrahepatic bile ducts                                 Cholangiocarcinoma      7\n",
      "55                                            Stomach                    Adenocarcinoma, intestinal type      5\n",
      "56                                            Stomach                            Carcinoma, diffuse type      5\n",
      "17                                  Bronchus and lung                            Mucinous adenocarcinoma      5\n",
      "51                              Rectosigmoid junction                                Adenocarcinoma, NOS      4\n",
      "12                                  Bronchus and lung                 Adenocarcinoma with mixed subtypes      4\n",
      "63                                      Thyroid gland                 Papillary carcinoma, columnar cell      3\n",
      "0                                       Adrenal gland                              Pheochromocytoma, NOS      3\n",
      "18                                  Bronchus and lung                      Papillary adenocarcinoma, NOS      3\n",
      "15                                  Bronchus and lung        Bronchiolo-alveolar carcinoma, non-mucinous      3\n",
      "59                                            Stomach                             Tubular adenocarcinoma      3\n",
      "47                                           Pancreas                                Adenocarcinoma, NOS      2\n",
      "48                                           Pancreas                   Infiltrating duct carcinoma, NOS      2\n",
      "20                                  Bronchus and lung         Squamous cell carcinoma, keratinizing, NOS      2\n",
      "9                                              Breast                           Medullary carcinoma, NOS      2\n",
      "14                                  Bronchus and lung                   Basaloid squamous cell carcinoma      2\n",
      "33                                     Floor of mouth                       Squamous cell carcinoma, NOS      2\n",
      "1                                      Base of tongue                       Squamous cell carcinoma, NOS      2\n",
      "6                                              Breast  Infiltrating duct mixed with other types of ca...      2\n",
      "2                                             Bladder              Papillary transitional cell carcinoma      1\n",
      "7                                              Breast  Infiltrating lobular mixed with other types of...      1\n",
      "57                                            Stomach                            Mucinous adenocarcinoma      1\n",
      "58                                            Stomach                      Papillary adenocarcinoma, NOS      1\n",
      "60                                             Thymus                        Thymoma, type B1, malignant      1\n",
      "61                                             Thymus                        Thymoma, type B2, malignant      1\n",
      "34                                     Floor of mouth         Squamous cell carcinoma, keratinizing, NOS      1\n",
      "40                  Liver and intrahepatic bile ducts  Combined hepatocellular carcinoma and cholangi...      1\n",
      "26                                              Colon                            Mucinous adenocarcinoma      1\n",
      "27    Connective, subcutaneous and other soft tissues                                Leiomyosarcoma, NOS      1\n",
      "29                                          Esophagus                   Basaloid squamous cell carcinoma      1\n",
      "30                                          Esophagus                            Mucinous adenocarcinoma      1\n",
      "31                                          Esophagus                       Squamous cell carcinoma, NOS      1\n",
      "32                                          Esophagus         Squamous cell carcinoma, keratinizing, NOS      1\n",
      "43       Other and unspecified parts of biliary tract                                 Cholangiocarcinoma      1\n",
      "44               Other and unspecified parts of mouth         Squamous cell carcinoma, keratinizing, NOS      1\n",
      "46              Other and unspecified parts of tongue         Squamous cell carcinoma, keratinizing, NOS      1\n",
      "21                                  Bronchus and lung  Squamous cell carcinoma, small cell, nonkerati...      1\n",
      "22                                       Cervix uteri                                Adenocarcinoma, NOS      1\n",
      "23                                       Cervix uteri                       Squamous cell carcinoma, NOS      1\n",
      "24                                       Cervix uteri         Squamous cell carcinoma, keratinizing, NOS      1\n",
      "10                                             Breast                         Metaplastic carcinoma, NOS      1\n",
      "11                                             Breast  Paget disease and infiltrating duct carcinoma ...      1\n",
      "16                                  Bronchus and lung                     Clear cell adenocarcinoma, NOS      1\n",
      "50                                     Prostate gland                   Infiltrating duct carcinoma, NOS      1\n",
      "52                                             Rectum                                Adenocarcinoma, NOS      1\n",
      "53                     Retroperitoneum and peritoneum                       Dedifferentiated liposarcoma      1\n"
     ]
    }
   ],
   "source": [
    "pair_counts = tumor_normal_metadata_df.groupby(['primary_site', 'primary_diagnosis']).size().reset_index(name='count').sort_values(by='count', ascending=False)\n",
    "print(pair_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                          primary_site                                  primary_diagnosis ajcc_pathologic_stage_simplified  count\n",
      "10                                              Breast                   Infiltrating duct carcinoma, NOS                                2     53\n",
      "108                                      Thyroid gland                      Papillary adenocarcinoma, NOS                                1     31\n",
      "58                                              Kidney                     Clear cell adenocarcinoma, NOS                                1     25\n",
      "38                                   Bronchus and lung                       Squamous cell carcinoma, NOS                                1     24\n",
      "25                                   Bronchus and lung                                Adenocarcinoma, NOS                                1     23\n",
      "61                                              Kidney                     Clear cell adenocarcinoma, NOS                                4     20\n",
      "75                   Liver and intrahepatic bile ducts                      Hepatocellular carcinoma, NOS                                1     18\n",
      "11                                              Breast                   Infiltrating duct carcinoma, NOS                                3     18\n",
      "39                                   Bronchus and lung                       Squamous cell carcinoma, NOS                                2     17\n",
      "60                                              Kidney                     Clear cell adenocarcinoma, NOS                                3     16\n",
      "9                                               Breast                   Infiltrating duct carcinoma, NOS                                1     16\n",
      "62                                              Kidney                      Papillary adenocarcinoma, NOS                                1     15\n",
      "64                                              Kidney                      Papillary adenocarcinoma, NOS                                3     12\n",
      "46                                               Colon                                Adenocarcinoma, NOS                                2     12\n",
      "77                   Liver and intrahepatic bile ducts                      Hepatocellular carcinoma, NOS                                3     12\n",
      "59                                              Kidney                     Clear cell adenocarcinoma, NOS                                2     11\n",
      "96                                             Stomach                                Adenocarcinoma, NOS                                2     11\n",
      "76                   Liver and intrahepatic bile ducts                      Hepatocellular carcinoma, NOS                                2     10\n",
      "66                                              Kidney             Renal cell carcinoma, chromophobe type                                1     10\n",
      "27                                   Bronchus and lung                                Adenocarcinoma, NOS                                3      8\n",
      "5                                              Bladder                        Transitional cell carcinoma                                4      8\n",
      "26                                   Bronchus and lung                                Adenocarcinoma, NOS                                2      8\n",
      "67                                              Kidney             Renal cell carcinoma, chromophobe type                                2      8\n",
      "109                                      Thyroid gland                      Papillary adenocarcinoma, NOS                                2      7\n",
      "110                                      Thyroid gland                      Papillary adenocarcinoma, NOS                                3      7\n",
      "80   Other and ill-defined sites in lip, oral cavit...                       Squamous cell carcinoma, NOS                                2      6\n",
      "4                                              Bladder                        Transitional cell carcinoma                                3      6\n",
      "72                   Liver and intrahepatic bile ducts                                 Cholangiocarcinoma                                1      6\n",
      "71                                              Larynx                       Squamous cell carcinoma, NOS                                4      6\n",
      "16                                              Breast                             Lobular carcinoma, NOS                                2      5\n",
      "70                                              Larynx                       Squamous cell carcinoma, NOS                                2      5\n",
      "87               Other and unspecified parts of tongue                       Squamous cell carcinoma, NOS                                3      5\n",
      "7                                               Breast            Infiltrating duct and lobular carcinoma                                2      5\n",
      "88               Other and unspecified parts of tongue                       Squamous cell carcinoma, NOS                                4      4\n",
      "114                                      Thyroid gland            Papillary carcinoma, follicular variant                                1      4\n",
      "82   Other and ill-defined sites in lip, oral cavit...                       Squamous cell carcinoma, NOS                                4      4\n",
      "65                                              Kidney                      Papillary adenocarcinoma, NOS                                4      4\n",
      "69                                              Kidney             Renal cell carcinoma, chromophobe type                                4      4\n",
      "50                                           Esophagus                                Adenocarcinoma, NOS                                1      4\n",
      "40                                   Bronchus and lung                       Squamous cell carcinoma, NOS                                3      4\n",
      "48                                               Colon                                Adenocarcinoma, NOS                                4      4\n",
      "45                                               Colon                                Adenocarcinoma, NOS                                1      4\n",
      "111                                      Thyroid gland                      Papillary adenocarcinoma, NOS                                4      4\n",
      "3                                              Bladder                        Transitional cell carcinoma                                2      4\n",
      "97                                             Stomach                                Adenocarcinoma, NOS                                3      3\n",
      "95                                             Stomach                                Adenocarcinoma, NOS                                1      3\n",
      "34                                   Bronchus and lung                            Mucinous adenocarcinoma                                2      3\n",
      "47                                               Colon                                Adenocarcinoma, NOS                                3      3\n",
      "115                                      Thyroid gland            Papillary carcinoma, follicular variant                                3      3\n",
      "8                                               Breast            Infiltrating duct and lobular carcinoma                                3      3\n",
      "31                                   Bronchus and lung        Bronchiolo-alveolar carcinoma, non-mucinous                                1      3\n",
      "102                                            Stomach                            Carcinoma, diffuse type                                2      3\n",
      "68                                              Kidney             Renal cell carcinoma, chromophobe type                                3      3\n",
      "92                               Rectosigmoid junction                                Adenocarcinoma, NOS                                1      2\n",
      "52                                           Esophagus                                Adenocarcinoma, NOS                                3      2\n",
      "51                                           Esophagus                                Adenocarcinoma, NOS                                2      2\n",
      "93                               Rectosigmoid junction                                Adenocarcinoma, NOS                                3      2\n",
      "99                                             Stomach                    Adenocarcinoma, intestinal type                                2      2\n",
      "91                                            Pancreas                   Infiltrating duct carcinoma, NOS                                2      2\n",
      "86               Other and unspecified parts of tongue                       Squamous cell carcinoma, NOS                                2      2\n",
      "90                                            Pancreas                                Adenocarcinoma, NOS                                2      2\n",
      "12                                              Breast                   Infiltrating duct carcinoma, NOS                                4      2\n",
      "37                                   Bronchus and lung                      Papillary adenocarcinoma, NOS                                3      2\n",
      "23                                   Bronchus and lung                 Adenocarcinoma with mixed subtypes                                2      2\n",
      "17                                              Breast                             Lobular carcinoma, NOS                                3      2\n",
      "81   Other and ill-defined sites in lip, oral cavit...                       Squamous cell carcinoma, NOS                                3      2\n",
      "101                                            Stomach                    Adenocarcinoma, intestinal type                                4      2\n",
      "28                                   Bronchus and lung                                Adenocarcinoma, NOS                                4      2\n",
      "113                                      Thyroid gland                 Papillary carcinoma, columnar cell                                3      2\n",
      "103                                            Stomach                            Carcinoma, diffuse type                                3      2\n",
      "1                                       Base of tongue                       Squamous cell carcinoma, NOS                                3      1\n",
      "2                                              Bladder              Papillary transitional cell carcinoma                                3      1\n",
      "6                                               Breast            Infiltrating duct and lobular carcinoma                                1      1\n",
      "105                                            Stomach                      Papillary adenocarcinoma, NOS                                1      1\n",
      "106                                            Stomach                             Tubular adenocarcinoma                                1      1\n",
      "107                                            Stomach                             Tubular adenocarcinoma                                2      1\n",
      "112                                      Thyroid gland                 Papillary carcinoma, columnar cell                                1      1\n",
      "100                                            Stomach                    Adenocarcinoma, intestinal type                                3      1\n",
      "104                                            Stomach                            Mucinous adenocarcinoma                                2      1\n",
      "0                                       Base of tongue                       Squamous cell carcinoma, NOS                                2      1\n",
      "57                                      Floor of mouth                       Squamous cell carcinoma, NOS                                4      1\n",
      "63                                              Kidney                      Papillary adenocarcinoma, NOS                                2      1\n",
      "29                                   Bronchus and lung                   Basaloid squamous cell carcinoma                                1      1\n",
      "30                                   Bronchus and lung                   Basaloid squamous cell carcinoma                                3      1\n",
      "32                                   Bronchus and lung                     Clear cell adenocarcinoma, NOS                                3      1\n",
      "18                                              Breast                           Medullary carcinoma, NOS                                1      1\n",
      "19                                              Breast                           Medullary carcinoma, NOS                                2      1\n",
      "20                                              Breast                         Metaplastic carcinoma, NOS                                2      1\n",
      "21                                              Breast  Paget disease and infiltrating duct carcinoma ...                                3      1\n",
      "22                                   Bronchus and lung                 Adenocarcinoma with mixed subtypes                                1      1\n",
      "24                                   Bronchus and lung                 Adenocarcinoma with mixed subtypes                                3      1\n",
      "73                   Liver and intrahepatic bile ducts                                 Cholangiocarcinoma                                2      1\n",
      "74                   Liver and intrahepatic bile ducts  Combined hepatocellular carcinoma and cholangi...                                2      1\n",
      "78                   Liver and intrahepatic bile ducts                      Hepatocellular carcinoma, NOS                                4      1\n",
      "79   Other and ill-defined sites in lip, oral cavit...                       Squamous cell carcinoma, NOS                                1      1\n",
      "83        Other and unspecified parts of biliary tract                                 Cholangiocarcinoma                                4      1\n",
      "84                Other and unspecified parts of mouth         Squamous cell carcinoma, keratinizing, NOS                                4      1\n",
      "85               Other and unspecified parts of tongue                       Squamous cell carcinoma, NOS                                1      1\n",
      "13                                              Breast  Infiltrating duct mixed with other types of ca...                                1      1\n",
      "14                                              Breast  Infiltrating duct mixed with other types of ca...                                3      1\n",
      "15                                              Breast  Infiltrating lobular mixed with other types of...                                1      1\n",
      "89               Other and unspecified parts of tongue         Squamous cell carcinoma, keratinizing, NOS                                4      1\n",
      "53                                           Esophagus                   Basaloid squamous cell carcinoma                                1      1\n",
      "54                                           Esophagus                       Squamous cell carcinoma, NOS                                1      1\n",
      "55                                           Esophagus         Squamous cell carcinoma, keratinizing, NOS                                2      1\n",
      "56                                      Floor of mouth                       Squamous cell carcinoma, NOS                                2      1\n",
      "33                                   Bronchus and lung                            Mucinous adenocarcinoma                                1      1\n",
      "35                                   Bronchus and lung                            Mucinous adenocarcinoma                                3      1\n",
      "36                                   Bronchus and lung                      Papillary adenocarcinoma, NOS                                1      1\n",
      "41                                   Bronchus and lung                       Squamous cell carcinoma, NOS                                4      1\n",
      "42                                   Bronchus and lung         Squamous cell carcinoma, keratinizing, NOS                                1      1\n",
      "43                                   Bronchus and lung         Squamous cell carcinoma, keratinizing, NOS                                3      1\n",
      "44                                   Bronchus and lung  Squamous cell carcinoma, small cell, nonkerati...                                1      1\n",
      "49                                               Colon                            Mucinous adenocarcinoma                                3      1\n",
      "94                                              Rectum                                Adenocarcinoma, NOS                                4      1\n",
      "98                                             Stomach                                Adenocarcinoma, NOS                                4      1\n"
     ]
    }
   ],
   "source": [
    "stage_dict = {\n",
    "    \"Stage I\": \"1\",\n",
    "    \"Stage IA\": \"1\",\n",
    "    \"Stage IB\": \"1\",\n",
    "    \"Stage IC\": \"1\",\n",
    "    \"Stage II\": \"2\",\n",
    "    \"Stage IIA\": \"2\",\n",
    "    \"Stage IIB\": \"2\",\n",
    "    \"Stage IIC\": \"2\",\n",
    "    \"Stage III\": \"3\",\n",
    "    \"Stage IIIA\": \"3\",\n",
    "    \"Stage IIIB\": \"3\",\n",
    "    \"Stage IIIC\": \"3\",\n",
    "    \"Stage IV\": \"4\",\n",
    "    \"Stage IVA\": \"4\",\n",
    "    \"Stage IVB\": \"4\",\n",
    "    \"Stage IVC\": \"4\",\n",
    "}\n",
    "\n",
    "tumor_normal_metadata_df['ajcc_pathologic_stage_simplified'] = tumor_normal_metadata_df['ajcc_pathologic_stage'].map(\n",
    "    lambda x: stage_dict.get(x, x)\n",
    ")\n",
    "\n",
    "trio_counts = tumor_normal_metadata_df.groupby(['primary_site', 'primary_diagnosis', 'ajcc_pathologic_stage_simplified']).size().reset_index(name='count').sort_values(by='count', ascending=False)\n",
    "print(trio_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'strand': 1,\n",
       "  'transcript_consequences': [{'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 184,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 185,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'hgvsp': 'ENSP00000312236.5:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000352993',\n",
       "    'hgvsc': 'ENST00000352993.7:c.66dup'},\n",
       "   {'transcript_id': 'ENST00000354071',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000354071.8:n.130dup',\n",
       "    'strand': -1,\n",
       "    'impact': 'MODIFIER',\n",
       "    'cdna_start': 129,\n",
       "    'cdna_end': 130,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'biotype': 'retained_intron',\n",
       "    'variant_allele': 'T',\n",
       "    'consequence_terms': ['non_coding_transcript_exon_variant']},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000356906',\n",
       "    'distance': 1388,\n",
       "    'strand': 1,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'consequence_terms': ['upstream_gene_variant']},\n",
       "   {'cdna_start': 178,\n",
       "    'impact': 'HIGH',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 179,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000357654',\n",
       "    'hgvsc': 'ENST00000357654.9:c.66dup',\n",
       "    'hgvsp': 'ENSP00000350283.3:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22},\n",
       "   {'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'strand': 1,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000460115',\n",
       "    'distance': 1578},\n",
       "   {'cdna_end': 192,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000461221.5:c.66dup',\n",
       "    'transcript_id': 'ENST00000461221',\n",
       "    'hgvsp': 'ENSP00000418548.1:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 191,\n",
       "    'consequence_terms': ['frameshift_variant', 'NMD_transcript_variant'],\n",
       "    'biotype': 'nonsense_mediated_decay',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'cdna_start': 184,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'cds_start': 65,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 185,\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000417241.2:p.Glu23ArgfsTer18',\n",
       "    'hgvsc': 'ENST00000461574.2:c.66dup',\n",
       "    'transcript_id': 'ENST00000461574',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'cdna_start': 178,\n",
       "    'impact': 'HIGH',\n",
       "    'consequence_terms': ['frameshift_variant', 'NMD_transcript_variant'],\n",
       "    'biotype': 'nonsense_mediated_decay',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000461798.5:c.66dup',\n",
       "    'transcript_id': 'ENST00000461798',\n",
       "    'hgvsp': 'ENSP00000417988.1:p.Glu23ArgfsTer18',\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 179,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1'},\n",
       "   {'distance': 1612,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000467245',\n",
       "    'strand': 1,\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691'},\n",
       "   {'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000468300.5:c.66dup',\n",
       "    'transcript_id': 'ENST00000468300',\n",
       "    'hgvsp': 'ENSP00000417148.1:p.Glu23ArgfsTer18',\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 260,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 259,\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding'},\n",
       "   {'cdna_start': 246,\n",
       "    'impact': 'HIGH',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 247,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000470026.6:c.66dup',\n",
       "    'transcript_id': 'ENST00000470026',\n",
       "    'hgvsp': 'ENSP00000419274.2:p.Glu23ArgfsTer18',\n",
       "    'protein_end': 22,\n",
       "    'strand': -1},\n",
       "   {'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 298,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'hgvsp': 'ENSP00000418960.2:p.Glu23ArgfsTer18',\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000471181.7:c.66dup',\n",
       "    'transcript_id': 'ENST00000471181',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 297,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'hgvsp': 'ENSP00000420201.2:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000473961',\n",
       "    'hgvsc': 'ENST00000473961.6:c.66dup',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 179,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 178,\n",
       "    'impact': 'HIGH'},\n",
       "   {'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 175,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65,\n",
       "    'transcript_id': 'ENST00000476777',\n",
       "    'hgvsc': 'ENST00000476777.6:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'hgvsp': 'ENSP00000417554.2:p.Glu23ArgfsTer18',\n",
       "    'cdna_start': 174,\n",
       "    'impact': 'HIGH',\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 179,\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000419988.2:p.Glu23ArgfsTer18',\n",
       "    'transcript_id': 'ENST00000477152',\n",
       "    'hgvsc': 'ENST00000477152.6:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'cdna_start': 178,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'hgvsc': 'ENST00000478531.6:c.66dup',\n",
       "    'transcript_id': 'ENST00000478531',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'hgvsp': 'ENSP00000420412.2:p.Glu23ArgfsTer18',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 168,\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'cdna_start': 167,\n",
       "    'impact': 'HIGH',\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant']},\n",
       "   {'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 158,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'cds_start': 65,\n",
       "    'hgvsc': 'ENST00000484087.6:c.66dup',\n",
       "    'transcript_id': 'ENST00000484087',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000419481.2:p.Glu23ArgfsTer18',\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 157,\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 224,\n",
       "    'impact': 'HIGH',\n",
       "    'hgvsp': 'ENSP00000420781.2:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000489037',\n",
       "    'hgvsc': 'ENST00000489037.2:c.66dup',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 225,\n",
       "    'codons': 'tta/ttAa'},\n",
       "   {'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 165,\n",
       "    'protein_end': 22,\n",
       "    'strand': -1,\n",
       "    'hgvsp': 'ENSP00000420705.2:p.Glu23ArgfsTer18',\n",
       "    'transcript_id': 'ENST00000491747',\n",
       "    'hgvsc': 'ENST00000491747.6:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'cdna_start': 164,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'cdna_start': 114,\n",
       "    'impact': 'HIGH',\n",
       "    'consequence_terms': ['frameshift_variant', 'NMD_transcript_variant'],\n",
       "    'biotype': 'nonsense_mediated_decay',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 115,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000492859.5:c.66dup',\n",
       "    'transcript_id': 'ENST00000492859',\n",
       "    'hgvsp': 'ENSP00000420253.1:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'consequence_terms': ['5_prime_UTR_variant'],\n",
       "    'cdna_start': 210,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 211,\n",
       "    'strand': -1,\n",
       "    'hgvsc': 'ENST00000493795.5:c.-22dup',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000493795'},\n",
       "   {'cdna_start': 210,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 211,\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'protein_coding',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'consequence_terms': ['5_prime_UTR_variant'],\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000493919.6:c.-22dup',\n",
       "    'transcript_id': 'ENST00000493919',\n",
       "    'strand': -1},\n",
       "   {'cdna_start': 258,\n",
       "    'impact': 'HIGH',\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 259,\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'cds_start': 65,\n",
       "    'hgvsc': 'ENST00000494123.6:c.66dup',\n",
       "    'transcript_id': 'ENST00000494123',\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000419103.2:p.Glu23ArgfsTer18'},\n",
       "   {'transcript_id': 'ENST00000497488',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000497488.2:c.-219+1240dup',\n",
       "    'strand': -1,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T',\n",
       "    'consequence_terms': ['intron_variant']},\n",
       "   {'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['intron_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'strand': -1,\n",
       "    'transcript_id': 'ENST00000586385',\n",
       "    'hgvsc': 'ENST00000586385.5:c.4+1151dup',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'strand': -1,\n",
       "    'transcript_id': 'ENST00000591534',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000591534.5:c.-44+1240dup',\n",
       "    'consequence_terms': ['intron_variant'],\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'impact': 'MODIFIER'},\n",
       "   {'transcript_id': 'ENST00000591849',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000591849.5:c.-99+1240dup',\n",
       "    'strand': -1,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['intron_variant'],\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T'},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 712,\n",
       "    'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'transcript_id': 'ENST00000618469',\n",
       "    'hgvsc': 'ENST00000618469.2:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000478114.2:p.Glu23ArgfsTer18',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 713,\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'cdna_start': 204,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 205,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000489431.2:p.Glu23ArgfsTer18',\n",
       "    'transcript_id': 'ENST00000634433',\n",
       "    'hgvsc': 'ENST00000634433.2:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22},\n",
       "   {'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 152,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000642945',\n",
       "    'hgvsc': 'ENST00000642945.1:c.66dup',\n",
       "    'hgvsp': 'ENSP00000495897.1:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'cdna_start': 151,\n",
       "    'impact': 'HIGH',\n",
       "    'consequence_terms': ['frameshift_variant', 'NMD_transcript_variant'],\n",
       "    'biotype': 'nonsense_mediated_decay',\n",
       "    'gene_symbol_source': 'HGNC'},\n",
       "   {'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 178,\n",
       "    'impact': 'HIGH',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'cds_start': 65,\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'amino_acids': 'L/LX',\n",
       "    'cdna_end': 179,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'hgvsp': 'ENSP00000496570.2:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'protein_start': 22,\n",
       "    'cds_end': 66,\n",
       "    'hgvsc': 'ENST00000644379.2:c.66dup',\n",
       "    'transcript_id': 'ENST00000644379'},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000644555.2:c.-221dup',\n",
       "    'transcript_id': 'ENST00000644555',\n",
       "    'strand': -1,\n",
       "    'cdna_end': 185,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_start': 184,\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['5_prime_UTR_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100'},\n",
       "   {'biotype': 'protein_coding',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'consequence_terms': ['5_prime_UTR_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'cdna_start': 291,\n",
       "    'cdna_end': 292,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'strand': -1,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000652672.2:c.-195dup',\n",
       "    'transcript_id': 'ENST00000652672'},\n",
       "   {'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'lncRNA',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'strand': 1,\n",
       "    'transcript_id': 'ENST00000657841',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'distance': 1519},\n",
       "   {'hgvsp': 'ENSP00000514849.1:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000700182',\n",
       "    'hgvsc': 'ENST00000700182.1:c.66dup',\n",
       "    'cds_start': 65,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 158,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding',\n",
       "    'cdna_start': 157,\n",
       "    'impact': 'HIGH'},\n",
       "   {'biotype': 'nonsense_mediated_decay',\n",
       "    'consequence_terms': ['frameshift_variant', 'NMD_transcript_variant'],\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 157,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'cds_start': 65,\n",
       "    'codons': 'tta/ttAa',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 158,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'hgvsp': 'ENSP00000514850.1:p.Glu23ArgfsTer18',\n",
       "    'transcript_id': 'ENST00000700183',\n",
       "    'hgvsc': 'ENST00000700183.1:c.66dup',\n",
       "    'cds_end': 66,\n",
       "    'protein_start': 22},\n",
       "   {'cdna_start': 308,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 309,\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'biotype': 'retained_intron',\n",
       "    'variant_allele': 'T',\n",
       "    'consequence_terms': ['non_coding_transcript_exon_variant'],\n",
       "    'transcript_id': 'ENST00000700184',\n",
       "    'hgvsc': 'ENST00000700184.1:n.309dup',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'strand': -1},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000700185.1:n.185dup',\n",
       "    'transcript_id': 'ENST00000700185',\n",
       "    'strand': -1,\n",
       "    'cdna_start': 184,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 185,\n",
       "    'biotype': 'protein_coding_CDS_not_defined',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'consequence_terms': ['non_coding_transcript_exon_variant']},\n",
       "   {'impact': 'MODIFIER',\n",
       "    'cdna_start': 184,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 185,\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'biotype': 'protein_coding_CDS_not_defined',\n",
       "    'variant_allele': 'T',\n",
       "    'consequence_terms': ['non_coding_transcript_exon_variant'],\n",
       "    'transcript_id': 'ENST00000700186',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'hgvsc': 'ENST00000700186.1:n.185dup',\n",
       "    'strand': -1},\n",
       "   {'cds_end': 66,\n",
       "    'protein_start': 22,\n",
       "    'transcript_id': 'ENST00000713676',\n",
       "    'hgvsc': 'ENST00000713676.1:c.66dup',\n",
       "    'hgvsp': 'ENSP00000518978.1:p.Glu23ArgfsTer18',\n",
       "    'strand': -1,\n",
       "    'protein_end': 22,\n",
       "    'gene_id': 'ENSG00000012048',\n",
       "    'cdna_end': 139,\n",
       "    'amino_acids': 'L/LX',\n",
       "    'codons': 'tta/ttAa',\n",
       "    'cds_start': 65,\n",
       "    'hgnc_id': 'HGNC:1100',\n",
       "    'gene_symbol': 'BRCA1',\n",
       "    'variant_allele': 'T',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'impact': 'HIGH',\n",
       "    'cdna_start': 138,\n",
       "    'consequence_terms': ['frameshift_variant'],\n",
       "    'biotype': 'protein_coding'},\n",
       "   {'distance': 1563,\n",
       "    'transcript_id': 'ENST00000790195',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'strand': 1,\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T'},\n",
       "   {'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'strand': 1,\n",
       "    'transcript_id': 'ENST00000790196',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'distance': 1569},\n",
       "   {'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'lncRNA',\n",
       "    'distance': 1610,\n",
       "    'transcript_id': 'ENST00000790197',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'strand': 1},\n",
       "   {'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'lncRNA',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790198',\n",
       "    'distance': 2340,\n",
       "    'strand': 1},\n",
       "   {'distance': 1555,\n",
       "    'transcript_id': 'ENST00000790199',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'strand': 1,\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T'},\n",
       "   {'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790200',\n",
       "    'distance': 1607,\n",
       "    'strand': 1,\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'lncRNA',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'consequence_terms': ['upstream_gene_variant']},\n",
       "   {'strand': 1,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790201',\n",
       "    'distance': 1535,\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496'},\n",
       "   {'strand': 1,\n",
       "    'distance': 1558,\n",
       "    'transcript_id': 'ENST00000790202',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'variant_allele': 'T',\n",
       "    'biotype': 'lncRNA',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER'},\n",
       "   {'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496',\n",
       "    'strand': 1,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790203',\n",
       "    'distance': 1565},\n",
       "   {'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'distance': 1569,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790204',\n",
       "    'strand': 1},\n",
       "   {'gene_id': 'ENSG00000198496',\n",
       "    'impact': 'MODIFIER',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'distance': 1611,\n",
       "    'transcript_id': 'ENST00000790205',\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'strand': 1},\n",
       "   {'strand': 1,\n",
       "    'gene_symbol_source': 'HGNC',\n",
       "    'transcript_id': 'ENST00000790206',\n",
       "    'distance': 1602,\n",
       "    'biotype': 'lncRNA',\n",
       "    'variant_allele': 'T',\n",
       "    'hgnc_id': 'HGNC:20691',\n",
       "    'gene_symbol': 'NBR2',\n",
       "    'consequence_terms': ['upstream_gene_variant'],\n",
       "    'impact': 'MODIFIER',\n",
       "    'gene_id': 'ENSG00000198496'}],\n",
       "  'minimised': 1,\n",
       "  'colocated_variants': [{'phenotype_or_disease': 1,\n",
       "    'var_synonyms': {'ClinVar': ['RCV000478017',\n",
       "      'RCV000240681',\n",
       "      'RCV000049087',\n",
       "      'RCV000163427',\n",
       "      'RCV000031272',\n",
       "      'VCV000037691',\n",
       "      'RCV001554246',\n",
       "      'RCV000785198',\n",
       "      'RCV000778174']},\n",
       "    'end': 43124031,\n",
       "    'seq_region_name': '17',\n",
       "    'frequencies': {'T': {'gnomade_nfe': 1.8e-06,\n",
       "      'gnomade_sas': 0,\n",
       "      'gnomade_asj': 0,\n",
       "      'gnomade_afr': 0,\n",
       "      'gnomade_mid': 0,\n",
       "      'gnomade_remaining': 0,\n",
       "      'gnomade_amr': 0,\n",
       "      'gnomade_fin': 0,\n",
       "      'gnomade': 1.369e-06,\n",
       "      'gnomade_eas': 0}},\n",
       "    'strand': 1,\n",
       "    'pubmed': [25741868,\n",
       "     24033266,\n",
       "     22703879,\n",
       "     26467025,\n",
       "     26845104,\n",
       "     22006311,\n",
       "     20104584,\n",
       "     16683254,\n",
       "     21203900,\n",
       "     15617999,\n",
       "     11597388,\n",
       "     21324516,\n",
       "     22009639,\n",
       "     23199084,\n",
       "     26295337,\n",
       "     23683081,\n",
       "     24916970,\n",
       "     11179017,\n",
       "     16168118,\n",
       "     18627636,\n",
       "     27257965,\n",
       "     16998791,\n",
       "     15024741,\n",
       "     17688236,\n",
       "     12181777,\n",
       "     21559243,\n",
       "     33643918,\n",
       "     19329713,\n",
       "     22535016,\n",
       "     25863477,\n",
       "     21643751,\n",
       "     22430266,\n",
       "     9145676,\n",
       "     12473589,\n",
       "     15994883,\n",
       "     20301425,\n",
       "     11606101,\n",
       "     10952777,\n",
       "     16267036,\n",
       "     23867111,\n",
       "     8807330,\n",
       "     15383404,\n",
       "     21503673,\n",
       "     7837387,\n",
       "     22516946,\n",
       "     23233716,\n",
       "     8531968,\n",
       "     8533757,\n",
       "     7611277,\n",
       "     7894492,\n",
       "     20950396,\n",
       "     7550349,\n",
       "     8595420,\n",
       "     8642955,\n",
       "     8651293,\n",
       "     8764110,\n",
       "     9042909,\n",
       "     9921907,\n",
       "     10739756,\n",
       "     11352856,\n",
       "     11896106,\n",
       "     16941470,\n",
       "     17307836,\n",
       "     18034184,\n",
       "     18594935,\n",
       "     19147582,\n",
       "     19906413,\n",
       "     20651350,\n",
       "     20711688,\n",
       "     22274685,\n",
       "     22763381,\n",
       "     23086583,\n",
       "     23531862,\n",
       "     23788959,\n",
       "     23855721,\n",
       "     23941904,\n",
       "     24764757,\n",
       "     31454914,\n",
       "     32733560,\n",
       "     35393538,\n",
       "     35464868,\n",
       "     35693198,\n",
       "     28392550,\n",
       "     29907814,\n",
       "     34290354,\n",
       "     33471991,\n",
       "     28176296,\n",
       "     29339979,\n",
       "     27553291,\n",
       "     35220195,\n",
       "     29446198,\n",
       "     29433453,\n",
       "     29470806,\n",
       "     28993434,\n",
       "     28492532,\n",
       "     29176636,\n",
       "     20189727,\n",
       "     32341426,\n",
       "     31492746,\n",
       "     35918668,\n",
       "     29152070],\n",
       "    'id': 'rs80357783',\n",
       "    'allele_string': 'T/TT',\n",
       "    'start': 43124031,\n",
       "    'clin_sig': ['pathogenic']}],\n",
       "  'allele_string': '-/T',\n",
       "  'end': 43124031,\n",
       "  'start': 43124032,\n",
       "  'most_severe_consequence': 'frameshift_variant',\n",
       "  'seq_region_name': '17',\n",
       "  'assembly_name': 'GRCh38',\n",
       "  'input': 'rs80357713',\n",
       "  'id': 'rs80357713'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000346315.3:c.66dup\n",
      "ENST00000351666.3:c.66dup\n",
      "ENST00000352993.3:c.66dup\n",
      "ENST00000354071.3:c.66dup\n",
      "ENST00000357654.3:c.66dup\n",
      "ENST00000461221.1:c.66dup\n",
      "ENST00000461798.1:c.66dup\n",
      "ENST00000468300.1:c.66dup\n",
      "ENST00000470026.1:c.66dup\n",
      "ENST00000471181.2:c.66dup\n",
      "ENST00000476777.1:c.66dup\n",
      "ENST00000477152.1:c.66dup\n",
      "ENST00000478531.1:c.66dup\n",
      "ENST00000489037.1:c.66dup\n",
      "ENST00000491747.2:c.66dup\n",
      "ENST00000492859.1:c.66dup\n",
      "ENST00000494123.1:c.66dup\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "rsid = \"rs80357713\"\n",
    "url = f\"https://grch37.rest.ensembl.org/vep/human/id/{rsid}?hgvs=1\"\n",
    "headers = {\"Content-Type\": \"application/json\"}\n",
    "response = requests.get(url, headers=headers)\n",
    "# print(response.json())\n",
    "\n",
    "for entry in response.json():\n",
    "    for tx in entry.get(\"transcript_consequences\", []):\n",
    "        if \"hgvsc\" in tx:\n",
    "            hgvsc = tx[\"hgvsc\"]\n",
    "            if \":n.\" in hgvsc or \"-\" in hgvsc or \"+\" in hgvsc:\n",
    "                continue\n",
    "            print(hgvsc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'releases': [113]}\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000352993.7:c.66dup\n",
      "ENST00000357654.9:c.66dup\n",
      "ENST00000461221.5:c.66dup\n",
      "ENST00000461574.2:c.66dup\n",
      "ENST00000461798.5:c.66dup\n",
      "ENST00000468300.5:c.66dup\n",
      "ENST00000470026.6:c.66dup\n",
      "ENST00000471181.7:c.66dup\n",
      "ENST00000473961.6:c.66dup\n",
      "ENST00000476777.6:c.66dup\n",
      "ENST00000477152.6:c.66dup\n",
      "ENST00000478531.6:c.66dup\n",
      "ENST00000484087.6:c.66dup\n",
      "ENST00000489037.2:c.66dup\n",
      "ENST00000491747.6:c.66dup\n",
      "ENST00000492859.5:c.66dup\n",
      "ENST00000494123.6:c.66dup\n",
      "ENST00000618469.2:c.66dup\n",
      "ENST00000634433.2:c.66dup\n",
      "ENST00000642945.1:c.66dup\n",
      "ENST00000644379.2:c.66dup\n",
      "ENST00000700182.1:c.66dup\n",
      "ENST00000700183.1:c.66dup\n",
      "ENST00000713676.1:c.66dup\n"
     ]
    }
   ],
   "source": [
    "import requests, sys\n",
    "\n",
    "# # verify release - currently 113 \n",
    "# server = \"https://rest.ensembl.org\"\n",
    "# ext = \"/info/data/?\"\n",
    " \n",
    "# r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    " \n",
    "# if not r.ok:\n",
    "#   r.raise_for_status()\n",
    "#   sys.exit()\n",
    " \n",
    "# decoded = r.json()\n",
    "# print(repr(decoded))\n",
    "\n",
    "rsid = \"rs80357713\"\n",
    "url = f\"https://rest.ensembl.org/vep/human/id/{rsid}?hgvs=1\"\n",
    "headers = {\"Content-Type\": \"application/json\"}\n",
    "response = requests.get(url, headers=headers)\n",
    "# print(response.json())\n",
    "\n",
    "for entry in response.json():\n",
    "    for tx in entry.get(\"transcript_consequences\", []):\n",
    "        if \"hgvsc\" in tx:\n",
    "            hgvsc = tx[\"hgvsc\"]\n",
    "            if \":n.\" in hgvsc or \"-\" in hgvsc or \"+\" in hgvsc:\n",
    "                continue\n",
    "            print(hgvsc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'display_name': 'BRCA1-229',\n",
       " 'is_canonical': 0,\n",
       " 'species': 'homo_sapiens',\n",
       " 'gencode_primary': 0,\n",
       " 'id': 'ENST00000642945',\n",
       " 'object_type': 'Transcript',\n",
       " 'logic_name': 'havana_homo_sapiens',\n",
       " 'end': 43125343,\n",
       " 'Parent': 'ENSG00000012048',\n",
       " 'seq_region_name': '17',\n",
       " 'Translation': {'Parent': 'ENST00000642945',\n",
       "  'version': 1,\n",
       "  'db_type': 'core',\n",
       "  'start': 43112495,\n",
       "  'id': 'ENSP00000495897',\n",
       "  'length': 81,\n",
       "  'species': 'homo_sapiens',\n",
       "  'end': 43124096,\n",
       "  'object_type': 'Translation'},\n",
       " 'assembly_name': 'GRCh38',\n",
       " 'db_type': 'core',\n",
       " 'biotype': 'nonsense_mediated_decay',\n",
       " 'version': 1,\n",
       " 'length': 1255,\n",
       " 'Exon': [{'id': 'ENSE00004020662',\n",
       "   'species': 'homo_sapiens',\n",
       "   'end': 43125343,\n",
       "   'object_type': 'Exon',\n",
       "   'strand': -1,\n",
       "   'seq_region_name': '17',\n",
       "   'db_type': 'core',\n",
       "   'version': 2,\n",
       "   'start': 43125277,\n",
       "   'assembly_name': 'GRCh38'},\n",
       "  {'id': 'ENSE00003559512',\n",
       "   'species': 'homo_sapiens',\n",
       "   'end': 43124115,\n",
       "   'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'seq_region_name': '17',\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'start': 43124017,\n",
       "   'assembly_name': 'GRCh38'},\n",
       "  {'object_type': 'Exon',\n",
       "   'strand': -1,\n",
       "   'end': 43115779,\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00003510592',\n",
       "   'start': 43115726,\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'seq_region_name': '17'},\n",
       "  {'assembly_name': 'GRCh38',\n",
       "   'start': 43112487,\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'seq_region_name': '17',\n",
       "   'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'end': 43112606,\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00003824040'},\n",
       "  {'end': 43106533,\n",
       "   'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'id': 'ENSE00004011535',\n",
       "   'species': 'homo_sapiens',\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'start': 43106456,\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'seq_region_name': '17'},\n",
       "  {'end': 43104956,\n",
       "   'object_type': 'Exon',\n",
       "   'strand': -1,\n",
       "   'id': 'ENSE00004011537',\n",
       "   'species': 'homo_sapiens',\n",
       "   'version': 1,\n",
       "   'db_type': 'core',\n",
       "   'start': 43104868,\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'seq_region_name': '17'},\n",
       "  {'seq_region_name': '17',\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'start': 43104122,\n",
       "   'version': 1,\n",
       "   'db_type': 'core',\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00003618767',\n",
       "   'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'end': 43104261},\n",
       "  {'seq_region_name': '17',\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'start': 43099775,\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00004011539',\n",
       "   'object_type': 'Exon',\n",
       "   'strand': -1,\n",
       "   'end': 43099880},\n",
       "  {'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'end': 43097289,\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00003673367',\n",
       "   'start': 43097244,\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'version': 1,\n",
       "   'db_type': 'core',\n",
       "   'seq_region_name': '17'},\n",
       "  {'object_type': 'Exon',\n",
       "   'strand': -1,\n",
       "   'end': 43095922,\n",
       "   'species': 'homo_sapiens',\n",
       "   'id': 'ENSE00003518965',\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'start': 43095846,\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'seq_region_name': '17'},\n",
       "  {'end': 43094860,\n",
       "   'strand': -1,\n",
       "   'object_type': 'Exon',\n",
       "   'id': 'ENSE00003817389',\n",
       "   'species': 'homo_sapiens',\n",
       "   'db_type': 'core',\n",
       "   'version': 1,\n",
       "   'start': 43094482,\n",
       "   'assembly_name': 'GRCh38',\n",
       "   'seq_region_name': '17'}],\n",
       " 'strand': -1,\n",
       " 'source': 'havana',\n",
       " 'start': 43094482}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transcript ID: ENST00000642945\n",
      "Version: 1\n",
      "Genome Assembly: GRCh38\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'release'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 13\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mVersion: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdata[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mversion\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     12\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mGenome Assembly: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdata[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124massembly_name\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 13\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEnsembl Release: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[43mdata\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mrelease\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'release'"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "tx_id = \"ENST00000642945\"\n",
    "url = f\"https://rest.ensembl.org/lookup/id/{tx_id}?expand=1\"\n",
    "headers = {\"Content-Type\": \"application/json\"}\n",
    "\n",
    "r = requests.get(url, headers=headers)\n",
    "data = r.json()\n",
    "\n",
    "print(f\"Transcript ID: {data['id']}\")\n",
    "print(f\"Version: {data['version']}\")\n",
    "print(f\"Genome Assembly: {data['assembly_name']}\")\n",
    "print(f\"Ensembl Release: {data['release']}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RLSRWP_2025",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
